---
title: "geoduck_tagseq_analysis"
author: "Samuel Gurr"
date: "January 29, 2021"
output:
  html_document:
    toc: true
    toc_float: true
---

# Setup: 

```{r setup, include=TRUE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE, cache = TRUE)
```

### Load libraries
```{r load_libraries, include = TRUE}
# load libraries - notes show the install command needed to install (pre installed)
library(DESeq2) # note: this was previously installed with the command `BiocManager::install("DESeq2")`
library(edgeR)
library(goseq)
library(dplyr)
library(GenomicFeatures)
library(data.table)
library(calibrate)

library(affycoretools) # note: this was previously installed with the BiocManager::install("affycoretools")

library(data.table)
library(vsn)
library(tidybulk)
# Plotting
library(ggplot2)
library(cowplot)
library(pheatmap)
library(gplots)
library(RColorBrewer)
library(EnhancedVolcano)  # note: this was previously installed with the command `BiocManager::install("EnhancedVolcano")`
library(pcaExplorer) 

```

### Set working directory
```{r set_wd, include = TRUE}

# SET WORKING DIRECTORY AND LOAD DATA
setwd("C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/")

path_out = 'C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/RAnalysis/DESeq2/output/'
```

# LOAD DATA:

### Gene count data - convert to matrix
```{r count_data, include = TRUE}
getwd()
### filtered counts tables - format matrix after upload [from Count_Matrix_Stats.Filter.R] 
d0.counts_matrix  <- read.csv(file="../Data/filtered_counts/day0.counts.filtered_3cpm50perc.csv", sep=',', header=TRUE) 
d0.counts_matrix <- data.frame(d0.counts_matrix[,-1], row.names=d0.counts_matrix[,1]) 

d7.counts_matrix  <- read.csv(file="../Data/filtered_counts/day7.counts.filtered_3cpm50perc.csv", sep=',', header=TRUE) 
d7.counts_matrix <- data.frame(d7.counts_matrix[,-1], row.names=d7.counts_matrix[,1]) 

d14.counts_matrix <- read.csv(file="../Data/filtered_counts/day14.counts.filtered_3cpm50perc.csv", sep=',', header=TRUE) 
d14.counts_matrix <- data.frame(d14.counts_matrix[,-1], row.names=d14.counts_matrix[,1]) 

d21.counts_matrix <- read.csv(file="../Data/filtered_counts/day21.counts.filtered_3cpm50perc.csv", sep=',', header=TRUE) 
d21.counts_matrix <- data.frame(d21.counts_matrix[,-1], row.names=d21.counts_matrix[,1]) 

```

### Sample metadata - Experimental treatments/groups
```{r experiment_data, include = TRUE}
### experiment metadata [from Count_Matrix_Stats.Filter.R]  - convert characaters to factors for DESeq2
all.exp_data  <- read.csv(file="../Data/filtered_counts/all.exp.data.csv", sep=',', header=TRUE) %>%   mutate_if(is.character, as.factor)

d0.exp_data   <- read.csv(file="../Data/filtered_counts/day0.exp.data.csv", sep=',', header=TRUE) %>%   mutate_if(is.character, as.factor)

d7.exp_data   <- read.csv(file="../Data/filtered_counts/day7.exp.data.csv", sep=',', header=TRUE) %>%   mutate_if(is.character, as.factor)

d14.exp_data  <- read.csv(file="../Data/filtered_counts/day14.exp.data.csv", sep=',', header=TRUE) %>%   mutate_if(is.character, as.factor)

d21.exp_data  <- read.csv(file="../Data/filtered_counts/day21.exp.data.csv", sep=',', header=TRUE) %>%   mutate_if(is.character, as.factor)
```

# edgeR attempt
```{r edgeR_Day0} 
# for edgeR
write.table(read.csv("../Data/filtered_counts/ day0.counts.filtered_matrix.csv", sep=","), "../Data/filtered_counts/ day0.counts.filtered_matrix.tab") # day0 count matrix
d0.counts_matrix_egdeR  <- read.table(file="../Data/filtered_counts/ day0.counts.filtered_matrix.tab")
d0.counts_matrix_egdeR<- data.frame(d0.counts_matrix_egdeR[,-1], row.names=d0.counts_matrix_egdeR[,1]) 

colnames(d0.counts_matrix_egdeR) # same order
d0.exp_data[2] # same order

group <- factor(d0.exp_data$Primary_Treatment)
y <- DGEList(counts=d0.counts_matrix_egdeR, group=group)
y <- calcNormFactors(y)
design <- model.matrix(~group)
y <- estimateDisp(y,design)
# To perform quasi-likelihood F-tests:
fit <- glmQLFit(y,design)
qlf <- glmQLFTest(fit,coef=2)
topTags(qlf)
summary(decideTests(qlf))
FDR <- p.adjust(qlf$table$PValue, method="BH")
sum(FDR < 0.05)
# To perform likelihood ratio tests:
fit <- glmFit(y,design)
lrt <- glmLRT(fit,coef=2)
topTags(lrt)
summary(decideTests(lrt))
# plot
plotMD(qlf)
abline(h=c(-2,2), col="blue")
#Here's a closer look at the individual counts-per-million for the top genes. The top genes are very consistent across the three replicates
top <- rownames(topTags(qlf))
cpm(y)[top,]
```


```{r edgeR_Day7} 
# for edgeR
write.table(read.csv("../Data/filtered_counts/ day7.counts.filtered_matrix.csv", sep=","), "../Data/filtered_counts/ day7.counts.filtered_matrix.tab") # day0 count matrix
d7.counts_matrix_egdeR  <- read.table(file="../Data/filtered_counts/ day7.counts.filtered_matrix.tab")
d7.counts_matrix_egdeR<- data.frame(d7.counts_matrix_egdeR[,-1], row.names=d7.counts_matrix_egdeR[,1]) 

colnames(d7.counts_matrix_egdeR) # same order
d7.exp_data[2] # same order



# Prim.Treat <- factor(d7.exp_data$Primary_Treatment)
# # Prim.Treat <- relevel(Prim.Treat, ref="A")
# Sec.Treat <- factor(d7.exp_data$Second_Treament)

Group <- factor(d7.exp_data$All_Treatment)
design <- model.matrix(~0+Group)
colnames(design) <- levels(Group)
y <- DGEList(counts=d7.counts_matrix_egdeR, group=Group) 
y <- estimateDisp(y,design) # get dispersion value
# y$samples
fit <- glmQLFit(y, design)
colnames(fit)
# Then we can make any comparisons we wish. For example, we might wish to make the following
# contrasts:
my.contrasts <- makeContrasts(
 P.AvM = (AM+AA+AS)-(MM+MA+MS),
# Drug.2vs0 = Drug.2h-Drug.0h,
# Placebo.1vs0 = Placebo.1h-Placebo.0h,
# Placebo.2vs0 = Placebo.2h-Placebo.0h,
# DrugvsPlacebo.0h = Drug.0h-Placebo.0h,
# DrugvsPlacebo.1h = (Drug.1h-Drug.0h)-(Placebo.1h-Placebo.0h),
# DrugvsPlacebo.2h = (Drug.2h-Drug.0h)-(Placebo.2h-Placebo.0h),
levels=design)

qlf <- glmQLFTest(fit, contrast=my.contrasts[,"P.AvM"], coef=2)
qlf <- glmQLFTest(fit, coef=2)
topTags(qlf)
summary(decideTests(qlf))
```

```{r edgeR_Day14} 
# for edgeR
write.table(read.csv("../Data/filtered_counts/ day14.counts.filtered_matrix.csv", sep=","), "../Data/filtered_counts/ day14.counts.filtered_matrix.tab") # day0 count matrix
d14.counts_matrix_egdeR  <- read.table(file="../Data/filtered_counts/ day14.counts.filtered_matrix.tab")
d14.counts_matrix_egdeR<- data.frame(d14.counts_matrix_egdeR[,-1], row.names=d14.counts_matrix_egdeR[,1]) 

colnames(d14.counts_matrix_egdeR) # same order
d14.exp_data[2] # same order



# Prim.Treat <- factor(d14.exp_data$Primary_Treatment)
# # Prim.Treat <- relevel(Prim.Treat, ref="A")
# Sec.Treat <- factor(d14.exp_data$Second_Treament)

Group <- factor(d14.exp_data$All_Treatment)
# Group <- factor(d14.exp_data$Primary_Treatment)
design <- model.matrix(~Group)
colnames(design) <- levels(Group)
y <- DGEList(counts=d14.counts_matrix_egdeR, group=Group) 
y <- estimateDisp(y,design) # get dispersion value
# y$samples
fit <- glmQLFit(y, design)
colnames(fit)
# Then we can make any comparisons we wish. For example, we might wish to make the following
# contrasts:
my.contrasts <- makeContrasts(
 P.AvM = (AM+AA+AS)-(MM+MA+MS),
# Drug.2vs0 = Drug.2h-Drug.0h,
# Placebo.1vs0 = Placebo.1h-Placebo.0h,
# Placebo.2vs0 = Placebo.2h-Placebo.0h,
# DrugvsPlacebo.0h = Drug.0h-Placebo.0h,
# DrugvsPlacebo.1h = (Drug.1h-Drug.0h)-(Placebo.1h-Placebo.0h),
# DrugvsPlacebo.2h = (Drug.2h-Drug.0h)-(Placebo.2h-Placebo.0h),
levels=design)

qlf <- glmQLFTest(fit, contrast=my.contrasts[,"P.AvM"], coef=2)
#qlf <- glmQLFTest(fit, coef=1)
topTags(qlf)
summary(decideTests(qlf)) # total number of DEGs at 5% FDR 
```

# DESeq2 datasets

### DEsign DESeqDatasets
```{r build_dds_datasets}
# ========================================================== 
# DAY 0 FULL MODEL ==  design = ~ Primary_Treatment
# ========================================================== #
#####
d0.metadata <- d0.exp_data %>% dplyr::select(c('Sample.Name', 'Primary_Treatment', 'All_Treatment')) # coondense dataset to build target matrix
d0.metadata <- data.frame(d0.metadata[,-1], row.names=d0.metadata[,1]) # move Sample.Name column as row names  
d0.metadata.mtx <- as.matrix(d0.metadata, row.names="Geoduck.ID") # create matrix 
# check for 'TRUE' in each - check before proceeding  design
d0.metadata.mtx <- d0.metadata.mtx[match(colnames(d0.counts_matrix),rownames(d0.metadata.mtx)), ]
# all(rownames(d0.metadata.mtx) %in% colnames(d0.counts_matrix)) # should be TRUE
# all(rownames(d0.metadata.mtx) == colnames(d0.counts_matrix)) # should be TRUE
# all(rownames(d0.metadata.mtx) == colnames(d0.counts_matrix))  # should be TRUE
# build dds
dds.d0 <- DESeqDataSetFromMatrix(countData = d0.counts_matrix,
                                      colData = d0.metadata.mtx,
                                      design = ~ Primary_Treatment) # DESeq Data Set (dds) - design as ~Primary_Treatment
 
# ========================================================== 
# DAY 7 - FULL MODEL     ==  design = ~ Primary_Treatment+Second_Treament+Primary_Treatment:Second_Treament
#       - ADDITIVE MODEL ==  design = ~ Primary_Treatment+Second_Treament
#       - GROUP    MODEL ==  design = ~ Primary_Treatment+Second_Treament
# ========================================================== #
#####
d7.metadata <- d7.exp_data %>% dplyr::select(c('Sample.Name', 'Primary_Treatment', 'Second_Treament', 'All_Treatment')) # condense dataset to   build target matrix
d7.metadata <- data.frame(d7.metadata[,-1], row.names=d7.metadata[,1]) # move Sample.Name column as row names  
d7.metadata.mtx <- as.matrix(d7.metadata, row.names="Geoduck.ID") # create matrix 
# CKECK THE cts.matrix AND THE exp.data.mtx
d7.metadata.mtx <- d7.metadata.mtx[match(colnames(d7.counts_matrix),rownames(d7.metadata.mtx)), ]
# all(rownames(d7.metadata.mtx) %in% colnames(d7.counts_matrix)) # should be TRUE
# all(rownames(d7.metadata.mtx) == colnames(d7.counts_matrix)) # should be TRUE
# all(rownames(d7.metadata.mtx) == colnames(d7.counts_matrix))  # should be TRUE
# build dds
# FULL MODEL 
dds.d7 <- DESeqDataSetFromMatrix(countData = d7.counts_matrix,
                                  colData = d7.metadata.mtx,
                                  design = ~ Second_Treament+Primary_Treatment+Second_Treament:Primary_Treatment) # DESeq Data Set (dds) - design as ~Primary_Treatment

# ADDITIVE MODEL  
dds.d7.main <- DESeqDataSetFromMatrix(countData = d7.counts_matrix,
                                       colData = d7.metadata.mtx,
                                       design = ~ Second_Treament+Primary_Treatment)
# GROUP MODEL- note! ~group -1` removed the intercept to call each pariwise interaction
dds.d7.group <- DESeqDataSetFromMatrix(countData = d7.counts_matrix,
                                        colData = d7.metadata.mtx,
                                        design = ~ All_Treatment-1) 
# ========================================================== 
# DAY 14  - FULL MODEL     ==  design = ~ Primary_Treatment+Second_Treament+Primary_Treatment:Second_Treament
#       - ADDITIVE MODEL ==  design = ~ Primary_Treatment+Second_Treament
#       - GROUP    MODEL ==  design = ~ Primary_Treatment+Second_Treament
# ========================================================== #
#####
d14.metadata <- d14.exp_data %>% dplyr::select(c('Sample.Name', 'Primary_Treatment', 'Second_Treament', 'All_Treatment')) # condense dataset to build target matrix
d14.metadata <- data.frame(d14.metadata[,-1], row.names=d14.metadata[,1]) # move Sample.Name column as row names  
d14.metadata.mtx <- as.matrix(d14.metadata, row.names="Geoduck.ID") # create matrix 
# CKECK THE cts.matrix AND THE exp.data.mtx
d14.metadata.mtx <- d14.metadata.mtx[match(colnames(d14.counts_matrix),rownames(d14.metadata.mtx)), ]
# all(rownames(d14.metadata.mtx) %in% colnames(d14.counts_matrix)) # should be TRUE
# all(rownames(d14.metadata.mtx) == colnames(d14.counts_matrix)) # should be TRUE
# all(rownames(d14.metadata.mtx) == colnames(d14.counts_matrix))  # should be TRUE
# build dds
# FULL MODEL 
dds.d14 <- DESeqDataSetFromMatrix(countData = d14.counts_matrix,
                                     colData = d14.metadata.mtx,
                                     design = ~ Second_Treament+Primary_Treatment+Second_Treament:Primary_Treatment) # DESeq Data Set (dds) - design as ~Primary_Treatment
# GROUP MDOEL
dds.d14.group <- DESeqDataSetFromMatrix(countData = d14.counts_matrix,
                                        colData = d14.metadata.mtx,
                                        design = ~ All_Treatment-1) 
# ADDITIVE MODEL 
dds.d14.main <- DESeqDataSetFromMatrix(countData = d14.counts_matrix,
                                       colData = d14.metadata.mtx,
                                       design = ~ Second_Treament+Primary_Treatment) 
# ========================================================== 
# DAY 21 -  FULL MODEL == design = ~Primary_Treatment+Second_Treament+Third_Treatment+Primary_Treatment:Second_Treament+Primary_Treatment:Third_Treatment+Second_Treament:Third_Treatment
#---- INPUT: Experiment/design matrix == 'exp.data.d21'
#---- INPUT: TagSeq count matrix == 'cts.matrix.d21.filtered' (3 CPM in 50% samples)
# ========================================================== 
#####
d21.metadata <- d21.exp_data %>% dplyr::select(c('Sample.Name', 'Primary_Treatment','Second_Treament', 'Third_Treatment', 'All_Treatment')) # coondense dataset to build target matrix
d21.metadata <- data.frame(d21.metadata[,-1], row.names=d21.metadata[,1]) # move Sample.Name column as row names  
d21.metadata.mtx <- as.matrix(d21.metadata, row.names="Geoduck.ID") # create matrix 
# CKECK THE cts.matrix AND THE exp.data.mtx
d21.metadata.mtx <- d21.metadata.mtx[match(colnames(d21.counts_matrix),rownames(d21.metadata.mtx)), ]
# all(rownames(d21.metadata.mtx) %in% colnames(d21.counts_matrix)) # should be TRUE
# all(rownames(d21.metadata.mtx) == colnames(d21.counts_matrix)) # should be TRUE
# all(rownames(d21.metadata.mtx) == colnames(d21.counts_matrix))  # should be TRUE
# build dds
# FULL MODEL
dds.d21 <- DESeqDataSetFromMatrix(countData = d21.counts_matrix,
                                        colData = d21.metadata.mtx,
                                        design = ~ Primary_Treatment+
                                        Second_Treament+
                                        Third_Treatment+
                                        Primary_Treatment:Second_Treament+
                                        Primary_Treatment:Third_Treatment+
                                        Second_Treament:Third_Treatment) 
# GROUP MDOEL
dds.d21.group <- DESeqDataSetFromMatrix(countData = d21.counts_matrix,
                                        colData = d21.metadata.mtx,
                                        design = ~ All_Treatment-1) 
# ADDITIVE MODEL 
dds.d21.main <- DESeqDataSetFromMatrix(countData = d21.counts_matrix,
                                       colData = d21.metadata.mtx,
                                       design = ~ Third_Treatment+Second_Treament+Primary_Treatment) 

```

### Plot reads per gene and reads per sample in each dds
```{r ddsPlots_read.gene_reads.sample}
# DAY 0 PLOT THE (1) READS PER SAMPLE (2) READS PER GENE - FOR EACH DDS
d0.nsamples <- ncol(counts(dds.d0)) # Number of samples - for the plot label
d0.rps <- qplot(colSums(counts(dds.d0))) +# reads of reads per sample 
  labs(x = "Mapped reads per sample", y = "Number of samples",
       title = "Day 0: Mapped reads per sample") +
  geom_label(aes(x = 8e5, y = 1, label = paste(d0.nsamples, "samples")))
d0.ngenes <- nrow(counts(dds.d0)) # Number of genes
d0.ngenes_min <-min(rowSums(counts(dds.d0))) #  minimum reads
d0.ngenes_mean.min <- min(rowMeans2(counts(dds.d0))) # minimum row mean
d0.ngenes_max <-max(rowSums(counts(dds.d0))) #  maximum reads
d0.rpg <- qplot(log10(rowSums(counts(dds.d0))), bins = 16) + # number of reads per gene
  labs(x = "log10(Mapped reads per gene)", y = "Number of genes",
       title = "Day 0: Mapped reads per gene") +
  geom_label(aes(x = 4, y = 3000, label = paste(d0.ngenes, "total genes"))) +
  geom_label(aes(x = 4, y = 2700, label = paste("max count =", d0.ngenes_max))) +
  geom_label(aes(x = 4, y = 2400, label = paste("min count =", d0.ngenes_min))) +
  geom_label(aes(x = 4, y = 2100, label = paste("min mean =", d0.ngenes_mean.min)))
d0.countfig <- plot_grid(d0.rps, d0.rpg)

# DAY 7 PLOT THE (1) READS PER SAMPLE (2) READS PER GENE - FOR EACH DDS
d7.nsamples <- ncol(counts(dds.d7)) # Number of samples - for the plot label
d7.rps <- qplot(colSums(counts(dds.d7))) +# reads of reads per sample 
  labs(x = "Mapped reads per sample", y = "Number of samples",
       title = "Day 7: Mapped reads per sample") +
  geom_label(aes(x = 6e5, y = 4, label = paste(d7.nsamples, "samples")))
d7.ngenes <- nrow(counts(dds.d7)) # Number of genes
d7.ngenes_min <-min(rowSums(counts(dds.d7))) #  minimum reads
d7.ngenes_max <-max(rowSums(counts(dds.d7))) #  maximum reads
d7.ngenes_mean.min <- min(rowMeans2(counts(dds.d7))) # minimum row mean
d7.rpg <- qplot(log10(rowSums(counts(dds.d7))), bins = 16) + # number of reads per gene
  labs(x = "log10(Mapped reads per gene)", y = "Number of genes",
       title = "Day 7: Mapped reads per gene") +
  geom_label(aes(x = 4, y = 3000, label = paste(d7.ngenes, "total genes"))) +
  geom_label(aes(x = 4, y = 2700, label = paste("max count =", d7.ngenes_max))) +
  geom_label(aes(x = 4, y = 2400, label = paste("min count =", d7.ngenes_min)))  +
  geom_label(aes(x = 4, y = 2100, label = paste("min mean =", d7.ngenes_mean.min)))
d7.countfig <- plot_grid(d7.rps, d7.rpg)


# DAY 14 PLOT THE (1) READS PER SAMPLE (2) READS PER GENE - FOR EACH DDS
d14.nsamples <- ncol(counts(dds.d14)) # Number of samples - for the plot label
d14.rps <- qplot(colSums(counts(dds.d14))) +# reads of reads per sample 
  labs(x = "Mapped reads per sample", y = "Number of samples",
       title = "Day 14: Mapped reads per sample") +
  geom_label(aes(x = 6e5, y = 4, label = paste(d14.nsamples, "samples")))
d14.ngenes <- nrow(counts(dds.d14)) # Number of genes
d14.ngenes_min <-min(rowSums(counts(dds.d14))) #  minimum reads
d14.ngenes_max <-max(rowSums(counts(dds.d14))) #  maximum reads
d14.ngenes_mean.min <- min(rowMeans2(counts(dds.d14))) # minimum row mean
d14.rpg <- qplot(log10(rowSums(counts(dds.d14))), bins = 16) + # number of reads per gene
  labs(x = "log10(Mapped reads per gene)", y = "Number of genes",
       title = "Day 14: Mapped reads per gene") +
  geom_label(aes(x = 4, y = 3000, label = paste(d14.ngenes, "total genes"))) +
  geom_label(aes(x = 4, y = 2700, label = paste("max count =", d14.ngenes_max))) +
  geom_label(aes(x = 4, y = 2400, label = paste("min count =", d14.ngenes_min))) +
  geom_label(aes(x = 4, y = 2100, label = paste("min mean =", d14.ngenes_mean.min)))
d14.countfig <- plot_grid(d14.rps, d14.rpg)

# DAY 21 PLOT THE (1) READS PER SAMPLE (2) READS PER GENE - FOR EACH DDS
d21.nsamples <- ncol(counts(dds.d21)) # Number of samples - for the plot label
d21.rps <- qplot(colSums(counts(dds.d21))) +# reads of reads per sample 
  labs(x = "Mapped reads per sample", y = "Number of samples",
       title = "Day 21: Mapped reads per sample") +
  geom_label(aes(x = 6e5, y = 11, label = paste(d21.nsamples, "samples")))
d21.ngenes <- nrow(counts(dds.d21)) # Number of genes
d21.ngenes_min <-min(rowSums(counts(dds.d21))) #  minimum reads
d21.ngenes_max <-max(rowSums(counts(dds.d21))) #  maximum reads
d21.ngenes_mean.min <- min(rowMeans2(counts(dds.d21))) # minimum row mean
d21.countfig <- qplot(log10(rowSums(counts(dds.d21))), bins = 16) + # number of reads per gene
  labs(x = "log10(Mapped reads per gene)", y = "Number of genes",
       title = "Day 21: Mapped reads per gene") +
  geom_label(aes(x = 4, y = 3000, label = paste(d21.ngenes, "total genes"))) +
  geom_label(aes(x = 4, y = 2700, label = paste("max count =", d21.ngenes_max))) +
  geom_label(aes(x = 4, y = 2400, label = paste("min mean =", d21.ngenes_min))) +
  geom_label(aes(x = 4, y = 2100, label = paste("min mean =", d21.ngenes_mean.min)))
d21.countfig <- plot_grid(d21.rps, d21.countfig)

png("../Data/DESeq2/dds_GeneCounts_3cpm.png", 1000, 1000, pointsize=20)
plot_grid(d0.countfig, d7.countfig, d14.countfig, d21.countfig)
dev.off()
```

### RUN DESEQ2 model - wait for this to complete...
```{r run_dds}


dds.d0 <- DESeq(dds.d0) # wait for this to complete....

# RUN DESEQ2 model - view all the pariwise comparisons
dds.d7 <- DESeq(dds.d7) #  full model                 wait for this to complete....
dds.d7.group <- DESeq(dds.d7.group) # group model     wait for this to complete....
dds.d7.main <- DESeq(dds.d7.main) # main effect model    

# RUN DESEQ2 model - view all the pariwise comparisons
dds.d14 <- DESeq(dds.d14) #  full model               wait for this to complete....
dds.d14.group <- DESeq(dds.d14.group) # group model   wait for this to complete....
dds.d14.main <- DESeq(dds.d14.main) # main effect model    

# RUN DESEQ2 model - view all the pariwise comparisons
dds.d21 <- DESeq(dds.d21) #  full model                  wait for this to complete....
dds.d21.group <- DESeq(dds.d21.group) # group model      wait for this to complete....
dds.d21.main <- DESeq(dds.d21.main) # main effect model    
```

# Analysis (DEG tables)
Rationale for thresholds in the following cluster:
With the thousands of pairwise comparisons between genes, it is critical to acknowledge the prevalence of spurious discoveries that must be guarded against!
In the following, I will run each model with 'alpha = 0.05' to assume a False Discovery Rate of 5%. I reinforce this thinking by graphing all pvalues 
as a histogram to show the distrubution - abline(s) show the FDR cut-off and the true padj values at 0.05 and 0.1 thresholds. 

Strader et al. 2018 - 48 total samples, genes with mean count <10 across all samples werre ommitted
                    - rlog transformated (rlog fxn)
                    - adonic and vegan package to assess multivariate analysis of variance 
                    - Used a 10% false discovery rate threshold
                    - GO enrichments performend using the stat value output from DESeq2 using Mann Whitney U test (here: https://github.com/z0on/ GO_MWU)
Johnson and Kelly 2021 - 40 samples 
                       - used edgeR to assess pairwise changes in gene expression between treatments
                       - filtered genes with fewer than 3 CPM across 50%  *n=20_ of all samples
                       used R program vegan to assess PCoA with Euclidean distances from log2+1 transformed normalized counts from cpm function in egde R
                       - used Mann Whitney U test to identify enriched ontologies - than this also after identifying modules via WGCNA with physiological measurements




# Primary treatment: padj < 0.05; FDR 5%; LFC >1 (<-1)
#Day 0 - NOTE: this model is run is the oppositoe direction of the other (M v A as opposed to A v M)
```{r Day0_PRIMARY_TREATMENT}
# prep a table for knitr
DEGs.table <- data.frame(matrix(nrow = 4, ncol = 6)) # create a new data table
colnames(DEGs.table)<-c('Day', 'Treatment','Pairwise.compare', 'DEGs_total', 'DEGs_UP', 'DEGs_DOWN') 
    
# DAY 0 - main model
resd0.primary <- results(dds.d0, name="Primary_Treatment_M_vs_A", alpha = 0.05)
d0.numDEGs_padj <-data.frame(table(resd0.primary$padj<0.05))[2,2] # DEGs - NOT considering LFC - just p adj
resd0.primary.ordered <- resd0.primary[order(resd0.primary$padj), ] # Order by adjusted p-value
d0.num.DownReg <- sum((resd0.primary.ordered$log2FoldChange[1:d0.numDEGs_padj] >= 1) == TRUE) #  LFC >= 1
d0.num.UpReg <- sum((resd0.primary.ordered$log2FoldChange[1:d0.numDEGs_padj] <= -1) == TRUE) # LFC >= 1
d0.total <- sum(d0.num.DownReg,d0.num.UpReg) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1)
# histogram of FDR
png("../Output/DESeq2/Day0/Day0.FDR_hist.png", 1000, 1000, pointsize=20)
hist(resd0.primary$pvalue, breaks=20, col="grey") # view histogram
abline(h=c( (nrow(resd0.primary)*0.05), 
            ((table(resd0.primary$padj < 0.1)[2]) + (nrow(resd0.primary)*0.1)),
            ((table(resd0.primary$padj < 0.05)[2]) + (nrow(resd0.primary)*0.05)) ),
                  col=c("blue", "red", "red"), lty=c(1,2,1), lwd=c(1, 3, 1)) # add line at 
dev.off()
# Write results - covert to as.data.frame for the ordered results
resdata.d0.primary  <- merge(as.data.frame(resd0.primary.ordered), as.data.frame(counts(dds.d0, normalized=TRUE)), by="row.names", sort=FALSE) ## Merge with normalized count data
names(resdata.d0.primary)[1] <- "Gene"
resdata.d0.primary <- resdata.d0.primary[order(resdata.d0.primary$padj), ] # Order by adjusted p-value
path_out.d0 = 'C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day0/' # call path
write.csv(resdata.d0.primary, paste(path_out.d0,"Day0.Primary_DESeq2results.csv")) # write

# volcano plot 
png("../Output/DESeq2/Day0/Day0.PrimaryTreatment-VolcanoPlot.png", 1000, 1000, pointsize=20)
EnhancedVolcano(resd0.primary,
                lab = rownames(resd0.primary),
                x = 'log2FoldChange',
                y = 'padj',
                title = 'Day 0: Primary Treatment (M v A)',
                subtitle = "DESeq2 - Differential expression",
                FCcutoff = 1,
                pCutoff = 0.05,
                pointSize = 4.0,
                labSize = 6.0,
                colAlpha = 1,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 4.0,
                widthConnectors = 0.75)
dev.off()


# Plot dispersions 
png("../Output/DESeq2/Day0/Day0.dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds.d0, main="Day 0 dispersions")
dev.off()

# Data transformations for heatmap and PCA visuals ======================================================================================= #
# rlog - regularized log transformation of origin count data to log2 scale - fit for each sample and dist. of coefficients in the data
rlog.d0<- rlogTransformation(dds.d0) # rlog transform (regularized log)

# diagnostiscs
png("../Output/DESeq2/Day0/Day0.rlog_histogram.png", 1000, 1000, pointsize=20)# diagnostics of transformation # Histogram and sd plot
hist(assay(rlog.d0)) # view histogram 
dev.off()
png("../Output/DESeq2/Day0/Day0.rlog_mean_sd.png", 1000, 1000, pointsize=20)
meanSdPlot(assay(rlog.d0)) # shows the sd y axis (sq root of varaince in all samples) - flat curve may seem like a goals, BUT may be unreasonable in cases with MANY true DEGs from experimental conditions
dev.off()

# PCA plot rlog ------------------------ #
pcaData_d0 <- plotPCA(rlog.d0, intgroup = "Primary_Treatment", returnData = TRUE)
percentVar <- round(100 * attr(pcaData_d0, "percentVar"))
png("../Output/DESeq2/Day0/Day0.rlog_PCA.png", 1000, 1000, pointsize=20)
ggplot(pcaData_d0, aes(x = PC1, y = PC2, color = Primary_Treatment, label=name)) +
  scale_shape_manual(values = c(4, 19, 17)) +
  geom_text(aes(label=name),hjust=0.2, vjust=1.4, size=5) +
  geom_point(size =6) +
  theme_classic() +
  theme(text = element_text(size=15)) +
  ggtitle("PCA: Day0 (rlog)") +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
dev.off()


# Plot pheatmap map rlog------------------------ #
# save pheatmap fxn
save_pheatmap <- function(x, filename, width=1000, height=960) { # template for saving pheatmap outputs
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  png(filename,width = width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
# parameters for the pheatmap
topgenes.IDs <- head(rownames(resd0.primary.ordered),13) # call the first 13 genes that are differentially expressed, FDR < 0.05 correct pdj
topgenes_rlog.counts <- assay(rlog.d0)[topgenes.IDs,] 
topgenes_corrected <- topgenes_rlog.counts - rowMeans(topgenes_rlog.counts) # substract from the row mean to get +/- 0 to normalizze and ease the asthetic
df_annot.col <- as.data.frame(colData(dds.d0)[c("Primary_Treatment")])


# colorblind palette
# colorBlindBlack8  <- c("#000000", "#E69F00", "#56B4E9", "#009E73", 
#                        "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
# pie(rep(1, 8), col = colorBlindBlack8)

ann_colors = list(
    Primary_Treatment = c(A="turquoise", M="coral1"))

d0.rlog.heatmap <- pheatmap(topgenes_corrected, 
                            annotation_col=df_annot.col, 
                            main = "Day 0  Ambient versus Moderate (13 total DEGs)",
                            cutree_cols = 2,
                            #cutree_rows = 2,
                            annotation_legend = TRUE,
                            annotation_colors = ann_colors,
                            show_rownames = FALSE,
                            labels_col=df_annot.col$Primary_Treatment, 
                            angle_col = 0,
                            fontsize = 8,
                            legend = TRUE)
#Save ppheatmap
save_pheatmap(d0.rlog.heatmap, filename = "../Output/DESeq2/Day0/Day0.PRIMARY.rlog_heatmap.png")

```

#Day 7
```{r Day7_PRIMARY_TREATMENT}
# DAY 7 - group model 
res.d7_P.AvM_group <- results(dds.d7.group, contrast = list(c("All_TreatmentAA", "All_TreatmentAM", "All_TreatmentAS"),c("All_TreatmentMA", "All_TreatmentMM", "All_TreatmentMS")),  alpha= 0.05)
# table(res.d7_P.MvA_group$padj<0.05) # 111 DEGs
d7.numDEGs_padj <-data.frame(table(res.d7_P.AvM_group$padj<0.05))[2,2] # DEGs - NOT considering LFC - just p adj
res.d7_Prim_AvM_group.ordered <- res.d7_P.AvM_group[order(res.d7_P.AvM_group$padj), ] # Order by adjusted p-value - 110 DEGs
d7.num.UpReg <- sum((res.d7_Prim_AvM_group.ordered$log2FoldChange[1:d7.numDEGs_padj] >= 1) == TRUE) #  LFC >= 1
d7.num.DownReg <- sum((res.d7_Prim_AvM_group.ordered$log2FoldChange[1:d7.numDEGs_padj] <= -1) == TRUE) # LFC >= 1
d7.total <- sum(d7.num.DownReg,d7.num.UpReg) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1)

png("../Output/DESeq2/Day7/Day7.Primary.FDR_hist.png", 1000, 1000, pointsize=20)
hist(res.d7_P.AvM_group$pvalue, breaks=20, col="grey") # view histogram,  
abline(h=c( (nrow(res.d7_P.AvM_group)*0.05), 
            ((table(res.d7_P.AvM_group$padj < 0.1)[2]) + (nrow(res.d7_P.AvM_group)*0.05)),
            ((table(res.d7_P.AvM_group$padj < 0.05)[2]) + (nrow(res.d7_P.AvM_group)*0.05)) ),
                  col=c("blue", "red", "red"), lty=c(1,2), lwd=c(1, 3)) # add line at 
dev.off()
# Write results - covert to as.data.frame for the ordered results
resdata.d7.primary  <- merge(as.data.frame(res.d7_Prim_AvM_group.ordered), as.data.frame(counts(dds.d7.group, normalized=TRUE)), by="row.names", sort=FALSE) ## Merge with normalized count data
names(resdata.d7.primary)[1] <- "Gene"
resdata.d7.primary <- resdata.d7.primary[order(resdata.d7.primary$padj), ] # Order by adjusted p-value
path_out.d7 = 'C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day7/' # call path
write.csv(resdata.d7.primary, paste(path_out.d7,"Day7.Primary_DESeq2results.csv")) # write

# volcano plot 
png("../Output/DESeq2/Day7/Day7.PrimaryTreatment-VolcanoPlot.png", 1000, 1000, pointsize=20)
EnhancedVolcano(res.d7_P.AvM_group,
                lab = rownames(res.d7_P.AvM_group),
                x = 'log2FoldChange',
                y = 'padj',
                title = 'Day 7: Primary Treatment (A v M)',
                subtitle = "DESeq2 - Differential expression",
                FCcutoff = 1,
                pCutoff = 0.05,
                pointSize = 4.0,
                labSize = 6.0,
                colAlpha = 1,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 4.0,
                widthConnectors = 0.75)
dev.off()


# Plot dispersions 
png("../Output/DESeq2/Day7/Day7.dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds.d7, main="Day 7 dispersions")
dev.off()



# Data transformations for heatmap and PCA visuals ======================================================================================= #


# rlog - regularized log transformation of origin count data to log2 scale - fit for each sample and dist. of coefficients in the data
rlog.d7<- rlogTransformation(dds.d7) # rlog transform (regularized log)

png("../Output/DESeq2/Day7/Day7.rlog_histogram.png", 1000, 1000, pointsize=20)# diagnostics of transformation # Histogram and sd plot
hist(assay(rlog.d7)) # view histogram 
dev.off()
png("../Output/DESeq2/Day7/Day7.rlog_mean_sd.png", 1000, 1000, pointsize=20)
meanSdPlot(assay(rlog.d7)) # shows the sd y axis (sq root of varaince in all samples) - flat curve may seem like a goals, BUT may be unreasonable in cases with MANY true DEGs from experimental conditions
dev.off()

# PCA plot rlog ------------------------ #
pcaData_d7 <- plotPCA(rlog.d7, intgroup = c( "Primary_Treatment", "Second_Treament"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData_d7, "percentVar"))
png("../Output/DESeq2/Day7/Day7.rlog_PCA.png", 1000, 1000, pointsize=20)
ggplot(pcaData_d7, aes(x = PC1, y = PC2, color = Primary_Treatment, shape = Second_Treament)) +
  scale_shape_manual(values = c(5, 19, 15)) +
  geom_text(aes(label=name),hjust=0.2, vjust=1.4, size=3) +
  geom_point(size =6) +
  theme_classic() +
  theme(text = element_text(size=15)) +
  ggtitle("PCA: Day7 (rlog)") +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
dev.off()
# Plot heat map rlog------------------------ #
save_pheatmap <- function(x, filename, width=1000, height=960) { # template for saving pheatmap outputs
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  png(filename,width = width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
# parameters for the pheatmap
topgenes.IDs <- head(rownames(res.d7_Prim_AvM_group.ordered),40) # call the first 111 genes that are differentially expressed, FDR < 0.05 correct pdj
topgenes_rlog.counts <- assay(rlog.d7)[topgenes.IDs,] 
topgenes_corrected <- topgenes_rlog.counts - rowMeans(topgenes_rlog.counts) # substract from the row mean to get +/- 0 to normalizze and ease the asthetic
df_annot.col <- as.data.frame(colData(dds.d7)[c("Primary_Treatment","Second_Treament")])


# Run pheatmap
# colorblind palette
# colorBlindBlack8  <- c("#000000", "#E69F00", "#56B4E9", "#009E73", 
#                        "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
# pie(rep(1, 8), col = colorBlindBlack8)

ann_colors = list(
    Primary_Treatment = c(A="#56B4E9", M="#E69F00"),
    Second_Treament = c(A = "#56B4E9", M = "#E69F00", S ="#D55E00"))

d7.rlog.heatmap <- pheatmap(topgenes_corrected, 
                            annotation_col=df_annot.col, 
                            main = "Day 7 PRIMARY Ambient versus Moderate (111 total DEGs; 40 shown)",
                            #color = colorRampPalette(c("limegreen", "white", "gray39"))(50),
                            cutree_cols = 4,
                            cutree_rows = 2,
                            annotation_legend = TRUE,
                            annotation_colors = ann_colors,
                            show_rownames = FALSE,
                            labels_col=(paste(df_annot.col$Primary_Treatment,df_annot.col$Second_Treament, sep ="_")), 
                            angle_col = 45,
                            fontsize = 8,
                            legend = TRUE)
#Save ppheatmap
save_pheatmap(d7.rlog.heatmap, filename = "../Output/DESeq2/Day7/Day7.PRIMARY.rlog_heatmap.png")
```

#Day 14
```{r Day14_PRIMARY_TREATMENT}
# DAY 14 - group model 
res.d14_P.AvM_group <- results(dds.d14.group, contrast = list(c("All_TreatmentAA", "All_TreatmentAM", "All_TreatmentAS"),  c("All_TreatmentMA", "All_TreatmentMM", "All_TreatmentMS")),  alpha= 0.05)
# table(res.d14_P.AvM_group$padj<0.05) 
d14.numDEGs_padj <-data.frame(table(res.d14_P.AvM_group$padj<0.05))[2,2] # DEGs - NOT considering LFC - just p adj
res.d14_Prim_AvM_group.ordered <- res.d14_P.AvM_group[order(res.d14_P.AvM_group$padj), ] # Order by adjusted p-value
d14.num.UpReg <- sum((res.d14_Prim_AvM_group.ordered$log2FoldChange[1:d14.numDEGs_padj] >= 1) == TRUE) #  LFC >= 1
d14.num.DownReg <- sum((res.d14_Prim_AvM_group.ordered$log2FoldChange[1:d14.numDEGs_padj] <= -1) == TRUE) # LFC >= 1
d14.total <- sum(d14.num.DownReg,d14.num.UpReg) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1) - 507

hist(res.d14_P.AvM_group$pvalue, breaks=20, col="grey") # view histogram, 
abline(h=c( (nrow(res.d14_P.AvM_group)*0.1), 
            ((table(res.d14_P.AvM_group$padj < 0.1)[2]) + (nrow(res.d14_P.AvM_group)*0.1)),
            ((table(res.d14_P.AvM_group$padj < 0.05)[2]) + (nrow(res.d14_P.AvM_group)*0.1)) ),
                  col=c("blue", "red", "red"), lty=c(1,2), lwd=c(1, 3)) # add line at 

# Write results - covert to as.data.frame for the ordered results
resdata.d14.primary  <- merge(as.data.frame(res.d14_Prim_AvM_group.ordered), as.data.frame(counts(dds.d14.group, normalized=TRUE)), by="row.names", sort=FALSE) ## Merge with normalized count data
names(resdata.d14.primary)[1] <- "Gene"
resdata.d14.primary <- resdata.d14.primary[order(resdata.d14.primary$padj), ] # Order by adjusted p-value
path_out.d14 = 'C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day14/' # call path
write.csv(resdata.d14.primary, paste(path_out.d14,"Day14.Primary_DESeq2results.csv")) # write

# volcano plot 
png("../Output/DESeq2/Day14/Day14.Primary-VolcanoPlot.png", 1000, 1000, pointsize=20)
EnhancedVolcano(res.d14_P.AvM_group,
                lab = rownames(res.d14_P.AvM_group),
                x = 'log2FoldChange',
                y = 'padj',
                title = 'Day 14: Primary Treatment (M v A)',
                subtitle = "DESeq2 - Differential expression",
                FCcutoff = 1,
                pCutoff = 0.05,
                pointSize = 4.0,
                labSize = 6.0,
                colAlpha = 1,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 4.0,
                widthConnectors = 0.145)
dev.off()


# Plot dispersions 
png("../Output/DESeq2/Day14/Day14.dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds.d14, main="Day 14 dispersions")
dev.off()

# Data transformations for heatmap and PCA visuals ======================================================================================= #
# rlog - regularized log transformation of origin count data to log2 scale - fit for each sample and dist. of coefficients in the data
rlog.d14<- rlogTransformation(dds.d14) # rlog transform (regularized log)

png("../Output/DESeq2/Day14/Day14.rlog_histogram.png", 1000, 1000, pointsize=20)# diagnostics of transformation # Histogram and sd plot
hist(assay(rlog.d14)) # view histogram 
dev.off()
png("../Output/DESeq2/Day14/Day14.rlog_mean_sd.png", 1000, 1000, pointsize=20)
meanSdPlot(assay(rlog.d14)) # shows the sd y axis (sq root of varaince in all samples) - flat curve may seem like a goals, BUT may be unreasonable in cases with MANY true DEGs from experimental conditions
dev.off()

# PCA plot rlog ------------------------ #
pcaData_d14 <- plotPCA(rlog.d14, intgroup = "Primary_Treatment", returnData = TRUE)
percentVar <- round(100 * attr(pcaData_d14, "percentVar"))
png("../Output/DESeq2/Day14/Day14.rlog_PCA.png", 1000, 1000, pointsize=20)
ggplot(pcaData_d14, aes(x = PC1, y = PC2, color = Primary_Treatment, label=name)) +
  scale_shape_manual(values = c(4, 19, 114)) +
  geom_text(aes(label=name),hjust=0.2, vjust=1.4, size=5) +
  geom_point(size =6) +
  theme_classic() +
  theme(text = element_text(size=15)) +
  ggtitle("PCA: Day14 (rlog)") +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
dev.off()
# Plot heat map rlog------------------------ #
save_pheatmap <- function(x, filename, width=1000, height=960) { # template for saving pheatmap outputs
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  png(filename,width = width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
# parameters for the pheatmap
topgenes.IDs <- head(rownames(res.d14_Prim_AvM_group.ordered),40) # call the first 13 genes that are differentially expressed, FDR < 0.05 correct pdj
topgenes_rlog.counts <- assay(rlog.d14)[topgenes.IDs,] 
topgenes_corrected <- topgenes_rlog.counts - rowMeans(topgenes_rlog.counts) # substract from the row mean to get +/- 0 to normalizze and ease the asthetic
df_annot.col <- as.data.frame(colData(dds.d14)[c("Primary_Treatment","Second_Treament")])


# colorblind palette
# colorBlindBlack8  <- c("#000000", "#E69F00", "#56B4E9", "#009E73", 
#                        "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
# pie(rep(1, 8), col = colorBlindBlack8)

ann_colors = list(
    Primary_Treatment = c(A="#56B4E9", M="#E69F00"),
    Second_Treament = c(A = "#56B4E9", M = "#E69F00", S ="#D55E00"))

d14.rlog.heatmap <- pheatmap(topgenes_corrected, 
                            annotation_col=df_annot.col, 
                            main = "Day 14 PRIMARY Ambient versus Moderate (507 total DEGs; 40 shown)",
                            #color = colorRampPalette(c("limegreen", "white", "gray39"))(50),
                            cutree_cols = 2,
                            cutree_rows = 2,
                            annotation_legend = TRUE,
                            annotation_colors = ann_colors,
                            show_rownames = FALSE,
                            labels_col=(paste(df_annot.col$Primary_Treatment,df_annot.col$Second_Treament, sep ="_")), 
                            angle_col = 45,
                            fontsize = 8,
                            legend = TRUE)
#Save ppheatmap
save_pheatmap(d14.rlog.heatmap, filename = "../Output/DESeq2/Day14/Day14.PRIMARY.rlog_heatmap.png")
```

#Day 21
```{r Day21_PRIMARY_TREATMENT}
# DAY 21 - group model 
res.d21_Prim_AvM_group <- results(dds.d21.group, contrast = list(c("All_TreatmentAAA", "All_TreatmentAAM", "All_TreatmentAMA", "All_TreatmentAMM", "All_TreatmentASA","All_TreatmentASM"), c("All_TreatmentMAA", "All_TreatmentMAM", "All_TreatmentMMA", "All_TreatmentMMM", "All_TreatmentMSA", "All_TreatmentMSM")),  alpha= 0.1)
d21.numDEGs_padj <-data.frame(table(res.d21_Prim_AvM_group$padj<0.05))[2,2] # DEGs - NOT considering LFC - just p adj
res.d21_Prim_AvM_group.ordered <- res.d21_Prim_AvM_group[order(res.d21_Prim_AvM_group$padj), ] # Order by adjusted p-value
d21.num.UpReg <- sum((res.d21_Prim_AvM_group.ordered$log2FoldChange[1:d21.numDEGs_padj] >= 1) == TRUE) #  LFC >= 1
d21.num.DownReg <- sum((res.d21_Prim_AvM_group.ordered$log2FoldChange[1:d21.numDEGs_padj] <= -1) == TRUE) # LFC >= 1
d21.total <- sum(d21.num.DownReg,d21.num.UpReg) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1) - 230 genes



hist(res.d21_Prim_AvM_group$pvalue, breaks=20, col="grey") # view histogram,  
abline(h=c( (nrow(res.d21_Prim_AvM_group)*0.1), 
            ((table(res.d21_Prim_AvM_group$padj < 0.1)[2]) + (nrow(res.d21_Prim_AvM_group)*0.1)),
            ((table(res.d21_Prim_AvM_group$padj < 0.05)[2]) + (nrow(res.d21_Prim_AvM_group)*0.1)) ),
                  col=c("blue", "red", "red"), lty=c(1,2), lwd=c(1, 3)) # add line at 


# Write results - covert to as.data.frame for the ordered results
resdata.d21.primary  <- merge(as.data.frame(res.d21_Prim_AvM_group.ordered), as.data.frame(counts(dds.d21.group, normalized=TRUE)), by="row.names", sort=FALSE) ## Merge with normalized count data
names(resdata.d21.primary)[1] <- "Gene"
resdata.d21.primary <- resdata.d21.primary[order(resdata.d21.primary$padj), ] # Order by adjusted p-value
path_out.d21 = 'C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day21/' # call path
write.csv(resdata.d21.primary, paste(path_out.d21,"Day21.Primary_DESeq2results.csv")) # write



# volcano plot 
png("../Output/DESeq2/Day21/Day21.PrimaryTreatment-VolcanoPlot.png", 1000, 1000, pointsize=20)
EnhancedVolcano(res.d21_Prim_AvM_group,
                lab = rownames(res.d21_Prim_AvM_group),
                x = 'log2FoldChange',
                y = 'padj',
                title = 'Day 21: Primary Treatment (M v A)',
                subtitle = "DESeq2 - Differential expression",
                FCcutoff = 1,
                pCutoff = 0.05,
                pointSize = 4.0,
                labSize = 6.0,
                colAlpha = 1,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 4.0,
                widthConnectors = 0.145)
dev.off()



# Plot dispersions 
png("../Output/DESeq2/Day21/Day21.dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds.d21, main="Day 21 dispersions")
dev.off()



# Data transformations for heatmap and PCA visuals ======================================================================================= #
# rlog - regularized log transformation of origin count data to log2 scale - fit for each sample and dist. of coefficients in the data
rlog.d21<- rlogTransformation(dds.d21) # rlog transform (regularized log)

png("../Output/DESeq2/Day21/Day21.rlog_histogram.png", 1000, 1000, pointsize=20)# diagnostics of transformation # Histogram and sd plot
hist(assay(rlog.d21)) # view histogram 
dev.off()
png("../Output/DESeq2/Day21/Day21.rlog_mean_sd.png", 1000, 1000, pointsize=20)
meanSdPlot(assay(rlog.d21)) # shows the sd y axis (sq root of varaince in all samples) - flat curve may seem like a goals, BUT may be unreasonable in cases with MANY true DEGs from experimental conditions
dev.off()



# PCA plot rlog ------------------------ #
pcaData_d21 <- plotPCA(rlog.d21, intgroup = "Primary_Treatment", returnData = TRUE)
percentVar <- round(100 * attr(pcaData_d21, "percentVar"))
png("../Output/DESeq2/Day21/Day21.rlog_PCA.png", 1000, 1000, pointsize=20)
ggplot(pcaData_d7, aes(x = PC1, y = PC2, color = Primary_Treatment, shape = Second_Treament)) +
  scale_shape_manual(values = c(4, 19, 17)) +
  geom_text(aes(label=name),hjust=0.2, vjust=1.4, size=3) +
  geom_point(size =6) +
  theme_classic() +
  theme(text = element_text(size=15)) +
  ggtitle("PCA: Day7 (rlog)") +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed()
dev.off()




# Plot heat map rlog------------------------ #
save_pheatmap <- function(x, filename, width=1000, height=960) { # template for saving pheatmap outputs
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  png(filename,width = width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
# parameters for the pheatmap
topgenes.IDs <- head(rownames(res.d21_Prim_AvM_group.ordered),40) # call the first 13 genes that are differentially expressed, FDR < 0.05 correct pdj
topgenes_rlog.counts <- assay(rlog.d21)[topgenes.IDs,] 
topgenes_corrected <- topgenes_rlog.counts - rowMeans(topgenes_rlog.counts) # substract from the row mean to get +/- 0 to normalizze and ease the asthetic
df_annot.col <- as.data.frame(colData(dds.d21)[c("Primary_Treatment","Second_Treament","Third_Treatment")])

# colorblind palette
# colorBlindBlack8  <- c("#000000", "#E69F00", "#56B4E9", "#009E73", 
#                        "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
# pie(rep(1, 8), col = colorBlindBlack8)

ann_colors = list(
    Primary_Treatment = c(A="#56B4E9", M="#E69F00"),
    Second_Treament = c(A = "#56B4E9", M = "#E69F00", S ="#D55E00"),
    Third_Treatment = c(A = "#56B4E9", M = "#E69F00"))

d21.rlog.heatmap <- pheatmap(topgenes_corrected, 
                            annotation_col=df_annot.col, 
                            main = "Day 21 PRIMARY Ambient versus Moderate (230 total DEGs; 40 shown)",
                            #color = colorRampPalette(c("limegreen", "white", "gray39"))(50),
                            cutree_cols = 2,
                            cutree_rows = 2,
                            annotation_legend = TRUE,
                            annotation_colors = ann_colors,
                            show_rownames = FALSE,
                            labels_col=(paste(df_annot.col$Primary_Treatment,df_annot.col$Second_Treament, df_annot.col$Third_Treatment)), 
                            angle_col = 45,
                            fontsize = 6,
                            legend = TRUE)
#Save ppheatmap
save_pheatmap(d21.rlog.heatmap, filename = "../Output/DESeq2/Day21/Day21.PRIMARY.rlog_heatmap.png")
```

```{r Output_tables_and_knitr_summary_tables}
# write to table
DEGs.table$Day <- c("0", "7", "14", "21")
DEGs.table$Treatment <- "Primary"
DEGs.table$Pairwise.compare <- "A_v_M"
DEGs.table$DEGs_total <- c(d0.total, d7.total, d14.total, d21.total)
DEGs.table$DEGs_UP <- c(d0.num.UpReg, d7.num.UpReg, d14.num.UpReg, d21.num.UpReg)
DEGs.table$DEGs_DOWN <- c(d0.num.DownReg, d7.num.DownReg, d14.num.DownReg, d21.num.DownReg)


# FDR histograms - red line shows the false discovery rate of 0.05 called in each model as "alpha = 0.05"
# peaks at pdj < 0.05 above this line are considering true in the models
# hist(resd0.primary$pvalue, breaks=20, col="grey")          + abline(h=(nrow(resd0.primary)*0.05),col="red")
# hist(res.d7_P.AvM_group$pvalue, breaks=20, col="grey")     + abline(h=(nrow(res.d7_P.AvM_group)*0.05),col="red")
# hist(res.d14_P.AvM_group$pvalue, breaks=20, col="grey")    + abline(h=(nrow(res.d14_P.AvM_group)*0.05),col="red")
# hist(res.d21_Prim_AvM_group$pvalue, breaks=20, col="grey") + abline(h=(nrow(res.d21_Prim_AvM_group)*0.05),col="red")

# output tables for DEGs 
# Day 0 
DE_d0.primary <- merge(as.data.frame(resd0.primary), as.data.frame(counts(dds.d0, normalized=FALSE)), by="row.names", sort=FALSE) %>% dplyr::filter(padj < 0.05) %>% mutate(up = log2FoldChange <= 1) %>% mutate(down = log2FoldChange >= -1) # remember! this model is M vA whereas the rest ate A v M
write.csv(DE_d0.primary, "../Output/DESeq2/Day0/DE_Day0_Primary.csv")

# Day 7 
DE_d7.primary <- merge(as.data.frame(res.d7_P.AvM_group), as.data.frame(counts(dds.d7, normalized=FALSE)), by="row.names", sort=FALSE) %>% dplyr::filter(padj < 0.05) %>% mutate(up = log2FoldChange > 1) %>% mutate(down = log2FoldChange < -1)
write.csv(DE_d7.primary, "../Output/DESeq2/Day7/DE_Day7_Primary.csv")

# Day 14 
DE_d14.primary <- merge(as.data.frame(res.d14_P.AvM_group), as.data.frame(counts(dds.d14, normalized=FALSE)), by="row.names", sort=FALSE) %>% dplyr::filter(padj < 0.05) %>% mutate(up = log2FoldChange > 1) %>% mutate(down = log2FoldChange < -1)
write.csv(DE_d14.primary, "../Output/DESeq2/Day14/DE_Day14_Primary.csv")

# Day 21
DE_d21.primary <- merge(as.data.frame(res.d21_Prim_AvM_group), as.data.frame(counts(dds.d21, normalized=FALSE)), by="row.names", sort=FALSE) %>% dplyr::filter(padj < 0.05) %>% mutate(up = log2FoldChange > 1) %>% mutate(down = log2FoldChange < -1)
write.csv(DE_d21.primary, "../Output/DESeq2/Day21/DE_Day21_Primary.csv")


d0_P.DEGs_boolean <- DE_d0.primary %>% dplyr::select(c("Row.names"
,"log2FoldChange","padj","up","down")) %>% dplyr::mutate(Day_Trmt = "Day0_Primary_AvM")

d7_P.DEGs_boolean <- DE_d7.primary %>% dplyr::select(c("Row.names"
,"log2FoldChange","padj","up","down")) %>% dplyr::mutate(Day_Trmt = "Day7_Primary_AvM")

d14_P.DEGs_boolean <- DE_d14.primary %>% dplyr::select(c("Row.names"
,"log2FoldChange","padj","up","down")) %>% dplyr::mutate(Day_Trmt = "Day14_Primary_AvM")

d21_P.DEGs_boolean <- DE_d21.primary %>% dplyr::select(c("Row.names"
,"log2FoldChange","padj","up","down")) %>% dplyr::mutate(Day_Trmt = "Day21_Primary_AvM")

All_DEGs_Primary.Ambieant_v_Moderate <- rbind(d0_P.DEGs_boolean,d7_P.DEGs_boolean, d14_P.DEGs_boolean,d21_P.DEGs_boolean)
write.csv(All_DEGs_Primary.Ambieant_v_Moderate, "../Output/DESeq2/DE_PrimaryTreatment.All.csv")

# Summary table for the frequency that DEGs occured 
# unique(All_DEGs_Primary.Ambieant_v_Moderate$Day_Trmt) # unique DAy_treat to filter
DE_freq.upreg <-All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(up == TRUE) %>% 
            group_by(Row.names, up) %>%
            summarise(Freq = n()) %>% 
            arrange(desc(Freq))
DE_freq.upreg <- DE_freq.upreg %>% group_by(Freq) %>% summarise(num.transcripts = n(),Direction = "upregulated")

DE_freq.downreg <-All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(down == TRUE) %>% 
            group_by(Row.names, down) %>%
            summarise(Freq = n()) %>%
            arrange(desc(Freq))
DE_freq.downreg <- DE_freq.downreg %>% group_by(Freq) %>% summarise(num.transcripts = n(),Direction = "downregulated")


knitr::kable(DEGs.table, caption = "Primary Treatment DEGs [3 cpm in 50%; FDR 0.05; padj <0.05; LFC >1]")
knitr::kable(rbind(DE_freq.downreg, DE_freq.upreg),caption = "Primary Treatment DEGs, occurances on days 7, 14, 21: (1 = unique; 2 = d days; 3 = all days)")

write.csv((rbind(DE_freq.downreg, DE_freq.upreg)), "../Output/DESeq2/Summary_Primary_DEGs_occurances_3cpm.csv")
write.csv(DEGs.table, "../Output/DESeq2/Summary_Primary_DEGs_3cpm.csv")
```





# Second treatment effects
#Day 7
```{r Day7_SECOND_TREATMENT}
# DAY 7 - group model 
### Seconc Treatment Ambient v Moderate 
res.Second.AvM_group <- results(dds.d7.group, contrast = list(c("All_TreatmentAA", "All_TreatmentMA"),c("All_TreatmentAM", "All_TreatmentMM")),  alpha= 0.05) # 114 DEGs
# table(res.d7_P.MvA_group$padj<0.05) # 111 DEGs
d7.numDEGs_padj <-data.frame(table(res.Second.AvM_group$padj<0.05))[2,2] # DEGs - NOT considering LFC - just p adj - 114
res.Second.AvM_group.ordered <- res.Second.AvM_group[order(res.Second.AvM_group$padj), ] # Order by adjusted p-value - 110 DEGs
d7.num.UpReg <- sum((res.Second.AvM_group.ordered$log2FoldChange[1:d7.numDEGs_padj] >= 1) == TRUE) #  LFC >= 1
d7.num.DownReg <- sum((res.Second.AvM_group.ordered$log2FoldChange[1:d7.numDEGs_padj] <= -1) == TRUE) # LFC >= 1
d7.total <- sum(d7.num.DownReg,d7.num.UpReg) # sum of DEGs with the criteria pdj < 0.05 + LFC>1 (< -1) 112

hist(res.Second.AvM_group$pvalue, breaks=20, col="grey") # view histogram,  
abline(h=c( (nrow(res.Second.AvM_group)*0.05), 
            ((table(res.Second.AvM_group$padj < 0.1)[2]) + (nrow(res.Second.AvM_group)*0.05)),
            ((table(res.Second.AvM_group$padj < 0.05)[2]) + (nrow(res.Second.AvM_group)*0.05)) ),
                  col=c("blue", "red", "red"), lty=c(1,2), lwd=c(1, 3)) # add line at 

# Write results - covert to as.data.frame for the ordered results
# resdata.d7.primary  <- merge(as.data.frame(res.d7_Prim_AvM_group.ordered), as.data.frame(counts(dds.d7.group, normalized=TRUE)), by="row.names", sort=FALSE) ## Merge with normalized count data
# names(resdata.d7.primary)[1] <- "Gene"
# resdata.d7.primary <- resdata.d7.primary[order(resdata.d7.primary$padj), ] # Order by adjusted p-value
# path_out.d7 = 'C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day7/' # call path
# write.csv(resdata.d7.primary, paste(path_out.d7,"Day7.Primary_DESeq2results.csv")) # write

# volcano plot 
png("../Output/DESeq2/Day7/Day7.Second_AvM-VolcanoPlot.png", 1000, 1000, pointsize=20)
EnhancedVolcano(res.Second.AvM_group,
                lab = rownames(res.Second.AvM_group),
                x = 'log2FoldChange',
                y = 'padj',
                title = 'Day 7: Second Treatment (A v M)',
                subtitle = "DESeq2 - Differential expression",
                FCcutoff = 1,
                pCutoff = 0.05,
                pointSize = 4.0,
                labSize = 6.0,
                colAlpha = 1,
                legendPosition = 'right',
                legendLabSize = 12,
                legendIconSize = 4.0,
                widthConnectors = 0.75)
dev.off()


# Plot dispersions 
png("../Output/DESeq2/Day7/Day7.dispersions.png", 1000, 1000, pointsize=20)
plotDispEsts(dds.d7, main="Day 7 dispersions")
dev.off()



# Data transformations for heatmap and PCA visuals ======================================================================================= #


# rlog - regularized log transformation of origin count data to log2 scale - fit for each sample and dist. of coefficients in the data
rlog.d7<- rlogTransformation(dds.d7) # rlog transform (regularized log)
# 
# png("../Output/DESeq2/Day7/Day7.rlog_histogram.png", 1000, 1000, pointsize=20)# diagnostics of transformation # Histogram and sd plot
# hist(assay(rlog.d7)) # view histogram 
# dev.off()
# png("../Output/DESeq2/Day0/Day0.rlog_mean_sd.png", 1000, 1000, pointsize=20)
# meanSdPlot(assay(rlog.d7)) # shows the sd y axis (sq root of varaince in all samples) - flat curve may seem like a goals, BUT may be unreasonable in cases with MANY true DEGs from experimental conditions
# dev.off()

# PCA plot rlog ------------------------ #
# pcaData_d7 <- plotPCA(rlog.d7, intgroup = c("Primary_Treatment","Second_Treament"), returnData = TRUE)
# percentVar <- round(100 * attr(pcaData_d7, "percentVar"))
# png("../Output/DESeq2/Day0/Day0.rlog_PCA.png", 1000, 1000, pointsize=20)
# ggplot(pcaData_d7, aes(x = PC1, y = PC2, color = Primary_Treatment, label=name)) +
#   scale_shape_manual(values = c(4, 19, 17)) +
#   geom_text(aes(label=name),hjust=0.2, vjust=1.4, size=5) +
#   geom_point(size =6) +
#   theme_classic() +
#   theme(text = element_text(size=15)) +
#   ggtitle("PCA: Day7 (rlog)") +
#   xlab(paste0("PC1: ", percentVar[1], "% variance")) +
#   ylab(paste0("PC2: ", percentVar[2], "% variance")) +
#   coord_fixed()
# dev.off()
# Plot heat map rlog------------------------ #
save_pheatmap <- function(x, filename, width=1000, height=960) { # template for saving pheatmap outputs
  stopifnot(!missing(x))
  stopifnot(!missing(filename))
  png(filename,width = width, height=height)
  grid::grid.newpage()
  grid::grid.draw(x$gtable)
  dev.off()
}
# parameters for the pheatmap
topgenes.IDs <- head(rownames(res.Second.AvM_group.ordered),20) # call the first 111 genes that are differentially expressed, FDR < 0.05 correct pdj
topgenes_rlog.counts <- assay(rlog.d7)[topgenes.IDs,] 
topgenes_corrected <- topgenes_rlog.counts - rowMeans(topgenes_rlog.counts) # substract from the row mean to get +/- 0 to normalizze and ease the asthetic
df_annot.col <- as.data.frame(colData(dds.d7)[c("Primary_Treatment","Second_Treament")])
# RRun pheatmap
d7.rlog.heatmap.S.AvM <- pheatmap(topgenes_corrected, 
                            annotation_col=df_annot.col, 
                            cutree_cols = 2, 
                            #cutree_rows = 2,
                            fontsize = 10,
                            show_rownames = F)
#Save ppheatmap
save_pheatmap(d7.rlog.heatmap.S.AvM, filename = "../Output/DESeq2/Day7/Day7.Second.AvMrlog_heatmap.png")









# secondary treatment effect for those primed under primary AMBIENT:------------------------------------------

########## # d7_SecondM.PrimaryM
d7_SecondM.PrimaryM <- results(dds.d7.main, contrast=list(c("Second_Treament_M_vs_A","Primary_Treatment_M_vs_A")),  alpha= 0.05) # Second Moderate treat effect for Moderate-primed animals
d7.numDEGs_padj<- data.frame(table(d7_SecondM.PrimaryM$padj<0.05))[2,2] # 32
d7_SecondM.PrimaryM.ordered <- d7_SecondM.PrimaryM[order(d7_SecondM.PrimaryM$padj), ] # Order by adjusted p-value
d7.SecondM.PrimaryM.UpReg <- sum((d7_SecondM.PrimaryM.ordered$log2FoldChange[1:d7.numDEGs_padj] >= 1) == TRUE) #  LFC >= 1
d7.SecondM.PrimaryM.DownReg <- sum((d7_SecondM.PrimaryM.ordered$log2FoldChange[1:d7.numDEGs_padj] <= -1) == TRUE) # LFC >= 1
d7.total <- sum(d7.SecondM.PrimaryM.UpReg,d7.SecondM.PrimaryM.DownReg) # 31
# Write results - covert to as.data.frame for the ordered results
DEGs.d7.SecM.PrimM <- merge(as.data.frame(d7_SecondM.PrimaryM.ordered), as.data.frame(counts(dds.d7.main, normalized=TRUE)), by="row.names", sort=FALSE) ## Merge with normalized count data
names(DEGs.d7.SecM.PrimM)[1] <- "Gene"
DEGs.d7.SecM.PrimM <- DEGs.d7.SecM.PrimM[order(DEGs.d7.SecM.PrimM$padj), ] # Order by adjusted p-value
path_out.d7 = 'C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day7/' # call path
write.csv(DEGs.d7.SecM.PrimM, paste(path_out.d7,"Day7.SecondM_PrimaryM_DESeq2results.csv")) # write

rlog.d7<- rlogTransformation(dds.d7) # rlog transform (regularized log)
topgenes <- head(rownames(d7_SecondM.PrimaryM.ordered),32) # call the first 13 genes that are differentially expressed
mat <- assay(rlog.d7)[topgenes,] 
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(dds.d7)[,c("Primary_Treatment", "Second_Treament")])
d7.rlog.heatmap <- pheatmap(mat, annotation_col=df,
                            cutree_cols = 2, 
                            cutree_rows = 2,
                            fontsize = 10,
                            show_rownames = F)

save_pheatmap(d21.rlog.heatmap, filename = "../Output/DESeq2/Day21/Day21.rlog_heatmap.png")


########## # d7_SecondS.PrimaryM
d7_SecondS.PrimaryM <- results(dds.d7.main, contrast=list(c("Second_Treament_S_vs_A","Primary_Treatment_M_vs_A")),  alpha= 0.05) # Second Severe treat effect for Moderate-primed animals
data.frame(table(d7_SecondS.PrimaryM$padj<0.05))[2,2] # 24
d7_SecondS.PrimaryM.ordered <- d7_SecondS.PrimaryM[order(d7_SecondS.PrimaryM$padj), ] # Order by adjusted p-value

topgenes <- head(rownames(d7_SecondS.PrimaryM.ordered),20) # call the first 13 genes that are differentially expressed
mat <- assay(rlog.d7)[topgenes,] 
mat <- mat - rowMeans(mat)
df <- as.data.frame(colData(dds.d7)[,c("Primary_Treatment", "Second_Treament")])
d7.rlog.heatmap <- pheatmap(mat, annotation_col=df,
                            cutree_cols = 2, 
                            cutree_rows = 2,
                            fontsize = 10,
                            show_rownames = F)


# secondary treatment effect for those primed under primary MODERATE:-----------------------------------------
########## # d7_SecondM.PrimaryA
d7_SecondM.PrimaryA<-results(dds.d7.main, contrast=c("Second_Treament","A","M"),  alpha= 0.05) # Second Moderate treat effect for Ambient-primed animals
data.frame(table(d7_SecondM.PrimaryA$padj<0.05))[2,2] # 118
d7_SecondM.PrimaryA.ordered <- d7_SecondM.PrimaryA[order(d7_SecondM.PrimaryA$padj), ] # Order by adjusted p-value

########## # d7_SecondS.PrimaryA
d7_SecondS.PrimaryA<-results(dds.d7.main, contrast=c("Second_Treament","A","S"),  alpha= 0.05) # Second Severe treat effect for Ambient-primed animals
data.frame(table(d7_SecondS.PrimaryA$padj<0.05))[2,2] # 1
d7_SecondS.PrimaryA.ordered <- d7_SecondS.PrimaryA[order(d7_SecondS.PrimaryA$padj), ] # Order by adjusted p-value

```











# ~group effecs; all pairwise interaction terms: padj < 0.05; FDR 5%; LFC >1 (<-1) 
```{r GROUP_Day 7, 14, and 21}
# Day 7: ALL PAIRWISE COMPARISONS
d7.GROUP.DEGs_table = data.frame() # start a data frame
df_d7_group<- data.frame(resultsNames(dds.d7.group))
unique.vars <- df_d7_group[1,]
for(i in 1:nrow(df_d7_group)) {
  
      var <- df_d7_group[i,]
      T_F <- (df_d7_group[,1]!=c(unique.vars,var))
      df_d7_group_2 <- data.frame(df_d7_group[T_F,])

      for(j in 1:nrow(df_d7_group_2)) {
          
          var2 <- df_d7_group_2[j,]
          res <- results(dds.d7.group, contrast = list((var),(var2)),  alpha= 0.05) 
          res.table<-as.data.frame(table(res$padj<0.05))
          DEGs_total <- res.table[2,2]
          DEGs_total[is.na(DEGs_total)] <- 0
          DEGs.table <- data.frame(matrix(nrow = 1, ncol = 6)) # create a new data table
          colnames(DEGs.table)<-c('Day', 'Treatment', 'Pairwise.compare', 'DEGs_total', 'DEGs_UP', 'DEGs_DOWN') 
          res.order <- res[order(res$padj), ] ## Order by adjusted p-value
          DEGs.table$Day <- "7"
          DEGs.table$Treatment <- "full_interaction"
          DEGs.table$Pairwise.compare <- paste((substr(var, 14,15)), (substr(var2, 14,15)), sep = "_")
          DEGs.table$DEGs_total <- DEGs_total
          DEGs.table$DEGs_UP <- sum((res.order$log2FoldChange[1:(DEGs_total)] >= 1) == TRUE) 
          DEGs.table$DEGs_DOWN <- sum((res.order$log2FoldChange[1:(DEGs_total)] <= -1) == TRUE) 
          df.group <- data.frame(DEGs.table) # name dataframe for this single row
          d7.GROUP.DEGs_table <- rbind(d7.GROUP.DEGs_table, df.group) # bind to a cumulative list dataframe
          unique.vars <- unique((substr(d7.GROUP.DEGs_table$Pairwise.compare,1,2)))
      } 
}
d7.full.interaction.DEGs <- d7.GROUP.DEGs_table[-grep("MM_A|MA_A|MS_A|MM_M|MS_MA", d7.GROUP.DEGs_table$Pairwise.compare),] # ommit redundant pairwise results - same pairwise in opposite direaction

# Day 14: ALL PAIRWISE COMPARISONS
d14.GROUP.DEGs_table = data.frame() # start a data frame
df_d14_group<- data.frame(resultsNames(dds.d14.group))
unique.vars <- df_d14_group[1,]
for(i in 1:nrow(df_d14_group)) {
  
      var <- df_d14_group[i,]
      T_F <- (df_d14_group[,1]!=c(unique.vars,var))
      df_d14_group_2 <- data.frame(df_d14_group[T_F,])

      for(j in 1:nrow(df_d14_group_2)) {
          
          var2 <- df_d14_group_2[j,]
          res <- results(dds.d14.group, contrast = list((var),(var2)),  alpha= 0.05) 
          res.table<-as.data.frame(table(res$padj<0.05))
          DEGs_total <- res.table[2,2]
          DEGs_total[is.na(DEGs_total)] <- 0
          DEGs.table <- data.frame(matrix(nrow = 1, ncol = 6)) # create a new data table
          colnames(DEGs.table)<-c('Day', 'Treatment', 'Pairwise.compare', 'DEGs_total', 'DEGs_UP', 'DEGs_DOWN') 
          res.order <- res[order(res$padj), ] ## Order by adjusted p-value
          DEGs.table$Day <- "14"
          DEGs.table$Treatment <- "full_interaction"
          DEGs.table$Pairwise.compare <- paste((substr(var, 14,15)), (substr(var2, 14,15)), sep = "_")
          DEGs.table$DEGs_total <- DEGs_total
          DEGs.table$DEGs_UP <- sum((res.order$log2FoldChange[1:(DEGs_total)] >= 1) == TRUE) 
          DEGs.table$DEGs_DOWN <- sum((res.order$log2FoldChange[1:(DEGs_total)] <= -1) == TRUE) 
          df.group <- data.frame(DEGs.table) # name dataframe for this single row
          d14.GROUP.DEGs_table <- rbind(d14.GROUP.DEGs_table, df.group) # bind to a cumulative list dataframe
          unique.vars <- unique((substr(d14.GROUP.DEGs_table$Pairwise.compare,1,2)))
      } 
}
d14.full.interaction.DEGs <- d14.GROUP.DEGs_table[-grep("MM_A|MA_A|MS_A|MM_M|MS_MA", d14.GROUP.DEGs_table$Pairwise.compare),] # ommit redundant pairwise results - same pairwise in opposite direaction


# Day 21: ALL PAIRWISE COMPARISONS
d21.GROUP.DEGs_table = data.frame() # start a data frame
df_d21_group<- data.frame(resultsNames(dds.d21.group))
unique.vars <- df_d21_group[1,]
for(i in 1:nrow(df_d21_group)) {
  
      var <- df_d21_group[i,]
      T_F <- (df_d21_group[,1]!=c(unique.vars,var))
      df_d21_group_2 <- data.frame(df_d21_group[T_F,])

      for(j in 1:nrow(df_d21_group_2)) {
          
          var2 <- df_d21_group_2[j,]
          res <- results(dds.d21.group, contrast = list((var),(var2)),  alpha= 0.05) 
          res.table<-as.data.frame(table(res$padj<0.05))
          DEGs_total <- res.table[2,2]
          DEGs_total[is.na(DEGs_total)] <- 0
          DEGs.table <- data.frame(matrix(nrow = 1, ncol = 6)) # create a new data table
          colnames(DEGs.table)<-c('Day', 'Treatment', 'Pairwise.compare', 'DEGs_total', 'DEGs_UP', 'DEGs_DOWN') 
          res.order <- res[order(res$padj), ] ## Order by adjusted p-value
          DEGs.table$Day <- "21"
          DEGs.table$Treatment <- "full_interaction"
          DEGs.table$Pairwise.compare <- paste((substr(var, 14,16)), (substr(var2, 14,16)), sep = "_")
          DEGs.table$DEGs_total <- DEGs_total
          DEGs.table$DEGs_UP <- sum((res.order$log2FoldChange[1:(DEGs_total)] >= 1) == TRUE) 
          DEGs.table$DEGs_DOWN <- sum((res.order$log2FoldChange[1:(DEGs_total)] <= -1) == TRUE) 
          df.group <- data.frame(DEGs.table) # name dataframe for this single row
          d21.GROUP.DEGs_table <- rbind(d21.GROUP.DEGs_table, df.group) # bind to a cumulative list dataframe
          unique.vars <- unique((substr(d21.GROUP.DEGs_table$Pairwise.compare,1,3)))
      } 
}

d21.full.interaction.DEGs <- d21.GROUP.DEGs_table[-grep("MAA_A|MAM_A|MMA_A|MMM_A|MSA_A|MSM_A|MMA_M|MMM_M|MSA_M", d21.GROUP.DEGs_table$Pairwise.compare),] # ommit redundant pairwise results - same pairwise in opposite direaction

# knitr table
knitr::kable(rbind(d7.full.interaction.DEGs, d14.full.interaction.DEGs, d21.full.interaction.DEGs), caption = "All pairwise DEGs in 'group' DESeq2 model")
```

# Plotting

### rlog transform dds data... this takes a bit!
```{r tranformations}
colData(dds.d0) <- droplevels(colData(dds.d0))
# Day 0 
rlog.d0 <- rlogTransformation(dds.d0)
vsd.d0 <- vst(dds.d0, blind = TRUE)

# Day 7 
rlog.d7 <- rlogTransformation(dds.d7)
vsd.d7 <- vst(dds.d7, blind = TRUE)

# Day 14 
rlog.d14 <- rlogTransformation(dds.d14)
vsd.d14 <- vst(dds.d14, blind = TRUE)

# Day 21 
rlog.d21 <- rlogTransformation(dds.d21)
vsd.d21 <- vst(dds.d21, blind = TRUE)

```

### 
```{r GO_analysis}
# call the GO terms
Geoduck_GOterms <- as.data.frame(Geoduck_annotation) %>% dplyr::select(c('V1','V8'))
colnames(Geoduck_GOterms)[1:2] <- c('transcript.ID', 'GO.terms') # call gene name and the GO terms - (Uniprot ID 'V5')
Geoduck_GOterms$GO.terms <- gsub(";", " ", Geoduck_GOterms$GO.terms) # remove the ; delimiter and replace with nothing - already tabbed
splitted <- strsplit(as.character(Geoduck_GOterms$GO.terms), ";") #slit into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(Geoduck_GOterms$transcript.ID, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row

d0.counts_matrix # all count data filtered for d0
DE_d0.primary # diff expressed genes from d0 in response to Primary treatment

# Vectors for goseq 
# 1. length vector  
GO_gene.length <- Geoduck_annotation %>% dplyr::mutate(length = V4-V3) %>%  dplyr::select(c("V1","length"))
names(GO_gene.length)[1] <- "gene.ID"



# 2.  Construct a named vector of DEGs suritable for goseq
# measured Genes 
all.counts <- tibble::rownames_to_column(all.counts_matrix, "gene.ID")
GO_unique.genes.all <- as.vector(unique(all.counts$gene.ID)) # call unique genes for GO analysis (goseq)

# GO_unique.genes.d0 <- as.vector(unique(row.names(d0.counts_matrix)))   # call unique genes for GO analysis (goseq)
# GO_unique.genes.d7 <- as.vector(unique(row.names(d7.counts_matrix)))   # call unique genes for GO analysis (goseq)
# GO_unique.genes.d14 <- as.vector(unique(row.names(d14.counts_matrix))) # call unique genes for GO analysis (goseq)
# GO_unique.genes.d21 <- as.vector(unique(row.names(d21.counts_matrix))) # call unique genes for GO analysis (goseq)

#call all DEGs in the target treatmet/day (i.e. Prim Mod vs. ambient "DE_d0.primary")
# and align to all unique mapped reads 'GO_unique.genes.all' to create a binary vvector 0 = not DE; 1 = DE
dim(DE_d0.primary) # day0 - should be 13
d0.primary_gene.vector=as.integer(GO_unique.genes.all%in%(DE_d0.primary$Row.names)) # Day 0 - Primary Treatment effect (Ambient vs. Moderate)
names(d0.primary_gene.vector)=GO_unique.genes.all

dim(DE_d7.primary) # day7 - should be 111
d7.primary_gene.vector=as.integer(GO_unique.genes.all%in%(DE_d7.primary$Row.names)) # Day 7 - Primary Treatment effect (Ambient vs. Moderate)
names(d7.primary_gene.vector)=GO_unique.genes.all 

dim(DE_d14.primary) # day14 - should be 511
d14.primary_gene.vector=as.integer(GO_unique.genes.all%in%(DE_d14.primary$Row.names)) # Day 14 - Primary Treatment effect (Ambient vs. Moderate)
names(d14.primary_gene.vector)=GO_unique.genes.all

dim(DE_d21.primary) # day21 - should be 233
d21.primary_gene.vector=as.integer(GO_unique.genes.all%in%(DE_d21.primary$Row.names)) # Day 21 - Primary Treatment effect (Ambient vs. Moderate)
names(d21.primary_gene.vector)=GO_unique.genes.all

# Review what we have for goseq....
# Binary DEG vectors 
head(d0.primary_gene.vector)
head(d7.primary_gene.vector)
head(d14.primary_gene.vector)
head(d21.primary_gene.vector)

# ID vector
head(GO_unique.genes.all)

#Make length vector
GO_gene.length_merge <- merge(GO_gene.length, all.counts, by = "gene.ID")
length_vector <- GO_gene.length_merge$length
head(length_vector)

#Calculate Probability Weighting Function
DEG.pwf<-nullp(d21.primary_gene.vector, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
head(DEG.pwf)
GO.wall<-goseq(DEG.pwf, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
#How many enriched GO terms do we have
class(GO.wall)
head(GO.wall)
tail(GO.wall)
nrow(GO.wall)
#Find only enriched GO terms that are statistically significant at cutoff 
enriched.GO.05.a<-GO.wall$category[GO.wall$over_represented_pvalue<.05]
enriched.GO.05<-data.frame(enriched.GO.05.a)
colnames(enriched.GO.05) <- c("category")
enriched.GO.05 <- merge(enriched.GO.05, GO.wall, by="category")
enriched.GO.05 <- enriched.GO.05[order(-enriched.GO.05$numDEInCat),]
enriched.GO.05$term <- as.factor(enriched.GO.05$term)
head(enriched.GO.05)

MF <- subset(enriched.GO.05, ontology=="MF")
MF <- MF[order(-MF$numDEInCat),]
CC <- subset(enriched.GO.05, ontology=="CC")
CC <- CC[order(-CC$numDEInCat),]
BP <- subset(enriched.GO.05, ontology=="BP")
BP <- BP[order(-BP$numDEInCat),]

library(forcats)
MFplot <- MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
MFplot
CCplot <- CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
CCplot
BPplot <- BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background
BPplot
library(gridExtra)
GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

GOplot2 <- enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
GOplot2
```



















dds.d7
vsd <- vst(dds.d21, blind = TRUE)
sampleDists <- dist(t(assay(dds.d21)), method = "manhattan")
sampleDistMatrix <- as.matrix(sampleDists)
## Calculate MDS
mds <- as.data.frame(colData(vsd)) %>% 
  cbind(cmdscale(sampleDistMatrix))
# Calculate group centroids for plotting
mds <- mds %>%
  group_by(Primary_Treatment) %>%
  dplyr::summarise(c1 = mean(`1`), c2 = mean(`2`)) %>%    
  full_join(mds)
# Calculate variance explained by each PC
MDS <- cmdscale(sampleDistMatrix, eig = TRUE)
vexpl <- round(MDS$eig*100/sum(MDS$eig),1)[1:2]
## ggplot theme
theme_custom <- function() {
 # theme_classic(base_size = 10) %+replace%    #, base_family = "Arial"
  theme_bw(base_size = 10) %+replace%    #, base_family = "Arial"
    theme(
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(), 
      panel.background = element_blank(),
      panel.border = element_rect(color = "black", fill = NA),
      legend.background = element_rect(fill = NA, colour = NA),
      axis.text.x = element_text(angle=45, hjust=1, vjust = 1)#,
    )
}
# Plot with spiders
pcoa <- ggplot(mds, aes(color = Primary_Treatment, shape = Second_Treament)) +
  geom_segment(mapping = aes(x = `1`, y = `2`, xend = c1, yend = c2),
               lwd = 0.25, col = "grey") +
  geom_point(size = 2, aes(x = c1, y = c2, fill = Primary_Treatment)) +
  geom_point(size = 2, aes(x = c1, y = c2)) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`, fill = Primary_Treatment, show.legend = F)) +
  geom_point(size = 0.7, aes(x = `1`, y = `2`), show.legend = F) +
  scale_shape_manual(values = c(21, 24, 19)) +
  #scale_alpha_manual(values = c(1, 0), name = "temp", labels = c("control", "heated")) +
  guides(shape = guide_legend(order = 2, override.aes = list(fill = "black"))) +
  #guides(alpha = guide_legend(override.aes = list(shape = 21, alpha = 1, fill = c("black", "white")))) +
  labs(x = paste0("PC1 [", vexpl[1],"%]"), y = paste0("PC2 [", vexpl[2],"%]")) +
  theme_custom() +
  theme(legend.spacing.y = unit(0, "cm"))
pcoa