---
title: "GO_Analysis"
author: "Samuel Gurr"
date: "February 1, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Load libraries
```{r load_libraries, include = TRUE}
# load libraries - notes show the install command needed to install (pre installed)
library(goseq)
library(dplyr)
library(forcats)
library(ggplot2)
library(gridExtra)
library(tidyr)
library(grDevices)
```


### Set working directory
```{r set_wd, include = TRUE}

# SET WORKING DIRECTORY AND LOAD DATA
setwd("C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/")
path_out='C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/GO/'
```



# LOAD DATA:

### Annotation and DE data (from DESeq2 Rmd script)
```{r Load_annot_and_DE_files, include = TRUE}
### Panopea generosa - load .fna ('Geoduck_annotation') and foramt GO terms ('Geoduck_GOterms') and vectors
Geoduck_annotation <- read.delim2(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Seq_details/Panopea-generosa-genes-annotations.txt", header=F)

# DE files - all genes and filtered raw counts
#  contains boolean for upregulated and downregulated DEGs
# (1) DEGs: Primary Ambient_v_Moderate 
DE_d0.primary <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day0/DE_Day0_Primary.csv", sep=',', header=TRUE) 
DE_d7.primary <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day7/DE_Day7_Primary.csv", sep=',', header=TRUE) 
DE_d14.primary <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day14/DE_Day14_Primary.csv", sep=',', header=TRUE) 
DE_d21.primary <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/Day21/DE_Day21_Primary.csv", sep=',', header=TRUE) 

# All mapped read counts (unfiltered)
all.counts_matrix <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/filtered_counts/all.counts.filtered_10cpm50perc.csv", sep=',', header=TRUE) 

All_DEGs_Primary.Ambieant_v_Moderate<-read.csv("C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/DE_PrimaryTreatment.All.csv")

```

```{r Primary_Treatment_constant_DEGs}
# Summary table for the frequency that DEGs occured 
# unique(All_DEGs_Primary.Ambieant_v_Moderate$Day_Trmt) # unique DAy_treat to filter
DE_freq.upreg <-All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(up == TRUE) %>% 
            group_by(Row.names, up) %>%
            summarise(Freq = n()) %>% 
            arrange(desc(Freq))
DE_freq.upreg.summary <- DE_freq.upreg %>% group_by(Freq) %>% summarise(num.transcripts = n(),Direction = "upregulated")

DE_freq.downreg <-All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(down == TRUE) %>% 
            group_by(Row.names, down) %>%
            summarise(Freq = n()) %>%
            arrange(desc(Freq))
DE_freq.downreg.summary  <- DE_freq.downreg %>% group_by(Freq) %>% summarise(num.transcripts = n(),Direction = "downregulated")

knitr::kable(rbind(DE_freq.upreg.summary, DE_freq.downreg.summary),caption = "Primary Treatment DEGs, occurances on days 7, 14, 21: (1 = unique; 2 = d days; 3 = all days)")

# This knitr table shows that there are 10 downregulated genes and 13 upregulated genes on days 7 14 and 21
# downregualted == higher gene expression by the Moderate history (relative to ambient)
# upreagulted == higher gene expression by the Ambient history (relative to moderate)
# Let's identify these genes for GO analysis 

Contant_upreg <- DE_freq.upreg %>% dplyr::filter(Freq %in% 3) 
Constant_upreg_DEGs <- Contant_upreg$Row.names

Contant_dwnreg <- DE_freq.downreg %>% dplyr::filter(Freq %in% 3) 
Contant_dwnreg_DEGs <- Contant_dwnreg$Row.names


# Calculate the mean logfold change in the constant degs 
Mean.LFC_downreg <- All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(down == TRUE) %>% 
            filter(Row.names %in% Contant_dwnreg$Row.names) %>% # call only rownames (Gene IDs) from the constant downreg dataset (8 total)
            group_by(Row.names) %>% 
            summarise(Freq = n(), # should be 3 for all - just a quick check that this works...
                      mean.LFC = mean(log2FoldChange),
                      sd.LFC = sd(log2FoldChange)) %>%
            arrange(desc(mean.LFC))

Mean.LFC_upreg <- All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(up == TRUE) %>% 
            filter(Row.names %in% Contant_upreg$Row.names) %>% # call only rownames (Gene IDs) from the constant downreg dataset (8 total)
            group_by(Row.names) %>% 
            summarise(Freq = n(), # should be 3 for all - just a quick check that this works...
                      mean.LFC = mean(log2FoldChange),
                      sd.LFC = sd(log2FoldChange)) %>%
            arrange(desc(mean.LFC))


# build annotation file to merge with the mean LFC tables
annot.condenced <- Geoduck_annotation[,c(1,3:9)]
annot.condenced$gene.length <- annot.condenced$V4 - annot.condenced$V3
annot.condenced <- annot.condenced[,-c(2,3)]
names(annot.condenced) <- c('Gene.ID', 'Uniprot', 'HGNC', 'fxn', 'Go.terms', 'Go.fxns','gene.length')

# merge LFC mean constant primary genes with the annotation file 
names(Mean.LFC_upreg)[1] <- "Gene.ID" # first change the name from Row.names
names(Mean.LFC_downreg)[1] <- "Gene.ID" # first change the name from Row.names
LFC_ALL.constant.primary <- rbind(Mean.LFC_upreg, Mean.LFC_downreg) # bind these
Constant.primary.MASTER <- merge(LFC_ALL.constant.primary, annot.condenced, by ="Gene.ID") # merge with the annotation for Go HGNC, fxn, gene length, etc. 

write.csv(Constant.primary.MASTER,paste(path_out,"Constant.Primary_master.csv"))
```


### GO_MWU
# Create all files for the GO_MWU.R script
# output two-column csv of just the gene.id and log2fc 
```{r GO_MWU files - Primary_Treatment}
DE_d0.P <- data.frame(DE_d0.primary[,c(2,4)])
write.csv(DE_d0.P,"C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d0.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d7.P <- data.frame(DE_d7.primary[,c(2,4)])
write.csv(DE_d7.P,"C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d7.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d14.P <- data.frame(DE_d14.primary[,c(2,4)])
write.csv(DE_d14.P,"C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d14.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d21.P <- data.frame(DE_d21.primary[,c(2,4)])
write.csv(DE_d21.P,"C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d21.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

```





### Prep data by Date and direction of DE 
```{r}
# Objective here is to call the DE genes (up and down) 
# DAY 0 ------------------------------------------- #
dim(DE_d0.primary) # day0 - should be 13
DE_d0.primary.UPREG  <- DE_d0.primary %>% dplyr::filter(up %in% TRUE)               # UPREG call
DE_d0.primary.DWNREG <- DE_d0.primary %>% dplyr::filter(down %in% TRUE)             # DWNREG call
# DAY 7 ------------------------------------------- #
dim(DE_d7.primary) # day7 - should be 111
DE_d7.primary.UPREG  <- DE_d7.primary %>% dplyr::filter(up %in% TRUE)               # UPREG call
DE_d7.primary.DWNREG <- DE_d7.primary %>% dplyr::filter(down %in% TRUE)             # DWNREG call
# DAY 14 ------------------------------------------- #
dim(DE_d14.primary) # day14 - should be 511
DE_d14.primary.UPREG  <- DE_d14.primary %>% dplyr::filter(up %in% TRUE)             # UPREG call
DE_d14.primary.DWNREG <- DE_d21.primary %>% dplyr::filter(down %in% TRUE)           # DWNREG call
# DAY 21 ------------------------------------------- #
dim(DE_d21.primary) # day21 - should be 233
DE_d21.primary.UPREG  <- DE_d21.primary %>% dplyr::filter(up %in% TRUE)             # UPREG call
DE_d21.primary.DWNREG <- DE_d21.primary %>% dplyr::filter(down %in% TRUE)           # DWNREG call

```




# Genewalk

link to tutorial: https://churchman.med.harvard.edu/genewalk/#tutorial
Citation (bioRxiv preprint): Ietswaart, R., Gyori, B. M., Bachman, J. A., Sorger, P. K., & Churchman, L. S. (2019). GeneWalk identifies relevant gene functions for a biological context using network representation learning. bioRxiv, 755579.
About: Genewalk is curently designed to perform function network analysis of DEGs with human and mouse identifiers



```{r Genewalk_files}
Geoduck.gene_HGNC.list <- Geoduck_annotation %>% dplyr::select(c("V1","V6")) # call the geoduck gene ID and the HGNC in the annotation file
colnames(Geoduck.gene_HGNC.list) <- c("gene.ID", "HGNC_symbol") # rename columns
# note that the HGNC column contains the HNC code with the model taxa as 'HUMAN, MOUSE, CHICK, etc.'
Geoduck.gene_HGNC.list$HGNC<- sapply(strsplit((Geoduck.gene_HGNC.list$HGNC), "_"), '[', 1) # this splits the HGNC by the delimiter "_" and calls just the HGNC code - note the model taxa is now absent from the dataframe! 


# change the col name of the gene.ID  in the DE datasets

# DAY 0 ------------------------------------------- #

# UPREGULATED
colnames(DE_d0.primary.UPREG)[2] <- "gene.ID"
GW.d0.primary.UPREG <- merge((DE_d0.primary.UPREG[order(DE_d0.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d0.upreg.percHGNC <- ((dim(na.omit(GW.d0.primary.UPREG))[1]) / (dim(DE_d0.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
colnames(DE_d0.primary.DWNREG)[2] <- "gene.ID"
GW.d0.primary.DWNREG <- merge((DE_d0.primary.DWNREG[order(DE_d0.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d0.downreg.percHGNC <- ((dim(na.omit(GW.d0.primary.DWNREG))[1]) / (dim(DE_d0.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 7 ------------------------------------------- #

# UPREGULATED
colnames(DE_d7.primary.UPREG)[2] <- "gene.ID"
GW.d7.primary.UPREG <- merge((DE_d7.primary.UPREG[order(DE_d7.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d7.upreg.percHGNC <- ((dim(na.omit(GW.d7.primary.UPREG))[1]) / (dim(DE_d7.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
colnames(DE_d7.primary.DWNREG)[2] <- "gene.ID"
GW.d7.primary.DWNREG <- merge((DE_d7.primary.DWNREG[order(DE_d7.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d7.downreg.percHGNC <- ((dim(na.omit(GW.d7.primary.DWNREG))[1]) / (dim(DE_d7.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 14 ------------------------------------------- #

# UPREGULATED
colnames(DE_d14.primary.UPREG)[2] <- "gene.ID"
GW.d14.primary.UPREG <- merge((DE_d14.primary.UPREG[order(DE_d14.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d14.upreg.percHGNC <- ((dim(na.omit(GW.d14.primary.UPREG))[1]) / (dim(DE_d14.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
colnames(DE_d14.primary.DWNREG)[2] <- "gene.ID"
GW.d14.primary.DWNREG <- merge((DE_d14.primary.DWNREG[order(DE_d14.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC") # sort to order by greatest logfold change and merge with HGNC list
d14.downreg.percHGNC <- ((dim(na.omit(GW.d14.primary.DWNREG))[1]) / (dim(DE_d14.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 21 ------------------------------------------- #

# UPREGULATED
colnames(DE_d21.primary.UPREG)[2] <- "gene.ID"
GW.d21.primary.UPREG <- merge((DE_d21.primary.UPREG[order(DE_d21.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d21.upreg.percHGNC <- ((dim(na.omit(GW.d21.primary.UPREG))[1]) / (dim(DE_d21.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
colnames(DE_d21.primary.DWNREG)[2] <- "gene.ID"
GW.d21.primary.DWNREG <- merge((DE_d21.primary.DWNREG[order(DE_d21.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d21.downreg.percHGNC <- ((dim(na.omit(GW.d21.primary.DWNREG))[1]) / (dim(DE_d21.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID


### OUTPUT Genewalk tables as csv wihtout header
GW.d0.primary.UPREG.om   <- na.omit(GW.d0.primary.UPREG)
GW.d0.primary.DWNREG.om  <- na.omit(GW.d0.primary.DWNREG)

GW.d7.primary.UPREG.om   <- na.omit(GW.d7.primary.UPREG)
GW.d7.primary.DWNREG.om  <- na.omit(GW.d7.primary.DWNREG) 

GW.d14.primary.UPREG.om  <- na.omit(GW.d14.primary.UPREG)
GW.d14.primary.DWNREG.om <- na.omit(GW.d14.primary.DWNREG)

GW.d21.primary.UPREG.om  <- na.omit(GW.d21.primary.UPREG)
GW.d21.primary.DWNREG.om <- na.omit(GW.d21.primary.DWNREG)

path_genewalk='C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/Genewalk_files/' # call path
# output files 
write.table(GW.d0.primary.UPREG.om,paste(path_genewalk,"Day0_upreg.csv", sep =''),sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d0.primary.DWNREG.om, paste(path_genewalk,"Day0_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)

write.table(GW.d7.primary.UPREG.om,paste(path_genewalk,"Day7_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d7.primary.DWNREG.om,paste(path_genewalk,"Day7_downreg.csv", sep =''), sep=",", col.names=FALSE, row.names=FALSE, quote=FALSE)

write.table(GW.d14.primary.UPREG.om,paste(path_genewalk,"Day14_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d14.primary.DWNREG.om,paste(path_genewalk,"Day14_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)

write.table(GW.d21.primary.UPREG.om,paste(path_genewalk,"Day21_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d21.primary.DWNREG.om,paste(path_genewalk,"Day21_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
```


```{r Genewalk_knitr_table}
GW_summarytable <- data.frame(matrix(nrow = 4, ncol = 8)) # create a new data table
colnames(GW_summarytable)<-c('Day', 'Treatment', 'DEGs_upreg', 'numHGNC_upreg', 'percHGNC_upreg', 'DEGs_dwnreg', 'numHGNC_dwnreg', 'percHGNC_dwnreg') 
GW_summarytable$Day <- c("0","7","14","21")
GW_summarytable$Treatment <- "Ambient_v_Moderate"
GW_summarytable$DEGs_upreg <- c((dim(DE_d0.primary.UPREG)[1]), (dim(DE_d7.primary.UPREG)[1]), (dim(DE_d14.primary.UPREG)[1]), (dim(DE_d21.primary.UPREG)[1]))
GW_summarytable$numHGNC_upreg <- c((dim(na.omit(GW.d0.primary.UPREG))[1]),(dim(na.omit(GW.d7.primary.UPREG))[1]),(dim(na.omit(GW.d14.primary.UPREG))[1]),(dim(na.omit(GW.d21.primary.UPREG))[1]))
GW_summarytable$percHGNC_upreg <- c(d0.upreg.percHGNC,d7.upreg.percHGNC,d14.upreg.percHGNC,d21.upreg.percHGNC)
GW_summarytable$DEGs_dwnreg <- c((dim(DE_d0.primary.DWNREG)[1]), (dim(DE_d7.primary.DWNREG)[1]), (dim(DE_d14.primary.DWNREG)[1]), (dim(DE_d21.primary.DWNREG)[1]))
GW_summarytable$numHGNC_dwnreg<- c((dim(na.omit(GW.d0.primary.DWNREG))[1]),(dim(na.omit(GW.d7.primary.DWNREG))[1]),(dim(na.omit(GW.d14.primary.DWNREG))[1]),(dim(na.omit(GW.d21.primary.DWNREG))[1]))
GW_summarytable$percHGNC_dwnreg <- c(d0.downreg.percHGNC,d7.downreg.percHGNC,d14.downreg.percHGNC,d21.downreg.percHGNC)

knitr::kable(GW_summarytable, caption = "Summary table of Genewalk files - total DEGs and mapping abund HGNC IDs")
# Write csv
write.csv(GW_summarytable,paste(path_genewalk,"Summary_HGNC_hits.csv"))
```




# goseq



Prepare dataframe(s) and vectors for goseq 
### (1) Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
```{r GO.terms_data}
# call the GO terms
Geoduck_GOterms <- as.data.frame(Geoduck_annotation) %>% dplyr::select(c('V1','V8'))
colnames(Geoduck_GOterms)[1:2] <- c('transcript.ID', 'GO.terms') # call gene name and the GO terms - (Uniprot ID 'V5')
splitted <- strsplit(as.character(Geoduck_GOterms$GO.terms), ";") #slit into multiple GO ids
GO.terms <- data.frame(v1 = rep.int(Geoduck_GOterms$transcript.ID, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
```

### (2) Unique Genes - vector based on all unique mapped reads 
```{r Unique_genes}
# Construct a named vector of DEGs suritable for goseq
# measured Genes 
all.counts_matrix <- data.frame(all.counts_matrix[,-1], row.names=all.counts_matrix[,1]) # ID column to rownames
all.counts <- tibble::rownames_to_column(all.counts_matrix, "gene.ID") # call the rownames as gene.ID
GO_unique.genes.all <- as.vector(unique(all.counts$gene.ID)) # call all unique genes for GO analysis (goseq)


# GO_unique.genes.d0 <- as.vector(unique(row.names(d0.counts_matrix)))   # call unique genes for GO analysis (goseq)
# GO_unique.genes.d7 <- as.vector(unique(row.names(d7.counts_matrix)))   # call unique genes for GO analysis (goseq)
# GO_unique.genes.d14 <- as.vector(unique(row.names(d14.counts_matrix))) # call unique genes for GO analysis (goseq)
# GO_unique.genes.d21 <- as.vector(unique(row.names(d21.counts_matrix))) # call unique genes for GO analysis (goseq)
```

### (3) Gene length 
```{r Gene_length}
# length vector  
GO_gene.length <- Geoduck_annotation %>% dplyr::mutate(length = V4-V3) %>%  dplyr::select(c("V1","length"))
names(GO_gene.length)[1] <- "gene.ID"

GO_gene.length_merge <- merge(GO_gene.length, all.counts, by = "gene.ID")
length_vector <- GO_gene.length_merge$length

```

### (4) Call DE datasets (from DESeq2) - merge to the genes list 'GO_unique.genes.all' to create a binary vector 0 = not DE'd; 1 = DE'd
```{r Binary_DEGs}
# (1) DEGs: Primary Ambient_v_Moderate 
# DAY 0 ------------------------------------------- 
DE.d0._Primary.UP    <-as.integer(GO_unique.genes.all%in%(DE_d0.primary.UPREG$gene.ID)) # Day 0 - Primary - Upregulated DEGs
names(DE.d0._Primary.UP)=GO_unique.genes.all

DE.d0._Primary.DOWN  <-as.integer(GO_unique.genes.all%in%(DE_d0.primary.DWNREG$gene.ID)) # Day 0 - Primary - Downregulated DEGs
names(DE.d0._Primary.DOWN)=GO_unique.genes.all

DE.d0._Primary.ALL   <-as.integer(GO_unique.genes.all%in%(DE_d0.primary$gene.ID)) # Day 0 - Primary - All DEGs
names(DE.d0._Primary.ALL)=GO_unique.genes.all

# DAY 7 ------------------------------------------- #
DE.d7._Primary.UP    <-as.integer(GO_unique.genes.all%in%(DE_d7.primary.UPREG$gene.ID)) # Day 0 - Primary - Upregulated DEGs
names(DE.d7._Primary.UP)=GO_unique.genes.all

DE.d7._Primary.DOWN  <-as.integer(GO_unique.genes.all%in%(DE_d7.primary.DWNREG$gene.ID)) # Day 0 - Primary - Downregulated DEGs
names(DE.d7._Primary.DOWN)=GO_unique.genes.all

DE.d7._Primary.ALL   <-as.integer(GO_unique.genes.all%in%(DE_d7.primary$gene.ID)) # Day 0 - Primary - All DEGs
names(DE.d7._Primary.ALL)=GO_unique.genes.all

# DAY 14 ------------------------------------------- #
DE.d14._Primary.UP    <-as.integer(GO_unique.genes.all%in%(DE_d14.primary.UPREG$gene.ID)) # Day 0 - Primary - Upregulated DEGs
names(DE.d14._Primary.UP)=GO_unique.genes.all

DE.d14._Primary.DOWN  <-as.integer(GO_unique.genes.all%in%(DE_d14.primary.DWNREG$gene.ID)) # Day 0 - Primary - Downregulated DEGs
names(DE.d14._Primary.DOWN)=GO_unique.genes.all

DE.d14._Primary.ALL   <-as.integer(GO_unique.genes.all%in%(DE_d14.primary$gene.ID)) # Day 0 - Primary - All DEGs
names(DE.d14._Primary.ALL)=GO_unique.genes.all

# DAY 21 ------------------------------------------- #
DE.d21._Primary.UP    <-as.integer(GO_unique.genes.all%in%(DE_d21.primary.UPREG$gene.ID)) # Day 0 - Primary - Upregulated DEGs
names(DE.d21._Primary.UP)=GO_unique.genes.all

DE.d21._Primary.DOWN  <-as.integer(GO_unique.genes.all%in%(DE_d21.primary.DWNREG$gene.ID)) # Day 0 - Primary - Downregulated DEGs
names(DE.d21._Primary.DOWN)=GO_unique.genes.all

DE.d21._Primary.ALL   <-as.integer(GO_unique.genes.all%in%(DE_d21.primary$gene.ID)) # Day 0 - Primary - All DEGs
names(DE.d21._Primary.ALL)=GO_unique.genes.all

#Constant DEGs ------------------------------------ #
# Note: these are degenes that were upregulated and downregulated throughout the 21-day experiment due to the effect of primary treatment
# data from the cluster {r Primary_Treatment_constant_DEGs}
DE_Contant_upreg <-as.integer(GO_unique.genes.all%in%(Contant_upreg$Row.names)) # Day 0 - Primary - All DEGs
names(DE_Contant_upreg)=GO_unique.genes.all

DE_Contant_dwnreg <-as.integer(GO_unique.genes.all%in%(Contant_dwnreg$Row.names)) # Day 0 - Primary - All DEGs
names(DE_Contant_dwnreg)=GO_unique.genes.all
```

### Let's review what we have for goseq
```{r NOTE_checklist}
# Review what we have for goseq....

# (1) Go terms ---------------- #
#  head(GO.terms)

# (2) ID vector --------------- #
# head(GO_unique.genes.all) # == ID.vector

# (3) Length vector ----------- #
# head(length_vector)

# (4) Binary DEG vectors ------ # 
# head(DE.d21._Primary.UP)    # example of d21 prmary ambient_v_moderate upregulated DEGs
# head(DE.d21._Primary.DOWN)  # example of d21 prmary ambient_v_moderate downregulated DEGs
# head(DE.d21._Primary.ALL)   # example of d21 prmary ambient_v_moderate all DEGs
```


# It's goseq time!!!
### Calculate Probability Weighting Function
```{r pwf}
#Calculate Probability Weighting Function (using 'nullp')

# (1) DEGs: Primary Ambient_v_Moderate 
# DAY 0 ------------------------------------------- #
DEG.pwf_d0.all<-nullp(DE.d0._Primary.ALL, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_d0.UP<-nullp(DE.d0._Primary.UP, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_d0.DOWN<-nullp(DE.d0._Primary.DOWN, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
# DAY 7 ------------------------------------------- #
DEG.pwf_d7.all<-nullp(DE.d7._Primary.ALL, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_d7.UP<-nullp(DE.d7._Primary.UP, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_d7.DOWN<-nullp(DE.d7._Primary.DOWN, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
# DAY 14 ------------------------------------------- #
DEG.pwf_d14.all<-nullp(DE.d14._Primary.ALL, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_d14.UP<-nullp(DE.d14._Primary.UP, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_d14.DOWN<-nullp(DE.d14._Primary.DOWN, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
# DAY 21 ------------------------------------------- #
DEG.pwf_d21.all<-nullp(DE.d21._Primary.ALL, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_d21.UP<-nullp(DE.d21._Primary.UP, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_d21.DOWN<-nullp(DE.d21._Primary.DOWN, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
# Constant DEGs ------------------------------------ #
DEG.pwf_Constant.UP<-nullp(DE_Contant_upreg, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
DEG.pwf_Constant.DOWN<-nullp(DE_Contant_dwnreg, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene

```

### Run goseq
```{r goseq}
# (1) DEGs: Primary Ambient_v_Moderate 
# DAY 0 ------------------------------------------- #
GO_d0.all<-goseq(DEG.pwf_d0.all, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d0.UP<-goseq(DEG.pwf_d0.UP, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d0.DOWN<-goseq(DEG.pwf_d0.DOWN, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# DAY 7 ------------------------------------------- #
GO_d7.all<-goseq(DEG.pwf_d7.all, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d7.UP<-goseq(DEG.pwf_d7.UP, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d7.DOWN<-goseq(DEG.pwf_d7.DOWN, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# DAY 14 ------------------------------------------- #
GO_d14.all<-goseq(DEG.pwf_d14.all, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d14.UP<-goseq(DEG.pwf_d14.UP, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d14.DOWN<-goseq(DEG.pwf_d14.DOWN, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# DAY 21 ------------------------------------------- #
GO_d21.all<-goseq(DEG.pwf_d21.all, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d21.UP<-goseq(DEG.pwf_d21.UP, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d21.DOWN<-goseq(DEG.pwf_d21.DOWN, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# Constant DEGs ------------------------------------ #
GO_Constant.UP<-goseq(DEG.pwf_Constant.UP, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_Constant.DOWN<-goseq(DEG.pwf_Constant.DOWN, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
```

```{r Constant.primary.GO table}
GO_Constant.UP_summary <- GO_Constant.UP %>% dplyr::select(c('category','numDEInCat','numInCat', 'term', 'ontology')) %>% dplyr::filter(!is.na(term)) %>% dplyr::filter(numDEInCat > 0)
GO_Constant.UP_summary$DE_dir <- "upregulated"
GO_Constant.DOWN_summary <- GO_Constant.DOWN %>% dplyr::select(c('category','numDEInCat','numInCat', 'term', 'ontology')) %>% dplyr::filter(!is.na(term)) %>% dplyr::filter(numDEInCat > 0)
GO_Constant.DOWN_summary$DE_dir <- "downregulated"

GO_constant.Master <- rbind(GO_Constant.UP_summary,GO_Constant.DOWN_summary)
GO_constant.Master$perc.occurance <- GO_constant.Master$numDEInCat / GO_constant.Master$numInCat *100

#knitr
knitr::kable(GO_constant.Master, caption = "Summary table of constant GO terms Days 7, 14, 21")
# Write csv
write.csv(GO_constant.Master,paste(path_out,"Constant.Primary_GO.terms.csv"))
```

### Plotting: Day 0
```{r plots_Day0.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d0.UP$category[GO_d0.UP$over_represented_pvalue<.05] # change twice here
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d0.UP, by="category") # change here
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- # 
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d0.DOWN$category[GO_d0.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d0.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d0.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d0.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 0 Upregulated DEGs") +
  geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 0 Downregulated DEGs") +
  geom_label(aes(x = 0.5, y = 1.5, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d0.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day0/GOplot_d0.Primary.pdf", GO.plot_d0.Primary, width = 21, height = 21, units = c("in"))
getwd()
```

### Plotting: Day 7
```{r plots_Day7.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d7.UP$category[GO_d7.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d7.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d7.DOWN$category[GO_d7.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d7.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d7.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d7.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 7 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 2, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 7 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 2, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d7.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day7/GOplot_d7.Primary.pdf", GO.plot_d7.Primary, width = 21, height = 21, units = c("in"))
```

### Plotting: Day 14
```{r plots_Day14.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d14.UP$category[GO_d14.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d14.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d14.DOWN$category[GO_d14.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d14.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d14.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d14.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 14 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 20, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 14 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 6, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d14.Primary<- grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day14/GOplot_d14.Primary.pdf", GO.plot_d14.Primary, width = 21, height = 21, units = c("in"))
```

### Plotting: Day 21
```{r plots_Day21.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d21.UP$category[GO_d21.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d21.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d21.DOWN$category[GO_d21.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d21.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d21.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d21.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 21 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 7, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 21 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 3, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d21.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day21/GOplot_d21.Primary.pdf", GO.plot_d21.Primary, width = 21, height = 21, units = c("in"))
```




### Plotting: Constant DEGs Days 7 14 21 affected by Primary treatment history
```{r Constant_All}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_Constant.UP$category[GO_Constant.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_Constant.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_Constant.DOWN$category[GO_Constant.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_Constant.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(Contant_upreg)[1] # call num upregulated genes
num_down <-dim(Contant_dwnreg)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Constant Upregulated DEGs") +
  geom_label(aes(x = 1, y = 3, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Constant Downregulated DEGs") +
  geom_label(aes(x = 1, y = 3, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_Constant<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/GO/GOplot_Constant.Primary.pdf", GO.plot_Constant, width = 21, height = 21, units = c("in"))
```