---
title: "GO_Analysis_DESeq2_10cpm"
author: "Samuel Gurr"
date: "February 1, 2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


NOTE: after runinng this once... you can load the liibraries (below) and skip to 'the 'Find GOslim terms' to load the .csv files for plotting
Load libraries
```{r load_libraries, include = TRUE}
# load libraries - notes show the install command needed to install (pre installed)
library(goseq)
library(dplyr)
library(forcats)
library(ggplot2)
library(gridExtra)
library(tidyr)
library(grDevices)
library(dplyr)
library(goseq)
library(reshape2)
library(ggplot2)
library(Rmisc)
library(ggpubr)
library(tibble)
library(hrbrthemes)
library(gridExtra)
library(ggplot2)
library(tidyr)
library(forcats) # for plotting later..
library(zoo)
library(ComplexHeatmap)
library(circlize)
library(GSEABase)
library(data.table)
```



Set working directory and outpath
```{r set_wd, include = TRUE}

# SET WORKING DIRECTORY AND LOAD DATA
setwd("C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/")
path_out='C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/GO/'
```



LOAD DATA:
Annotation and DE data (from DESeq2 Rmd script)
```{r Load_annot_and_DE_files, include = TRUE}
### Panopea generosa - load .fna ('Geoduck_annotation') and foramt GO terms ('Geoduck_GOterms') and vectors
Geoduck_annotation <- read.delim2(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Seq_details/Panopea-generosa-genes-annotations.txt", header=F)

# DE files - all genes and filtered raw counts
#  contains boolean for upregulated and downregulated DEGs
# DEGs: Primary Effect  - Ambient_v_Moderate
DE_d0.primary <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/10cpm/Day0/DE_Day0_Primary.csv", sep=',', header=TRUE)
colnames(DE_d0.primary)[2] <- "gene.ID" # rename Pgen gene ID column

DE_d7.primary <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/10cpm/Day7/DE_Day7_Primary.csv", sep=',', header=TRUE) 
colnames(DE_d7.primary)[2] <- "gene.ID"# rename Pgen gene ID column

DE_d14.primary <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/10cpm/Day14/DE_Day14_Primary.csv", sep=',', header=TRUE) 
colnames(DE_d14.primary)[2] <- "gene.ID"# rename Pgen gene ID column

DE_d21.primary <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/10cpm/Day21/DE_Day21_Primary.csv", sep=',', header=TRUE) 
colnames(DE_d21.primary)[2] <- "gene.ID"# rename Pgen gene ID column

#  mapped read counts for each Day (addressed separately in DESeq2)
# day0 filtered 10cpm in 50% samples
Day0_all.counts <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/Filtered_Counts/10cpm_50perc/day0.counts.filtered_10cpm50perc.csv", sep=',', header=TRUE) 
colnames(Day0_all.counts)[1] <- "gene.ID"# rename Pgen gene ID column

# day7 filtered 10cpm in 50% samples
Day7_all.counts <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/Filtered_Counts/10cpm_50perc/day7.counts.filtered_10cpm50perc.csv", sep=',', header=TRUE) 
colnames(Day7_all.counts)[1] <- "gene.ID"# rename Pgen gene ID column

# day14 filtered 10cpm in 50% samples
Day14_all.counts <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/Filtered_Counts/10cpm_50perc/day14.counts.filtered_10cpm50perc.csv", sep=',', header=TRUE) 
colnames(Day14_all.counts)[1] <- "gene.ID"# rename Pgen gene ID column

# day21 filtered 10cpm in 50% samples
Day21_all.counts <- read.csv(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/Filtered_Counts/10cpm_50perc/day21.counts.filtered_10cpm50perc.csv", sep=',', header=TRUE) 
colnames(Day21_all.counts)[1] <- "gene.ID"# rename Pgen gene ID column

All_DEGs_Primary.Ambieant_v_Moderate<-read.csv("C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/10cpm/DE_PrimaryTreatment.All.csv")

```



Prep Differential Expression data (from DESeq2) by Date and direction of DE 
```{r}
# Objective here is to call the DE genes (up and down) 
# DAY 0 ------------------------------------------- #
dim(DE_d0.primary) # day0 - should be 14
DE_d0.primary.UPREG  <- DE_d0.primary %>% dplyr::filter(up %in% TRUE)               # UPREG call
DE_d0.primary.DWNREG <- DE_d0.primary %>% dplyr::filter(down %in% TRUE)             # DWNREG call
# DAY 7 ------------------------------------------- #
dim(DE_d7.primary) # day7 - should be 108
DE_d7.primary.UPREG  <- DE_d7.primary %>% dplyr::filter(up %in% TRUE)               # UPREG call
DE_d7.primary.DWNREG <- DE_d7.primary %>% dplyr::filter(down %in% TRUE)             # DWNREG call
# DAY 14 ------------------------------------------- #
dim(DE_d14.primary) # day14 - should be 429
DE_d14.primary.UPREG  <- DE_d14.primary %>% dplyr::filter(up %in% TRUE)             # UPREG call
DE_d14.primary.DWNREG <- DE_d14.primary %>% dplyr::filter(down %in% TRUE)           # DWNREG call
# DAY 21 ------------------------------------------- #
dim(DE_d21.primary) # day21 - should be 155
DE_d21.primary.UPREG  <- DE_d21.primary %>% dplyr::filter(up %in% TRUE)             # UPREG call
DE_d21.primary.DWNREG <- DE_d21.primary %>% dplyr::filter(down %in% TRUE)           # DWNREG call

# All genes  ------------------------------------------- #
dim(All_DEGs_Primary.Ambieant_v_Moderate) # day21 - should be 155
All.primary.UPREG  <- All_DEGs_Primary.Ambieant_v_Moderate %>% dplyr::filter(up %in% TRUE)             # UPREG call
All.primary.DWNREG <- All_DEGs_Primary.Ambieant_v_Moderate %>% dplyr::filter(down %in% TRUE)           # DWNREG call
nrow(All.primary.DWNREG) ==sum( (nrow(DE_d21.primary.DWNREG)) , (nrow(DE_d14.primary.DWNREG)) , (nrow(DE_d7.primary.DWNREG)), (nrow(DE_d0.primary.DWNREG))) # SHOULD BE TRUE
nrow(All.primary.UPREG) ==sum( (nrow(DE_d21.primary.UPREG)) , (nrow(DE_d14.primary.UPREG)) , (nrow(DE_d7.primary.UPREG)), (nrow(DE_d0.primary.UPREG))) # SHOULD BE TRUE
```




```{r Primary_Treatment_constant_DEGs}
# Summary table for the frequency that DEGs occured 
# unique(All_DEGs_Primary.Ambieant_v_Moderate$Day_Trmt) # unique DAy_treat to filter
DE_freq.upreg <-All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(up == TRUE) %>% 
            group_by(Row.names, up) %>%
            dplyr::summarise(Freq = n()) %>% 
            arrange(desc(Freq))
DE_freq.upreg.summary <- DE_freq.upreg %>% group_by(Freq) %>% dplyr::summarise(num.transcripts = n(),Direction = "upregulated")

DE_freq.downreg <-All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(down == TRUE) %>% 
            group_by(Row.names, down) %>%
            dplyr::summarise(Freq = n()) %>%
            arrange(desc(Freq))
DE_freq.downreg.summary  <- DE_freq.downreg %>% group_by(Freq) %>% dplyr::summarise(num.transcripts = n(),Direction = "downregulated")

knitr::kable(rbind(DE_freq.upreg.summary, DE_freq.downreg.summary),caption = "Primary Treatment DEGs, occurances on days 7, 14, 21: (1 = unique; 2 = d days; 3 = all days)")

# This knitr table shows that there are 10 downregulated genes and 13 upregulated genes on days 7 14 and 21
# downregualted == higher gene expression by the Moderate history (relative to ambient)
# upreagulted == higher gene expression by the Ambient history (relative to moderate)
# Let's identify these genes for GO analysis 

Contant_upreg <- DE_freq.upreg %>% dplyr::filter(Freq %in% 3) 
Constant_upreg_DEGs <- Contant_upreg$Row.names

Contant_dwnreg <- DE_freq.downreg %>% dplyr::filter(Freq %in% 3) 
Contant_dwnreg_DEGs <- Contant_dwnreg$Row.names


# Calculate the mean logfold change in the constant degs 
Mean.LFC_downreg <- All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(down == TRUE) %>% 
            filter(Row.names %in% Contant_dwnreg$Row.names) %>% # call only rownames (Gene IDs) from the constant downreg dataset (8 total)
            group_by(Row.names) %>% 
            dplyr::summarise(Freq = n(), # should be 3 for all - just a quick check that this works...
                      mean.LFC = mean(log2FoldChange),
                      sd.LFC = sd(log2FoldChange)) %>%
            arrange(desc(mean.LFC))

Mean.LFC_upreg <- All_DEGs_Primary.Ambieant_v_Moderate %>%
            filter(Day_Trmt %in% c('Day7_Primary_AvM', 'Day14_Primary_AvM', 'Day21_Primary_AvM')) %>% 
            filter(up == TRUE) %>% 
            filter(Row.names %in% Contant_upreg$Row.names) %>% # call only rownames (Gene IDs) from the constant downreg dataset (8 total)
            group_by(Row.names) %>% 
            dplyr::summarise(Freq = n(), # should be 3 for all - just a quick check that this works...
                      mean.LFC = mean(log2FoldChange),
                      sd.LFC = sd(log2FoldChange)) %>%
            arrange(desc(mean.LFC))


# build annotation file to merge with the mean LFC tables
annot.condenced <- Geoduck_annotation[,c(1,3:9)]
annot.condenced$gene.length <- annot.condenced$V4 - annot.condenced$V3
annot.condenced <- annot.condenced[,-c(2,3)]
names(annot.condenced) <- c('Gene.ID', 'Uniprot', 'HGNC', 'fxn', 'Go.terms', 'Go.fxns','gene.length')

# merge LFC mean constant primary genes with the annotation file 
names(Mean.LFC_upreg)[1] <- "Gene.ID" # first change the name from Row.names
names(Mean.LFC_downreg)[1] <- "Gene.ID" # first change the name from Row.names
LFC_ALL.constant.primary <- rbind(Mean.LFC_upreg, Mean.LFC_downreg) # bind these
Constant.primary.MASTER <- merge(LFC_ALL.constant.primary, annot.condenced, by ="Gene.ID") # merge with the annotation for Go HGNC, fxn, gene length, etc. 

write.csv(Constant.primary.MASTER,paste(path_out,"Constant.Primary_master.csv"))
```


goseq
Prepare dataframe(s) and vectors for goseq 
(1) Format 'GO.term' for goseq from the P.generosa annotation .fna file 'Geoduck_annotation'
```{r GO.terms_data}
# call the GO terms
Geoduck_GOterms <- as.data.frame(Geoduck_annotation) %>% dplyr::select(c('V1','V8'))
colnames(Geoduck_GOterms)[1:2] <- c('transcript.ID', 'GO.terms') # call gene name and the GO terms - (Uniprot ID 'V5')
splitted <- strsplit(as.character(Geoduck_GOterms$GO.terms), "; ") #slit into multiple GO ids by delimiter'; ' remember the space after ; is needed here! without this you will only call the first listed GO term for each gene!
GO.terms <- data.frame(v1 = rep.int(Geoduck_GOterms$transcript.ID, sapply(splitted, length)), v2 = unlist(splitted)) #list all genes with each of their GO terms in a single row
```


(2) Unique Genes - vector based on all unique mapped reads  - Remember this is based on the genes for EACH sampling day becasue DEeq2 was run separately (i.e. days 0, 7, 14, and 21)
(3) Gene length 
Set ID and gene length vectors, and make a binary matrix indicating which genes are differentially expressed. These are used as input to nullp, which for calculates a Probability Weighting Function for each set of DEGs.
```{r Unique_genes}

# call unique IDvector(s) for each of the experiment days
IDvector.d0 <- as.vector(unique(Day0_all.counts$gene.ID))   # call unique genes (those filtered and used in DESEq2) on day0 - 'IDvector'
IDvector.d7 <- as.vector(unique(Day7_all.counts$gene.ID))  # call unique genes (those filtered and used in DESEq2) on day7 - 'IDvector'
IDvector.d14 <- as.vector(unique(Day14_all.counts$gene.ID)) # call unique genes (those filtered and used in DESEq2) on day14 - 'IDvector'
IDvector.d21 <- as.vector(unique(Day21_all.counts$gene.ID)) # call unique genes (those filtered and used in DESEq2) on day21 - 'IDvector'

#  length vectors (from entire annotation file - no genes ommitted from filtering)  
GO_gene.length <- Geoduck_annotation %>% dplyr::mutate(length = V4-V3) %>%  dplyr::select(c("V1","length"))
names(GO_gene.length)[1] <- "gene.ID" # cange the column name to merge with the gene counts on each day to call ID and length vectors UNIQUE to the experiment day

# merge length with counts data
GeneLength.d0   <- merge(GO_gene.length, Day0_all.counts, by = "gene.ID")   # merge day0 counts with 'GO_gene.length' 
GeneLength.d7   <- merge(GO_gene.length, Day7_all.counts, by = "gene.ID")  # merge day7 counts with 'GO_gene.length' 
GeneLength.d14  <- merge(GO_gene.length, Day14_all.counts, by = "gene.ID")  # merge day14 counts with 'GO_gene.length' 
GeneLength.d21  <- merge(GO_gene.length, Day21_all.counts, by = "gene.ID")  # merge day21 counts with 'GO_gene.length'

# call length values for goseq
length_vector.d0 <- GeneLength.d0$length    # length vector for all unique reads address in DESeq2 on day 0 
sum(sapply(length_vector.d0,length)) == dim(Day0_all.counts)[1] #should be TRUE
length_vector.d7 <- GeneLength.d7$length    # length vector for all unique reads address in DESeq2 on day 7
sum(sapply(length_vector.d7,length)) == dim(Day7_all.counts)[1] #should be TRUE
length_vector.d14 <- GeneLength.d14$length  # length vector for all unique reads address in DESeq2 on day 14
sum(sapply(length_vector.d14,length)) == dim(Day14_all.counts)[1] #should be TRUE
length_vector.d21 <- GeneLength.d21$length  # length vector for all unique reads address in DESeq2 on day 21
sum(sapply(length_vector.d21,length)) == dim(Day21_all.counts)[1] #should be TRUE
```



(4) Call DE datasets (from DESeq2) - merge to the genes list 'GO_unique.genes.all' to create a binary vector 0 = not DE'd; 1 = DE'd
```{r Binary_DEGs - must run the prep steps above}
# (1) DEGs: Primary Ambient_v_Moderate 
# DAY 0 ------------------------------------------- #
DE.d0._Primary.UP    <-as.integer(IDvector.d0%in%(DE_d0.primary.UPREG$gene.ID)) # Day 0 - Primary - Upregulated DEGs
names(DE.d0._Primary.UP)=IDvector.d0
sum(sapply(DE.d0._Primary.UP,length)) # should be 9207
sum(DE.d0._Primary.UP == 1) # should be 11

DE.d0._Primary.DOWN  <-as.integer(IDvector.d0%in%(DE_d0.primary.DWNREG$gene.ID)) # Day 0 - Primary - Downregulated DEGs
names(DE.d0._Primary.DOWN)=IDvector.d0
sum(sapply(DE.d0._Primary.DOWN,length)) # should be 9207
sum(DE.d0._Primary.DOWN == 1) # should be 3

DE.d0._Primary.ALL   <-as.integer(IDvector.d0%in%(DE_d0.primary$gene.ID)) # Day 0 - Primary - All DEGs
names(DE.d0._Primary.ALL)=IDvector.d0
sum(sapply(DE.d0._Primary.ALL,length)) # should be 9207
sum(DE.d0._Primary.ALL == 1) # should be 14

# DAY 7 ------------------------------------------- #
DE.d7._Primary.UP    <-as.integer(IDvector.d7%in%DE_d7.primary.UPREG$gene.ID) # Day 0 - Primary - Upregulated DEGs
names(DE.d7._Primary.UP)=IDvector.d7
sum(sapply(DE.d7._Primary.UP,length)) # should be 8548
sum(DE.d7._Primary.UP == 1) # should be 62

DE.d7._Primary.DOWN  <-as.integer(IDvector.d7%in%(DE_d7.primary.DWNREG$gene.ID)) # Day 0 - Primary - Downregulated DEGs
names(DE.d7._Primary.DOWN)=IDvector.d7
sum(sapply(DE.d7._Primary.DOWN,length)) # should be 8548
sum(DE.d7._Primary.DOWN == 1) # should be 46

DE.d7._Primary.ALL   <-as.integer(IDvector.d7%in%(DE_d7.primary$gene.ID)) # Day 0 - Primary - All DEGs
names(DE.d7._Primary.ALL)=IDvector.d7
sum(sapply(DE.d7._Primary.ALL,length)) # should be 8548
sum(DE.d7._Primary.ALL == 1) # should be 108

# DAY 14 ------------------------------------------- #
DE.d14._Primary.UP    <-as.integer(IDvector.d14%in%(DE_d14.primary.UPREG$gene.ID)) # Day 0 - Primary - Upregulated DEGs
names(DE.d14._Primary.UP)=IDvector.d14
sum(sapply(DE.d14._Primary.UP,length)) # should be 8626
sum(DE.d14._Primary.UP == 1) # should be 317

DE.d14._Primary.DOWN  <-as.integer(IDvector.d14%in%(DE_d14.primary.DWNREG$gene.ID)) # Day 0 - Primary - Downregulated DEGs
names(DE.d14._Primary.DOWN)=IDvector.d14
sum(sapply(DE.d14._Primary.DOWN,length)) # should be 8626
sum(DE.d14._Primary.DOWN == 1) # should be 112

DE.d14._Primary.ALL   <-as.integer(IDvector.d14%in%(DE_d14.primary$gene.ID)) # Day 0 - Primary - All DEGs
names(DE.d14._Primary.ALL)=IDvector.d14
sum(sapply(DE.d14._Primary.ALL,length)) # should be 8626
sum(DE.d14._Primary.ALL == 1) # should be 429

# DAY 21 ------------------------------------------- #
DE.d21._Primary.UP    <-as.integer(IDvector.d21%in%(DE_d21.primary.UPREG$gene.ID)) # Day 0 - Primary - Upregulated DEGs
names(DE.d21._Primary.UP)=IDvector.d21
sum(sapply(DE.d21._Primary.UP,length)) # should be 8421
sum(DE.d21._Primary.UP == 1) # should be 429

DE.d21._Primary.DOWN  <-as.integer(IDvector.d21%in%(DE_d21.primary.DWNREG$gene.ID)) # Day 0 - Primary - Downregulated DEGs
names(DE.d21._Primary.DOWN)=IDvector.d21
sum(sapply(DE.d21._Primary.DOWN,length)) # should be 8421
sum(DE.d21._Primary.DOWN == 1) # should be 54

DE.d21._Primary.ALL   <-as.integer(IDvector.d21%in%(DE_d21.primary$gene.ID)) # Day 0 - Primary - All DEGs
names(DE.d21._Primary.ALL)=IDvector.d21
sum(sapply(DE.d21._Primary.ALL,length)) # should be 8421
sum(DE.d21._Primary.ALL == 1) # should be 155

#Constant DEGs ------------------------------------ #
# Note: these are degenes that were upregulated and downregulated throughout the 21-day experiment due to the effect of primary treatment
# data from the cluster {r Primary_Treatment_constant_DEGs}
DE_Contant_upreg <-as.integer(GO_unique.genes.all%in%(Contant_upreg$Row.names)) # Day 0 - Primary - All DEGs
names(DE_Contant_upreg)=GO_unique.genes.all

DE_Contant_dwnreg <-as.integer(GO_unique.genes.all%in%(Contant_dwnreg$Row.names)) # Day 0 - Primary - All DEGs
names(DE_Contant_dwnreg)=GO_unique.genes.all
```



Let's review what we have for goseq
```{r NOTE_checklist}
# Review what we have for goseq....

# (1) Go terms ---------------- #
#  head(GO.terms)

# (2) ID vector --------------- #
# head(IDvector.d0)
# head(IDvector.d7)
# head(IDvector.d14)
# head(IDvector.d21)

# (3) Length vector ----------- #
# head(length_vector.d0)
# head(length_vector.d7)
# head(length_vector.d14)
# head(length_vector.d21)

# (4) Binary DEG vectors ------ # 
# head(DE.d21._Primary.UP)    # example of d21 prmary ambient_v_moderate upregulated DEGs
# head(DE.d21._Primary.DOWN)  # example of d21 prmary ambient_v_moderate downregulated DEGs
# head(DE.d21._Primary.ALL)   # example of d21 prmary ambient_v_moderate all DEGs
```



It's goseq time!!!
Calculate Probability Weighting Function
```{r pwf}
#Calculate Probability Weighting Function (using 'nullp')

# (1) DEGs: Primary Ambient_v_Moderate 
# DAY 0 ------------------------------------------- #
DEG.pwf_d0.all    <-nullp(DE.d0._Primary.ALL,   id=IDvector.d0, bias.data=length_vector.d0) #weight vector by length of gene
DEG.pwf_d0.UP     <-nullp(DE.d0._Primary.UP,    id=IDvector.d0, bias.data=length_vector.d0) #weight vector by length of gene
DEG.pwf_d0.DOWN   <-nullp(DE.d0._Primary.DOWN,  id=IDvector.d0, bias.data=length_vector.d0) #weight vector by length of gene

# DAY 7 ------------------------------------------- #
DEG.pwf_d7.all    <-nullp(DE.d7._Primary.ALL,   id=IDvector.d7, bias.data=length_vector.d7) #weight vector by length of gene
DEG.pwf_d7.UP     <-nullp(DE.d7._Primary.UP,    id=IDvector.d7, bias.data=length_vector.d7) #weight vector by length of gene
DEG.pwf_d7.DOWN   <-nullp(DE.d7._Primary.DOWN,  id=IDvector.d7, bias.data=length_vector.d7) #weight vector by length of gene

# DAY 14 ------------------------------------------- #
DEG.pwf_d14.all   <-nullp(DE.d14._Primary.ALL,  id=IDvector.d14, bias.data=length_vector.d14) #weight vector by length of gene
DEG.pwf_d14.UP    <-nullp(DE.d14._Primary.UP,   id=IDvector.d14, bias.data=length_vector.d14) #weight vector by length of gene
DEG.pwf_d14.DOWN  <-nullp(DE.d14._Primary.DOWN, id=IDvector.d14, bias.data=length_vector.d14) #weight vector by length of gene

# DAY 21 ------------------------------------------- #
DEG.pwf_d21.all   <-nullp(DE.d21._Primary.ALL,  id=IDvector.d21, bias.data=length_vector.d21) #weight vector by length of gene
DEG.pwf_d21.UP    <-nullp(DE.d21._Primary.UP,   id=IDvector.d21, bias.data=length_vector.d21) #weight vector by length of gene
DEG.pwf_d21.DOWN  <-nullp(DE.d21._Primary.DOWN, id=IDvector.d21, bias.data=length_vector.d21) #weight vector by length of gene

# Constant DEGs ------------------------------------ #
# DEG.pwf_Constant.UP<-nullp(DE_Contant_upreg, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
# DEG.pwf_Constant.DOWN<-nullp(DE_Contant_dwnreg, GO_unique.genes.all, bias.data=length_vector) #weight vector by length of gene
```



Run goseq
```{r goseq}
# (1) DEGs: Primary Ambient_v_Moderate 
# DAY 0 ------------------------------------------- #
GO_d0.all    <-goseq(DEG.pwf_d0.all,  IDvector.d0, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d0.UP     <-goseq(DEG.pwf_d0.UP,   IDvector.d0, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d0.DOWN   <-goseq(DEG.pwf_d0.DOWN, IDvector.d0, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# DAY 7 ------------------------------------------- #
GO_d7.all    <-goseq(DEG.pwf_d7.all,  IDvector.d7, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d7.UP     <-goseq(DEG.pwf_d7.UP,   IDvector.d7, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d7.DOWN   <-goseq(DEG.pwf_d7.DOWN, IDvector.d7, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# DAY 14 ------------------------------------------- #
GO_d14.all   <-goseq(DEG.pwf_d14.all,  IDvector.d14, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d14.UP    <-goseq(DEG.pwf_d14.UP,   IDvector.d14, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d14.DOWN  <-goseq(DEG.pwf_d14.DOWN, IDvector.d14, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# DAY 21 ------------------------------------------- #
GO_d21.all   <-goseq(DEG.pwf_d21.all,  IDvector.d21, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d21.UP    <-goseq(DEG.pwf_d21.UP,   IDvector.d21, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
GO_d21.DOWN  <-goseq(DEG.pwf_d21.DOWN, IDvector.d21, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# Constant DEGs ------------------------------------ #
# GO_Constant.UP<-goseq(DEG.pwf_Constant.UP, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
# GO_Constant.DOWN<-goseq(DEG.pwf_Constant.DOWN, ID.vector, gene2cat=GO.terms, test.cats=c("GO:CC", "GO:BP", "GO:MF"), method="Wallenius", use_genes_without_cat=TRUE)
```



Find only enriched GO terms that are statistically significant at cutoff
```{r}
# Day0 Primary Treatment effect
# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
d0.UP.05<-GO_d0.UP$category[GO_d0.UP$over_represented_pvalue<.05]
d0.UP.05<-data.frame(d0.UP.05)
colnames(d0.UP.05) <- c("category")
d0.UP.05 <- merge(d0.UP.05, GO_d0.UP, by="category")
d0.UP.05 <- d0.UP.05[order(d0.UP.05$ontology, d0.UP.05$over_represented_pvalue, -d0.UP.05$numDEInCat),]
d0.UP.05$term <- as.factor(d0.UP.05$term)
nrow(filter(d0.UP.05, ontology=="BP")) #number sig BP terms == 7
nrow(filter(d0.UP.05, ontology=="MF")) #number sig MF terms == 5
nrow(d0.UP.05) # == 12
#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
d0.DOWN.05<-GO_d0.DOWN$category[GO_d0.DOWN$over_represented_pvalue<.05]
d0.DOWN.05<-data.frame(d0.DOWN.05)
colnames(d0.DOWN.05) <- c("category")
d0.DOWN.05 <- merge(d0.DOWN.05, GO_d0.DOWN, by="category")
d0.DOWN.05 <- d0.DOWN.05[order(d0.DOWN.05$ontology, d0.DOWN.05$over_represented_pvalue, -d0.DOWN.05$numDEInCat),]
d0.DOWN.05$term <- as.factor(d0.DOWN.05$term)
nrow(filter(d0.DOWN.05, ontology=="BP")) #number sig BP terms == 7
nrow(filter(d0.DOWN.05, ontology=="MF")) #number sig MF terms == 5
nrow(d0.DOWN.05) # == 12


# Day7 Primary Treatment effect
# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
d7.UP.05<-GO_d7.UP$category[GO_d7.UP$over_represented_pvalue<.05]
d7.UP.05<-data.frame(d7.UP.05)
colnames(d7.UP.05) <- c("category")
d7.UP.05 <- merge(d7.UP.05, GO_d7.UP, by="category")
d7.UP.05 <- d7.UP.05[order(d7.UP.05$ontology, d7.UP.05$over_represented_pvalue, -d7.UP.05$numDEInCat),]
d7.UP.05$term <- as.factor(d7.UP.05$term)
nrow(filter(d7.UP.05, ontology=="BP")) #number sig BP terms == 85
nrow(filter(d7.UP.05, ontology=="MF")) #number sig MF terms == 28
nrow(d7.UP.05) # == 128
#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
d7.DOWN.05<-GO_d7.DOWN$category[GO_d7.DOWN$over_represented_pvalue<.05]
d7.DOWN.05<-data.frame(d7.DOWN.05)
colnames(d7.DOWN.05) <- c("category")
d7.DOWN.05 <- merge(d7.DOWN.05, GO_d7.DOWN, by="category")
d7.DOWN.05 <- d7.DOWN.05[order(d7.DOWN.05$ontology, d7.DOWN.05$over_represented_pvalue, -d7.DOWN.05$numDEInCat),]
d7.DOWN.05$term <- as.factor(d7.DOWN.05$term)
nrow(filter(d7.DOWN.05, ontology=="BP")) #number sig BP terms == 40
nrow(filter(d7.DOWN.05, ontology=="MF")) #number sig MF terms == 15
nrow(d7.DOWN.05) # == 60

# Day14 Primary Treatment effect
# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
d14.UP.05<-GO_d14.UP$category[GO_d14.UP$over_represented_pvalue<.05]
d14.UP.05<-data.frame(d14.UP.05)
colnames(d14.UP.05) <- c("category")
d14.UP.05 <- merge(d14.UP.05, GO_d14.UP, by="category")
d14.UP.05 <- d14.UP.05[order(d14.UP.05$ontology, d14.UP.05$over_represented_pvalue, -d14.UP.05$numDEInCat),]
d14.UP.05$term <- as.factor(d14.UP.05$term)
nrow(filter(d14.UP.05, ontology=="BP")) #number sig BP terms == 119
nrow(filter(d14.UP.05, ontology=="MF")) #number sig MF terms == 62
nrow(d14.UP.05) # == 208
#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
d14.DOWN.05<-GO_d14.DOWN$category[GO_d14.DOWN$over_represented_pvalue<.05]
d14.DOWN.05<-data.frame(d14.DOWN.05)
colnames(d14.DOWN.05) <- c("category")
d14.DOWN.05 <- merge(d14.DOWN.05, GO_d14.DOWN, by="category")
d14.DOWN.05 <- d14.DOWN.05[order(d14.DOWN.05$ontology, d14.DOWN.05$over_represented_pvalue, -d14.DOWN.05$numDEInCat),]
d14.DOWN.05$term <- as.factor(d14.DOWN.05$term)
nrow(filter(d14.DOWN.05, ontology=="BP")) #number sig BP terms == 146
nrow(filter(d14.DOWN.05, ontology=="MF")) #number sig MF terms == 53
nrow(d14.DOWN.05) # == 228

# Day21 Primary Treatment effect
# UPREGULATED GENES - those with higher expression from ambient (control) stress priming
d21.UP.05<-GO_d21.UP$category[GO_d21.UP$over_represented_pvalue<.05]
d21.UP.05<-data.frame(d21.UP.05)
colnames(d21.UP.05) <- c("category")
d21.UP.05 <- merge(d21.UP.05, GO_d21.UP, by="category")
d21.UP.05 <- d21.UP.05[order(d21.UP.05$ontology, d21.UP.05$over_represented_pvalue, -d21.UP.05$numDEInCat),]
d21.UP.05$term <- as.factor(d21.UP.05$term)
nrow(filter(d21.UP.05, ontology=="BP")) #number sig BP terms == 65
nrow(filter(d21.UP.05, ontology=="MF")) #number sig MF terms == 37
nrow(d21.UP.05) # == 116
#DOWNREGULATED GENES  - those with higher expression from moderate stress priming
d21.DOWN.05<-GO_d21.DOWN$category[GO_d21.DOWN$over_represented_pvalue<.05]
d21.DOWN.05<-data.frame(d21.DOWN.05)
colnames(d21.DOWN.05) <- c("category")
d21.DOWN.05 <- merge(d21.DOWN.05, GO_d21.DOWN, by="category")
d21.DOWN.05 <- d21.DOWN.05[order(d21.DOWN.05$ontology, d21.DOWN.05$over_represented_pvalue, -d21.DOWN.05$numDEInCat),]
d21.DOWN.05$term <- as.factor(d21.DOWN.05$term)
nrow(filter(d21.DOWN.05, ontology=="BP")) #number sig BP terms == 39
nrow(filter(d21.DOWN.05, ontology=="MF")) #number sig MF terms == 16
nrow(d21.DOWN.05) # == 67
```



Correct any un-annotated terms/ontologies
```{r}
NAs.ontology <- d0.UP.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- d0.DOWN.05 %>% subset(is.na(term))
print(NAs.ontology)

NAs.ontology <- d7.UP.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- d7.DOWN.05 %>% subset(is.na(term))
print(NAs.ontology)

NAs.ontology <- d14.UP.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- d14.DOWN.05 %>% subset(is.na(term))
print(NAs.ontology)

NAs.ontology <- d21.UP.05 %>% subset(is.na(term))
print(NAs.ontology)
NAs.ontology <- d21.DOWN.05 %>% subset(is.na(term))
print(NAs.ontology)

# NO CASES OF NA
```



Save significant terms
```{r, warning=FALSE}
write.csv(d0.UP.05, file    = "../Output/GO/DESeq2_goseq/Day0/GO.05.Day0_PrimEffect_Upregulated.csv", row.names = FALSE)
write.csv(d0.DOWN.05, file  = "../Output/GO/DESeq2_goseq/Day0/GO.05.Day0_PrimEffect_Downregulated.csv", row.names = FALSE)

write.csv(d7.UP.05, file    = "../Output/GO/DESeq2_goseq/Day7/GO.05.Day7_PrimEffect_Upregulated.csv", row.names = FALSE)
write.csv(d7.DOWN.05, file  = "../Output/GO/DESeq2_goseq/Day7/GO.05.Day7_PrimEffect_Downregulated.csv", row.names = FALSE)

write.csv(d14.UP.05, file   = "../Output/GO/DESeq2_goseq/Day14/GO.05.Day14_PrimEffect_Upregulated.csv", row.names = FALSE)
write.csv(d14.DOWN.05, file = "../Output/GO/DESeq2_goseq/Day14/GO.05.Day14_PrimEffect_Downregulated.csv", row.names = FALSE)

write.csv(d21.UP.05, file   = "../Output/GO/DESeq2_goseq/Day21/GO.05.Day21_PrimEffect_Upregulated.csv", row.names = FALSE)
write.csv(d21.DOWN.05, file = "../Output/GO/DESeq2_goseq/Day21/GO.05.Day21_PrimEffect_Downregulated.csv", row.names = FALSE)
```



GOslim analysis - data reduction from the many GO terms to more broad function/processes
Read in files if previous steps not run.
```{r}

d0.UP.05    <- read.csv("../Output/GO/goseq/Day0/GO.05.Day0_PrimEffect_Upregulated.csv")
d0.DOWN.05  <- read.csv("../Output/GO/goseq/Day0/GO.05.Day0_PrimEffect_Downregulated.csv")

d7.UP.05    <- read.csv("../Output/GO/goseq/Day7/GO.05.Day0_PrimEffect_Upregulated.csv")
d7.DOWN.05  <- read.csv("../Output/GO/goseq/Day7/GO.05.Day0_PrimEffect_Downregulated.csv")

d14.UP.05   <- read.csv("../Output/GO/goseq/Day14/GO.05.Day0_PrimEffect_Upregulated.csv")
d14.DOWN.05 <- read.csv("../Output/GO/goseq/Day14/GO.05.Day0_PrimEffect_Downregulated.csv")

d21.UP.05   <- read.csv("../Output/GO/goseq/Day21/GO.05.Day0_PrimEffect_Upregulated.csv")
d21.DOWN.05 <- read.csv("../Output/GO/goseq/Day21/GO.05.Day0_PrimEffect_Downregulated.csv")


d0.UP.05$dir    <- "Day0_Up"
d7.UP.05$dir    <- "Day7_Up"
d14.UP.05$dir   <- "Day14_Up"
d21.UP.05$dir   <- "Day21_Up"

d0.DOWN.05$dir  <- "Day0_Down"
d7.DOWN.05$dir  <- "Day7_Down"
d14.DOWN.05$dir <- "Day14_Down"
d21.DOWN.05$dir <- "Day21_Down"

all.UP.05<- rbind (d0.UP.05,
                 d7.UP.05,
                 d14.UP.05,
                 d21.UP.05)
all.DOWN.05<- rbind (d0.DOWN.05,
                 d7.DOWN.05,
                 d14.DOWN.05,
                 d21.DOWN.05)
```



Run GOslim to get broader categories
```{r}
slim <- getOBOCollection("http://current.geneontology.org/ontology/subsets/goslim_generic.obo") #get GO database

# DAY 0 treatment efect GO slim ===================================================================== #
# UPREGULATED
d0.UP_BP <- d0.UP.05 %>% # BP - all GO terms upregulated
  filter(ontology=="BP")
d0.UP_BP_GO_collection <- GOCollection(d0.UP_BP$category) #Make library of query terms
d0.UP_slims_bp <- data.frame(goSlim(d0.UP_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
d0.UP_slims_bp$category <- row.names(d0.UP_slims_bp) #save rownames as category
d0.UP_MF <- d0.UP.05 %>% # MF - all GO terms upregulated
  filter(ontology=="MF")
d0.UP_MF_GO_collection <- GOCollection(d0.UP_MF$category) #Make library of query terms
d0.UP_slims_mf <- data.frame(goSlim(d0.UP_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
d0.UP_slims_mf$category <- row.names(d0.UP_slims_mf) #save rownames as category
# DOWNREGULATED
d0.DOWN_BP <- d0.DOWN.05 %>% # BP - all GO terms DOWNREGULATED
  filter(ontology=="BP")
d0.DOWN_BP_GO_collection <- GOCollection(d0.DOWN_BP$category) #Make library of query terms
d0.DOWN_slims_bp <- data.frame(goSlim(d0.DOWN_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
d0.DOWN_slims_bp$category <- row.names(d0.DOWN_slims_bp) #save rownames as category
d0.DOWN_MF <- d0.DOWN.05 %>% # MF - all GO terms DOWNREGULATED
  filter(ontology=="MF")
d0.DOWN_MF_GO_collection <- GOCollection(d0.DOWN_MF$category) #Make library of query terms
d0.DOWN_slims_mf <- data.frame(goSlim(d0.DOWN_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
d0.DOWN_slims_mf$category <- row.names(d0.DOWN_slims_mf) #save rownames as category

# DAY 7 treatment efect GO slim ===================================================================== #
# UPREGULATED
d7.UP_BP <- d7.UP.05 %>% # BP - all GO terms upregulated
  filter(ontology=="BP")
d7.UP_BP_GO_collection <- GOCollection(d7.UP_BP$category) #Make library of query terms
d7.UP_slims_bp <- data.frame(goSlim(d7.UP_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
d7.UP_slims_bp$category <- row.names(d7.UP_slims_bp) #save rownames as category
d7.UP_MF <- d7.UP.05 %>% # MF - all GO terms upregulated
  filter(ontology=="MF")
d7.UP_MF_GO_collection <- GOCollection(d7.UP_MF$category) #Make library of query terms
d7.UP_slims_mf <- data.frame(goSlim(d7.UP_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
d7.UP_slims_mf$category <- row.names(d7.UP_slims_mf) #save rownames as category
# DOWNREGULATED
d7.DOWN_BP <- d7.DOWN.05 %>% # BP - all GO terms DOWNREGULATED
  filter(ontology=="BP")
d7.DOWN_BP_GO_collection <- GOCollection(d7.DOWN_BP$category) #Make library of query terms
d7.DOWN_slims_bp <- data.frame(goSlim(d7.DOWN_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
d7.DOWN_slims_bp$category <- row.names(d7.DOWN_slims_bp) #save rownames as category
d7.DOWN_MF <- d7.DOWN.05 %>% # MF - all GO terms DOWNREGULATED
  filter(ontology=="MF")
d7.DOWN_MF_GO_collection <- GOCollection(d7.DOWN_MF$category) #Make library of query terms
d7.DOWN_slims_mf <- data.frame(goSlim(d7.DOWN_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
d7.DOWN_slims_mf$category <- row.names(d7.DOWN_slims_mf) #save rownames as category

# DAY 14 treatment efect GO slim ===================================================================== #
# UPREGULATED
d14.UP_BP <- d14.UP.05 %>% # BP - all GO terms upregulated
  filter(ontology=="BP")
d14.UP_BP_GO_collection <- GOCollection(d14.UP_BP$category) #Make library of query terms
d14.UP_slims_bp <- data.frame(goSlim(d14.UP_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
d14.UP_slims_bp$category <- row.names(d14.UP_slims_bp) #save rownames as category
d14.UP_MF <- d14.UP.05 %>% # MF - all GO terms upregulated
  filter(ontology=="MF")
d14.UP_MF_GO_collection <- GOCollection(d14.UP_MF$category) #Make library of query terms
d14.UP_slims_mf <- data.frame(goSlim(d14.UP_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
d14.UP_slims_mf$category <- row.names(d14.UP_slims_mf) #save rownames as category
# DOWNREGULATED
d14.DOWN_BP <- d14.DOWN.05 %>% # BP - all GO terms DOWNREGULATED
  filter(ontology=="BP")
d14.DOWN_BP_GO_collection <- GOCollection(d14.DOWN_BP$category) #Make library of query terms
d14.DOWN_slims_bp <- data.frame(goSlim(d14.DOWN_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
d14.DOWN_slims_bp$category <- row.names(d14.DOWN_slims_bp) #save rownames as category
d14.DOWN_MF <- d14.DOWN.05 %>% # MF - all GO terms DOWNREGULATED
  filter(ontology=="MF")
d14.DOWN_MF_GO_collection <- GOCollection(d14.DOWN_MF$category) #Make library of query terms
d14.DOWN_slims_mf <- data.frame(goSlim(d14.DOWN_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
d14.DOWN_slims_mf$category <- row.names(d14.DOWN_slims_mf) #save rownames as category

# DAY 21 treatment efect GO slim ===================================================================== #
# UPREGULATED
d21.UP_BP <- d21.UP.05 %>% # BP - all GO terms upregulated
  filter(ontology=="BP")
d21.UP_BP_GO_collection <- GOCollection(d21.UP_BP$category) #Make library of query terms
d21.UP_slims_bp <- data.frame(goSlim(d21.UP_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
d21.UP_slims_bp$category <- row.names(d21.UP_slims_bp) #save rownames as category
d21.UP_MF <- d21.UP.05 %>% # MF - all GO terms upregulated
  filter(ontology=="MF")
d21.UP_MF_GO_collection <- GOCollection(d21.UP_MF$category) #Make library of query terms
d21.UP_slims_mf <- data.frame(goSlim(d21.UP_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
d21.UP_slims_mf$category <- row.names(d21.UP_slims_mf) #save rownames as category
# DOWNREGULATED
d21.DOWN_BP <- d21.DOWN.05 %>% # BP - all GO terms DOWNREGULATED
  filter(ontology=="BP")
d21.DOWN_BP_GO_collection <- GOCollection(d21.DOWN_BP$category) #Make library of query terms
d21.DOWN_slims_bp <- data.frame(goSlim(d21.DOWN_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
d21.DOWN_slims_bp$category <- row.names(d21.DOWN_slims_bp) #save rownames as category
d21.DOWN_MF <- d21.DOWN.05 %>% # MF - all GO terms DOWNREGULATED
  filter(ontology=="MF")
d21.DOWN_MF_GO_collection <- GOCollection(d21.DOWN_MF$category) #Make library of query terms
d21.DOWN_slims_mf <- data.frame(goSlim(d21.DOWN_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
d21.DOWN_slims_mf$category <- row.names(d21.DOWN_slims_mf) #save rownames as category


# all primary treatment efect GO slim  ===================================================================== #
# UPREGULATED
All.UP_BP <- all.UP.05 %>% # BP - all GO terms upregulated
  filter(ontology=="BP")
All.UP_BP_GO_collection <- GOCollection(All.UP_BP$category) #Make library of query terms
All.UP_slims_bp <- data.frame(goSlim(All.UP_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
All.UP_slims_bp$category <- row.names(All.UP_slims_bp) #save rownames as category
All.UP_MF <- all.UP.05 %>% # MF - all GO terms upregulated
  filter(ontology=="MF")
All.UP_MF_GO_collection <- GOCollection(All.UP_MF$category) #Make library of query terms
All.UP_slims_mf <- data.frame(goSlim(All.UP_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
All.UP_slims_mf$category <- row.names(All.UP_slims_mf) #save rownames as category
# DOWNREGULATED
All.DOWN_BP <- all.DOWN.05 %>% # BP - all GO terms downregulated
  filter(ontology=="BP")
All.DOWN_BP_GO_collection <- GOCollection(All.DOWN_BP$category) #Make library of query terms
All.DOWN_slims_bp <- data.frame(goSlim(All.DOWN_BP_GO_collection, slim, "BP")) #Find common parent terms to slim down our list
All.DOWN_slims_bp$category <- row.names(All.DOWN_slims_bp) #save rownames as category
All.DOWN_MF <- all.DOWN.05 %>% # MF - all GO terms downregulated
  filter(ontology=="MF")
All.DOWN_MF_GO_collection <- GOCollection(All.DOWN_MF$category) #Make library of query terms
All.DOWN_slims_mf <- data.frame(goSlim(All.DOWN_MF_GO_collection, slim, "MF")) #Find common parent terms to slim down our list
All.DOWN_slims_mf$category <- row.names(All.DOWN_slims_mf) #save rownames as category

# View(All.UP_slims_bp)
# View(All.UP_slims_mf)
# View(All.DOWN_slims_bp)
# View(All.DOWN_slims_mf)
```



Get mapped terms, using functions from Sam White's Biostars [post](https://support.bioconductor.org/p/128407/#128409).
```{r}
#Write function mappedIds to get the query terms that mapped to the slim categories - in other words, add acolumn to your slim dataframe with all the GO terms from goseq
mappedIds <-
  function(df, collection, OFFSPRING) #the command to run requires a dataframe of slim terms, like slims_MF above, your list of query terms, and the offspring from the GOCollection by goSlim
  {
    map <- as.list(OFFSPRING[rownames(df)]) # Subset GOcollection offspring by the rownames of your dataframe
    mapped <- lapply(map, intersect, ids(collection)) #Find the terms that intersect between the subset made above of your query terms and the GOids from the GO collection
    df[["go_terms"]] <- vapply(unname(mapped), paste, collapse = ";", character(1L)) #Add column "go_terms" with matching terms 
    df #show resulting dataframe
  }
#Run function for MF and BP terms
# DAY 0
d0.UP_BPslim <- mappedIds(d0.UP_slims_bp, d0.UP_BP_GO_collection, GOBPOFFSPRING)
d0.UP_MFslim <- mappedIds(d0.UP_slims_mf, d0.UP_MF_GO_collection, GOMFOFFSPRING)

d0.DOWN_BPslim <- mappedIds(d0.DOWN_slims_bp, d0.DOWN_BP_GO_collection, GOBPOFFSPRING)
d0.DOWN_MFslim <- mappedIds(d0.DOWN_slims_mf, d0.DOWN_MF_GO_collection, GOMFOFFSPRING)

# DAY 7
d7.UP_BPslim <- mappedIds(d7.UP_slims_bp, d7.UP_BP_GO_collection, GOBPOFFSPRING)
d7.UP_MFslim <- mappedIds(d7.UP_slims_mf, d7.UP_MF_GO_collection, GOMFOFFSPRING)

d7.DOWN_BPslim <- mappedIds(d7.DOWN_slims_bp, d7.DOWN_BP_GO_collection, GOBPOFFSPRING)
d7.DOWN_MFslim <- mappedIds(d7.DOWN_slims_mf, d7.DOWN_MF_GO_collection, GOMFOFFSPRING)

# DAY 14
d14.UP_BPslim <- mappedIds(d14.UP_slims_bp, d14.UP_BP_GO_collection, GOBPOFFSPRING)
d14.UP_MFslim <- mappedIds(d14.UP_slims_mf, d14.UP_MF_GO_collection, GOMFOFFSPRING)

d14.DOWN_BPslim <- mappedIds(d14.DOWN_slims_bp, d14.DOWN_BP_GO_collection, GOBPOFFSPRING)
d14.DOWN_MFslim <- mappedIds(d14.DOWN_slims_mf, d14.DOWN_MF_GO_collection, GOMFOFFSPRING)

# DAY 21
d21.UP_BPslim <- mappedIds(d21.UP_slims_bp, d21.UP_BP_GO_collection, GOBPOFFSPRING)
d21.UP_MFslim <- mappedIds(d21.UP_slims_mf, d21.UP_MF_GO_collection, GOMFOFFSPRING)

d21.DOWN_BPslim <- mappedIds(d21.DOWN_slims_bp, d21.DOWN_BP_GO_collection, GOBPOFFSPRING)
d21.DOWN_MFslim <- mappedIds(d21.DOWN_slims_mf, d21.DOWN_MF_GO_collection, GOMFOFFSPRING)

# all
All.UP_BPslim <- mappedIds(All.UP_slims_bp, All.UP_BP_GO_collection, GOBPOFFSPRING)
All.UP_MFslim <- mappedIds(All.UP_slims_mf, All.UP_MF_GO_collection, GOMFOFFSPRING)

All.DOWN_BPslim <- mappedIds(All.DOWN_slims_bp, All.DOWN_BP_GO_collection, GOBPOFFSPRING)
All.DOWN_MFslim <- mappedIds(All.DOWN_slims_mf, All.DOWN_MF_GO_collection, GOMFOFFSPRING)
```



Build final GO slim data set for plotting
- call the annotation file - build a master sheet of unique rows for genes and GO terms (multiple gene IDs for each GO term annotated)
- use this datasetto filter by upreg or down reg DE genes (DESEq2 directionality) - then for loop into the slim data to call unique genes in each GOslim bin
- output column 'Gene.Count' to the final GOslim data table (i.e. All.UP_BPslim_final == all DESeq2 primary effect genes table, upregualted DEGs only, Biological Process GOslim/GOterms only)
```{r}

# Build a master list of all genes and GO terms
Geoduck_annotation <- read.delim2(file="C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Seq_details/Panopea-generosa-genes-annotations.txt", header=F) # Load themaster Pgenerosa gene list 
Pgen_GOterms <- Geoduck_annotation %>% dplyr::select(c('V1','V8')) # select only two columns - those with the gene IDs and those with the GO terms
Pgen_GOterms2 <- strsplit(Pgen_GOterms$V8, split = "; ") # create a string splitting by delimiter '; ' - view the data to see that this separates each GO term entry in the string
Pgen_GOterms2 <- data.frame(gene.ID = rep(Pgen_GOterms$V1, sapply(Pgen_GOterms2, length)), Go.terms = unlist(Pgen_GOterms2)) # create new dataframe 'Pgen_GOterms2' listing genes for each GO term (MUCH longer!)
Pgen_GOterms2 <- na.omit(Pgen_GOterms2) # ommit the NAs  - genes without GO annotation

Pgen_GOterms_ALL_UPREG <- Pgen_GOterms2 %>%  dplyr::filter(gene.ID %in% All.primary.UPREG$Row.names) # 'All.primary.UPREG' lists all upregualted genes Days 0, 7, 14, 21 in response to primary treatment
Pgen_GOterms_ALL_DOWNREG <- Pgen_GOterms2 %>%  dplyr::filter(gene.ID %in% All.primary.DWNREG$Row.names) # 'All.primary.DWNREG' lists all upregualted genes Days 0, 7, 14, 21 in response to primary treatment

# All DEGs Primary Effect ========================== #
#UPREGULATED
#BP
All.UP_BPslim <- filter(All.UP_BPslim, Count>5 & Term!="biological_process") #filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(All.UP_BPslim$go_terms), ";") #split into multiple GO ids
All.UP_BPslim$BPsplitted <- BPsplitted
for (i in 1:nrow(All.UP_BPslim)) {
  table <- data.frame(GOlist = unlist(All.UP_BPslim[i,6])) # call the BPsplitted column of characters and create a small table to filter
  table <- unique(table)
  Pgen_loop <- Pgen_GOterms_ALL_UPREG %>% dplyr::filter(Go.terms %in% table$GOlist) # filter Gene IDs with the GO term
  Pgen_geneIDs <- Pgen_loop[-2] # ommit the GO.terms to call unique gene calls 
  Pgen_geneIDs <- unique(Pgen_geneIDs) # call unique Gene calls (unique genes that had the GO term within each of the GOslim bins)
  All.UP_BPslim$Gene.Count[i] <- nrow(Pgen_geneIDs)  # count of unique GeneIDs in each GOslim bin
  All.UP_BPslim$Gene.IDs[[i]] <- vapply((Pgen_geneIDs$gene.ID), paste, collapse = ";", character(1L))} # name of each unique gene.id in each GOslim bin
All.UP_BPslim_A <- data.frame(Term = rep.int(All.UP_BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
All.UP_BPslim_B <- merge(All.UP_BPslim_A, All.UP_BPslim, by="Term") #Add back counts, term, and category info
All.UP_BPslim_C <- unique(setDT(All.UP_BPslim_B)[order(go_term, -Gene.Count)], by = "category") #remove duplicate offspring terms, keeping only those in the larger umbrella term (Count number)
All.UP_BPslim_final <- data.frame(slim_term=All.UP_BPslim_C$Term, slim_cat=All.UP_BPslim_C$category, category=All.UP_BPslim_C$go_term, Gene.Count=All.UP_BPslim_C$Gene.Count, GO.Count=All.UP_BPslim_C$Count) #rename columns) #rename columns
#MF
All.UP_MFslim <- filter(All.UP_MFslim, Count>2 & Term!="molecular_function") #filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(All.UP_MFslim$go_terms), ";") #split into multiple GO ids
All.UP_MFslim$MFsplitted <- MFsplitted
for (i in 1:nrow(All.UP_MFslim)) {
  table <- data.frame(GOlist = unlist(All.UP_MFslim[i,6])) # call the BPsplitted column of characters and create a small table to filter
  table <- unique(table)
  Pgen_loop <- Pgen_GOterms_ALL_UPREG %>% dplyr::filter(Go.terms %in% table$GOlist) # filter Gene IDs with the GO term
  Pgen_geneIDs <- Pgen_loop[-2] # ommit the GO.terms to call unique gene calls 
  Pgen_geneIDs <- unique(Pgen_geneIDs) # call unique Gene calls (unique genes that had the GO term within each of the GOslim bins)
  All.UP_MFslim$Gene.Count[i] <- nrow(Pgen_geneIDs)  # count of unique GeneIDs in each GOslim bin
  All.UP_MFslim$Gene.IDs[[i]] <- vapply((Pgen_geneIDs$gene.ID), paste, collapse = ";", character(1L))} # name of each unique gene.id in each GOslim bin
All.UP_MFslim_A <- data.frame(Term = rep.int(All.UP_MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all - add total count for each term in 'Count'
All.UP_MFslim_B <- merge(All.UP_MFslim_A, All.UP_MFslim, by="Term")  #Add back counts, term, and category info
All.UP_MFslim_C <- unique(setDT(All.UP_MFslim_B)[order(go_term, Gene.Count)], by = "category")  #remove duplicate offspring terms, keeping only those in the larger umbrella term (Count number)
All.UP_MFslim_final <- data.frame(slim_term=All.UP_MFslim_C$Term, slim_cat=All.UP_MFslim_C$category, category=All.UP_MFslim_C$go_term, Gene.Count=All.UP_MFslim_C$Gene.Count, GO.Count=All.UP_MFslim_C$Count) #rename columns

# DOWNREGULATED
#BP
All.DOWN_BPslim <- filter(All.DOWN_BPslim, Count>5 & Term!="biological_process") #filter out empty slims and term "biological process"
BPsplitted <- strsplit(as.character(All.DOWN_BPslim$go_terms), ";") #split into multiple GO ids
All.DOWN_BPslim$BPsplitted <- BPsplitted
for (i in 1:nrow(All.DOWN_BPslim)) {
  table <- data.frame(GOlist = unlist(All.DOWN_BPslim[i,6])) # call the BPsplitted column of characters and create a small table to filter
  table <- unique(table)
  Pgen_loop <- Pgen_GOterms_ALL_DOWNREG %>% dplyr::filter(Go.terms %in% table$GOlist) # filter Gene IDs with the GO term
  Pgen_geneIDs <- Pgen_loop[-2] # ommit the GO.terms to call unique gene calls 
  Pgen_geneIDs <- unique(Pgen_geneIDs) # call unique Gene calls (unique genes that had the GO term within each of the GOslim bins)
  All.DOWN_BPslim$Gene.Count[i] <- nrow(Pgen_geneIDs)  # count of unique GeneIDs in each GOslim bin
  All.DOWN_BPslim$Gene.IDs[[i]] <- vapply((Pgen_geneIDs$gene.ID), paste, collapse = ";", character(1L))} # name of each unique gene.id in each GOslim bin
All.DOWN_BPslim_A <- data.frame(Term = rep.int(All.DOWN_BPslim$Term, sapply(BPsplitted, length)), go_term = unlist(BPsplitted)) #list all
All.DOWN_BPslim_B <- merge(All.DOWN_BPslim_A, All.DOWN_BPslim, by="Term") #Add back counts, term, and category info
All.DOWN_BPslim_C <- unique(setDT(All.DOWN_BPslim_B)[order(go_term, -Gene.Count)], by = "category") #remove dDOWNlicate offspring terms, keeping only those in the larger umbrella term (Count number)
All.DOWN_BPslim_final <- data.frame(slim_term=All.DOWN_BPslim_C$Term, slim_cat=All.DOWN_BPslim_C$category, category=All.DOWN_BPslim_C$go_term, Gene.Count=All.DOWN_BPslim_C$Gene.Count, GO.Count=All.DOWN_BPslim_C$Count) #rename columns
#MF
All.DOWN_MFslim <- filter(All.DOWN_MFslim, Count>2 & Term!="molecular_function") #filter out empty slims and term "molecular function"
MFsplitted <- strsplit(as.character(All.DOWN_MFslim$go_terms), ";") #split into multiple GO ids
All.DOWN_MFslim$MFsplitted <- MFsplitted
for (i in 1:nrow(All.DOWN_MFslim)) {
  table <- data.frame(GOlist = unlist(All.DOWN_MFslim[i,6])) # call the BPsplitted column of characters and create a small table to filter
  table <- unique(table)
  Pgen_loop <- Pgen_GOterms_ALL_DOWNREG %>% dplyr::filter(Go.terms %in% table$GOlist) # filter Gene IDs with the GO term
  Pgen_geneIDs <- Pgen_loop[-2] # ommit the GO.terms to call unique gene calls 
  Pgen_geneIDs <- unique(Pgen_geneIDs) # call unique Gene calls (unique genes that had the GO term within each of the GOslim bins)
  All.DOWN_MFslim$Gene.Count[i] <- nrow(Pgen_geneIDs)  # count of unique GeneIDs in each GOslim bin
  All.DOWN_MFslim$Gene.IDs[[i]] <- vapply((Pgen_geneIDs$gene.ID), paste, collapse = ";", character(1L))} # name of each unique gene.id in each GOslim bin
All.DOWN_MFslim_A <- data.frame(Term = rep.int(All.DOWN_MFslim$Term, sapply(MFsplitted, length)), go_term = unlist(MFsplitted)) #list all - add total count for each term in 'Count'
All.DOWN_MFslim_B <- merge(All.DOWN_MFslim_A, All.DOWN_MFslim, by="Term")  #Add back counts, term, and category info
All.DOWN_MFslim_C <- unique(setDT(All.DOWN_MFslim_B)[order(go_term, -Gene.Count)], by = "category")  #remove dDOWNlicate offspring terms, keeping only those in the larger umbrella term (Count number)
All.DOWN_MFslim_final <- data.frame(slim_term=All.DOWN_MFslim_C$Term, slim_cat=All.DOWN_MFslim_C$category, category=All.DOWN_MFslim_C$go_term, Gene.Count=All.DOWN_MFslim_C$Gene.Count, GO.Count=All.DOWN_MFslim_C$Count) #rename columns

```



Stacked bar plots -  Gene Counts within GOslim term (NOTE: there are likely duplicates (or more) BETWEEN the GOslim bins - however EACH bin contains UNIQUE genes)
```{r barplots} 
# create color palette  --------------------------------------------------------------------------------------- #
library(RColorBrewer)
cbp1 <- c("#999999", "#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7") # color blind pallette
OrangeRed <- brewer.pal(2, "OrRd") 
GreenBlue <- brewer.pal(2, "GnBu") 

# add ontology column  ---------------------------------------------------------------------------------------- #
All.UP_BPslim_final$Ontolgy <- "BP_Upregulated_all"
All.UP_MFslim_final$Ontolgy <- "MF_Upregulated_all"

All.DOWN_BPslim_final$Ontolgy <- "BP_Downregulated_all"
All.DOWN_MFslim_final$Ontolgy <- "MF_Downregulated_all"

# BIND DATA TO FACET THE PLOTS BY UP AND DOWN REG   ---------------------------------------------------------- #
# bp
BP_All <- rbind(All.UP_BPslim_final, All.DOWN_BPslim_final) # merge the data by binding rows 
BP_All$Ont <- "BP" # create a new common clumn to call in the plot 
BP_All$Ontolgy <- factor(BP_All$Ontolgy, levels = c("BP_Upregulated_all", "BP_Downregulated_all"))# reorder the facotr level for the facet wrap plot 
BP_All_filtered <- BP_All %>%  dplyr::filter(Gene.Count > 1) # ommit all with gene counts >1
# mf
MF_All <- rbind(All.UP_MFslim_final, All.DOWN_MFslim_final) # merge the data by binding rows 
MF_All$Ont <- "MF" # create a new common clumn to call in the plot 
MF_All$Ontolgy <- factor(MF_All$Ontolgy, levels = c("MF_Upregulated_all", "MF_Downregulated_all")) # reorder the facotr level for the facet wrap plot 
MF_All_filtered <- MF_All %>%  dplyr::filter(Gene.Count > 1) # ommit all with gene counts >1
# PLOTTING --------------------------------------------------------------------------------------------------- #
BP_Plot_PrimEffect_All <- ggplot(data = BP_All_filtered, aes(x = Ont, y = Gene.Count, fill=forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE))) + # BP plot
                               geom_bar(color = "white",size=1.5,stat="identity") + # call bar and create lines between each slim term
                               scale_fill_manual(values =  rev( colorRampPalette(OrangeRed)( (nrow(BP_All)) ) ) )+ # fill with pallette accomodating the max nmber of slim terms
                               geom_text(aes(label = forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE)), colour = "black", position="stack", size = 7, vjust = 1.25) + # add labels
                               theme_classic() + # classic theme
                               ylim(0, 300) + 
                               theme(legend.position = "none",text = element_text(size=25)) +
                               facet_wrap(~Ontolgy) # facet by the upregulated and downregulated datasets

MF_Plot_PrimEffect_All <- ggplot(data = MF_All_filtered, aes(x = Ont, y = Gene.Count, fill=forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE))) + # MF plot
                               geom_bar(color = "white",size=1.5,stat="identity") + # call bar and create lines between each slim term
                               scale_fill_manual(values =  rev( colorRampPalette(GreenBlue)( (nrow(MF_All)) ) ) )+ # fill with pallette accomodating the max nmber of slim terms
                               geom_text(aes(label = forcats::fct_reorder(slim_term,Gene.Count,.desc = TRUE)), colour = "black", position="stack", size = 7, vjust = 1.25) + # add labels
                               theme_classic() + # classic theme
                               ylim(0, 100) +
                               theme(legend.position = "none",text = element_text(size=25)) +
                               facet_wrap(~Ontolgy) # facet by the upregulated and downregulated datasets

# SAVE PLOTS  ------------------------------------------------------------------------------------------------- #
pdf(paste("../Output/GO/goseq/GOslim_All_PrimEff_BP.pdf", sep =''), width=15, height=20)
print(ggarrange(BP_Plot_PrimEffect_All,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()     

pdf(paste("../Output/GO/goseq/GOslim_All_PrimEff_MF.pdf", sep =''), width=15, height=20)
print(ggarrange(MF_Plot_PrimEffect_All,         
                plotlist = NULL,
                ncol = 1,
                nrow = 1,
            labels = NULL))
dev.off()     

```



Follow-up from the stacked blots above...
What are the genesin the GOslimss of interest
Extract genes in GOslim bins of interest and plots their LFC results (DESeq2)
```{r} 
# prep data for the for loop plotting...
# add column 'DE.direction'
All.DOWN_MFslim$DE.direction  <- "downregulated"
All.DOWN_BPslim$DE.direction  <- "downregulated"
All.UP_MFslim$DE.direction    <- "upregulated"
All.UP_BPslim$DE.direction    <- "upregulated"

# add column 'Experiment.note'
All.DOWN_MFslim$Experiment.note  <- "All_Days_PrimaryEfffect"
All.DOWN_BPslim$Experiment.note  <- "All_Days_PrimaryEfffect"
All.UP_MFslim$Experiment.note    <- "All_Days_PrimaryEfffect"
All.UP_BPslim$Experiment.note    <- "All_Days_PrimaryEfffect"

# ommit rownames
rownames(All.DOWN_MFslim)<-NULL
rownames(All.DOWN_BPslim)<-NULL
rownames(All.UP_MFslim)  <-NULL
rownames(All.UP_BPslim)  <-NULL

# merge data frames to build a master GOslim
Master_GOslim <- rbind(All.DOWN_MFslim[,c(10,7,9,8,3,4,5)], All.DOWN_BPslim[,c(10,7,9,8,3,4,5)], All.UP_MFslim[,c(10,7,9,8,3,4,5)], All.UP_BPslim[,c(10,7,9,8,3,4,5)]) # merge for a master table with rbind

# load DESeq2 results 
All_DEGs_Primary.Ambieant_v_Moderate<-read.csv("C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/DESeq2/10cpm/DE_PrimaryTreatment.All.csv")
All_DEGs_Primary.Ambieant_v_Moderate <- All_DEGs_Primary.Ambieant_v_Moderate[2:(ncol(All_DEGs_Primary.Ambieant_v_Moderate))] # ommit the first column - DESEq2 results data 
names(All_DEGs_Primary.Ambieant_v_Moderate)[1] <- 'Gene.ID' # rename the column 

Pgen_annot <- Geoduck_annotation[,c(1,7)]# call only the gene ID and gene names in the annotation csv
names(Pgen_annot) <- c('Gene.ID', 'Gene.name') # rename the columns 
Pgen_annot$Gene.name <- str_extract(Pgen_annot$Gene.name, "[^(]+") # call all gene names before the delimiter ( when the EC code is included)

# loop through the gene IDsin each GOslim bin, DESeq2 results heatmap facetted by sampling day in for loop
for (i in 1:nrow(Master_GOslim)) {
  
  genes <- data.frame(geneids = unlist(Master_GOslim[i,1]))
  
  if (Master_GOslim[i,4] == "downregulated") {
  
    DEG_data <- All_DEGs_Primary.Ambieant_v_Moderate %>% dplyr::filter(down %in% TRUE)  # filter only downregualted genes
    DEG_data_loop <- DEG_data %>% dplyr::filter(Gene.ID %in% genes$geneids)
    DEG_data_merge <- merge(Pgen_annot, DEG_data_loop, by ="Gene.ID")
    
    DEG_data_merge$Day_Trmt <- str_extract(DEG_data_merge$Day_Trmt, "[^_]+")
    
    Plot <- DEG_data_merge %>% #mutate(Day_Trmt = fct_relevel(Day_Trmt,  "Day7", "Day14", "Day21")) %>% 
              ggplot(aes(x = "NA", y = Gene.name)) + 
                geom_tile(aes(fill=log2FoldChange, width = 1)) + 
                scale_fill_gradient(low = "orangered3", high = "orange1") +
                facet_grid(~Day_Trmt, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                strip.text.y = element_text(angle=0, size = 11, face = "bold"),
                strip.text.x = element_text(size = 12, face = "bold"),
                axis.title.x = element_blank(),
                axis.title.y = element_text(size=15),
                axis.text = element_text(size = 12), legend.position = "right",
                plot.margin = unit(c(0,1,0,0.25), "cm"))+
                ggtitle(Master_GOslim[i,5])
    pdf(paste("../Output/GO/goseq/All_PrimEff_GOslim/Downregulated/",Master_GOslim[i,5],".pdf", sep =''), width=18, height=10)
    print(Plot)
    dev.off()
  
    } else {
        DEG_data <- All_DEGs_Primary.Ambieant_v_Moderate %>% dplyr::filter(up %in% TRUE)  # filter only downregualted genes
        DEG_data_loop <- DEG_data %>% dplyr::filter(Gene.ID %in% genes$geneids)
        DEG_data_merge <- merge(Pgen_annot, DEG_data_loop, by ="Gene.ID")
        
        DEG_data_merge$Day_Trmt <- str_extract(DEG_data_merge$Day_Trmt, "[^_]+")
        
        Plot <- DEG_data_merge %>% #mutate(Day_Trmt = fct_relevel(Day_Trmt,  "Day7", "Day14", "Day21")) %>% 
                  ggplot(aes(x = "NA", y = Gene.name)) + 
                    geom_tile(aes(fill=log2FoldChange, width = 1)) + 
                    scale_fill_gradient(low = "palegreen1", high = "green4") +
                    facet_grid( ~Day_Trmt, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
                    theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
                    panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
                    strip.text.y = element_text(angle=0, size = 11, face = "bold"),
                    strip.text.x = element_text(size = 12, face = "bold"),
                    axis.title.x = element_blank(),
                    axis.title.y = element_text(size=15),
                    axis.text = element_text(size = 12), legend.position = "right",
                    plot.margin = unit(c(0,1,0,0.25), "cm"))+
                    ggtitle(Master_GOslim[i,5])
        pdf(paste("../Output/GO/goseq/All_PrimEff_GOslim/Upregulated/",Master_GOslim[i,5],".pdf", sep =''), width=18, height=10)
        print(Plot)
        dev.off()
  } # end of if else statement

} # end of for loop


```



Save slim info with GO enrichment info for heatmap dataframes.
```{r}
ALL_Upreg_GO.BP <- right_join(All.UP_BPslim_final, filter(all.UP.05, ontology=="BP"), by="category") #add back GO enrichment info for each offspring term
ALL_Upreg_GO.BP <- na.omit(ALL_Upreg_GO.BP)
ALL_Upreg_GO.MF <- right_join(All.UP_MFslim_final, filter(all.UP.05, ontology=="MF"), by="category") #add back GO enrichment info for each offspring term

d14.Upreg_GO.BP <- right_join(d14.UP_BPslim, filter(d14.UP.05, ontology=="BP"), by="category") #add back GO enrichment info for each offspring term
d14.Upreg_GO.MF <- right_join(d14.UP_MFslim, filter(d14.UP.05, ontology=="MF"), by="category") #add back GO enrichment info for each offspring term


```



Make heatmap
```{r}
All_BPplot <- ALL_Upreg_GO.BP %>% 
  # filter(numInCat>5) %>% 
  ggplot(aes(x = dir, y = term)) + 
  geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
  facet_grid(slim_term ~ ontology, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
  strip.text.y = element_text(angle=0, size = 11, face = "bold"),
  strip.text.x = element_text(size = 12, face = "bold"),
  axis.title.x = element_blank(),
  axis.title.y = element_text(size=15),
  axis.text = element_text(size = 12), legend.position = "None",
  plot.margin = unit(c(0,1,0,0.25), "cm"))

ggsave("../Output/GO/goseq/All.PrimaryEffect.Upreg_BP.pdf", All_BPplot, width = 25, height = 45, units = c("in"))

All_MFplot <- ALL_Upreg_GO.MF %>% 
  filter(numInCat>5) %>% 
  ggplot(aes(x = dir, y = term)) + 
  geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
  facet_grid(slim_term ~ ontology, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
  strip.text.y = element_text(angle=0, size = 11, face = "bold"),
  strip.text.x = element_text(size = 12, face = "bold"),
  axis.title.x = element_blank(),
  axis.title.y = element_text(size=15),
  axis.text = element_text(size = 12), legend.position = "None",
  plot.margin = unit(c(0,1,0,0.25), "cm"))

ggsave("../Output/GO/goseq/All.PrimaryEffect.Upreg_MF.pdf", All_MFplot, width = 25, height = 35, units = c("in"))


# DAY 14
# Upregulated
Day14_BPplot <- d7.UP_BPslim %>% filter(numInCat>5) %>% ggplot(aes(x = dir, y = term)) + 
  geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
  facet_grid(slim_term ~ ontology, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
  strip.text.y = element_text(angle=0, size = 11, face = "bold"),
  strip.text.x = element_text(size = 12, face = "bold"),
  axis.title.x = element_blank(),
  axis.title.y = element_text(size=15),
  axis.text = element_text(size = 12), legend.position = "None",
  plot.margin = unit(c(0,1,0,0.25), "cm"))

ggsave("../Output/GO/goseq/Day14/Day14.PrimaryEffect.Upreg_BP.pdf", Day14_BPplot, width = 25, height = 45, units = c("in"))

Day14_MFplot <- d7.UP_MFslim %>% filter(numInCat>5) %>% ggplot(aes(x = dir, y = term)) + 
  geom_tile(aes(fill=over_represented_pvalue, width = 1)) + 
  facet_grid(slim_term ~ ontology, scales = "free_y", labeller = label_wrap_gen(width = 10, multi_line = TRUE))+
  theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(),
  panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"),
  strip.text.y = element_text(angle=0, size = 11, face = "bold"),
  strip.text.x = element_text(size = 12, face = "bold"),
  axis.title.x = element_blank(),
  axis.title.y = element_text(size=15),
  axis.text = element_text(size = 12), legend.position = "None",
  plot.margin = unit(c(0,1,0,0.25), "cm"))

ggsave("../Output/GO/goseq/Day14/Day14.PrimaryEffect.Upreg_MF.pdf", Day14_MFplot, width = 25, height = 35, units = c("in"))


```



```{r Constant.primary.GO table}
GO_Constant.UP_summary <- GO_Constant.UP %>% dplyr::select(c('category','numDEInCat','numInCat', 'term', 'ontology')) %>% dplyr::filter(!is.na(term)) %>% dplyr::filter(numDEInCat > 0)
GO_Constant.UP_summary$DE_dir <- "upregulated"
GO_Constant.DOWN_summary <- GO_Constant.DOWN %>% dplyr::select(c('category','numDEInCat','numInCat', 'term', 'ontology')) %>% dplyr::filter(!is.na(term)) %>% dplyr::filter(numDEInCat > 0)
GO_Constant.DOWN_summary$DE_dir <- "downregulated"

GO_constant.Master <- rbind(GO_Constant.UP_summary,GO_Constant.DOWN_summary)
GO_constant.Master$perc.occurance <- GO_constant.Master$numDEInCat / GO_constant.Master$numInCat *100

#knitr
knitr::kable(GO_constant.Master, caption = "Summary table of constant GO terms Days 7, 14, 21")
# Write csv
write.csv(GO_constant.Master,paste(path_out,"Constant.Primary_GO.terms.csv"))
```



Plotting: All Primary Treatmentt effects together! 
```{r All Upreg and Downreg GO terms in response to priming}

# call enriched GO terms and plot 
#  UPREGULATED GENES! 
# day 7 UPREGULATED
GO_d7.UP.05.a<-GO_d7.UP$category[GO_d7.UP$over_represented_pvalue<.05] # change twice here
GO_d7.UP.05<-data.frame(GO_d7.UP.05.a)
colnames(GO_d7.UP.05) <- c("category")
GO_d7.UP.05 <- merge(GO_d7.UP.05, GO_d7.UP, by="category") # change here
GO_d7.UP.05 <- GO_d7.UP.05[order(-GO_d7.UP.05$numDEInCat),]
GO_d7.UP.05$term <- as.factor(GO_d7.UP.05$term)

d7_UP_MF <- GO_d7.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d7_UP'))
d7_UP_BP <- GO_d7.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d7_UP'))

# day 14 UPREGULATED
GO_d14.UP.05.a<-GO_d14.UP$category[GO_d14.UP$over_represented_pvalue<.05] # change twice here
GO_d14.UP.05<-data.frame(GO_d14.UP.05.a)
colnames(GO_d14.UP.05) <- c("category")
GO_d14.UP.05 <- merge(GO_d14.UP.05, GO_d14.UP, by="category") # change here
GO_d14.UP.05 <- GO_d14.UP.05[order(-GO_d14.UP.05$numDEInCat),]
GO_d14.UP.05$term <- as.factor(GO_d14.UP.05$term)

d14_UP_MF <- GO_d14.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d14_UP'))
d14_UP_BP <- GO_d14.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d14_UP'))

# day 21 UPREGULATED
GO_d21.UP.05.a<-GO_d21.UP$category[GO_d21.UP$over_represented_pvalue<.05] # change twice here
GO_d21.UP.05<-data.frame(GO_d21.UP.05.a)
colnames(GO_d21.UP.05) <- c("category")
GO_d21.UP.05 <- merge(GO_d21.UP.05, GO_d21.UP, by="category") # change here
GO_d21.UP.05 <- GO_d21.UP.05[order(-GO_d21.UP.05$numDEInCat),]
GO_d21.UP.05$term <- as.factor(GO_d21.UP.05$term)

d21_UP_MF <- GO_d21.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d21_UP'))
d21_UP_BP <- GO_d21.UP.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_UP', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d21_UP'))


#======================================================================= #
# bind the rows of MF and BP in the clustered modules (those twith primary treatment effect in same pattern/directionality)
MF_UP <- rbind(d7_UP_MF, d14_UP_MF, d21_UP_MF) 
BP_UP <- rbind(d7_UP_BP, d14_UP_BP, d21_UP_BP)




MF_UP_plot  <- MF_UP %>% 
                          mutate(term = fct_reorder(term, ontology)) %>%
                          ggplot(aes(factor(ontology,level = c('MF_d7_UP', 'MF_d14_UP', 'MF_d21_UP')), term, fill= factor(ontology), alpha = over_represented_pvalue)) + 
                          geom_tile(fill = '#00BFC4') +
                          scale_alpha_continuous(range = c(1, 0.3)) +
                          theme_classic2() + 
                              ggtitle("GO Mol Fxn: Priming effect (> Exp in AMBIENT history)") +
                              xlab("GO.ontology_Sampling.Day") +
                              ylab('') +
                              theme(legend.position="none") 
BP_UP_plot <- BP_UP %>% 
                          mutate(term = fct_reorder(term, ontology)) %>%
                          ggplot(aes(factor(ontology,level = c('BP_d7_UP', 'BP_d14_UP', 'BP_d21_UP')), term, fill= factor(ontology), alpha = over_represented_pvalue)) + 
                          geom_tile(fill = '#F8766D') +
                          scale_alpha_continuous(range = c(1, 0.3)) +
                          theme_classic2() + 
                              ggtitle("GO Biol Proc: Priming effect (> Exp in AMBIENT history)") +
                              xlab("GO.ontology_Sampling.Day") +
                              ylab('') +
                              theme(legend.position="none") 










#  DOWNREGULATED GENES! 
# day 7 DOWNREGULATED
GO_d7.DOWN.05.a<-GO_d7.DOWN$category[GO_d7.DOWN$over_represented_pvalue<.05] # change twice here
GO_d7.DOWN.05<-data.frame(GO_d7.DOWN.05.a)
colnames(GO_d7.DOWN.05) <- c("category")
GO_d7.DOWN.05 <- merge(GO_d7.DOWN.05, GO_d7.DOWN, by="category") # change here
GO_d7.DOWN.05 <- GO_d7.DOWN.05[order(-GO_d7.DOWN.05$numDEInCat),]
GO_d7.DOWN.05$term <- as.factor(GO_d7.DOWN.05$term)

d7_DOWN_MF <- GO_d7.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d7_DOWN'))
d7_DOWN_BP <- GO_d7.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd7_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d7_DOWN'))

# day 14 DOWNREGULATED
GO_d14.DOWN.05.a<-GO_d14.DOWN$category[GO_d14.DOWN$over_represented_pvalue<.05] # change twice here
GO_d14.DOWN.05<-data.frame(GO_d14.DOWN.05.a)
colnames(GO_d14.DOWN.05) <- c("category")
GO_d14.DOWN.05 <- merge(GO_d14.DOWN.05, GO_d14.DOWN, by="category") # change here
GO_d14.DOWN.05 <- GO_d14.DOWN.05[order(-GO_d14.DOWN.05$numDEInCat),]
GO_d14.DOWN.05$term <- as.factor(GO_d14.DOWN.05$term)

d14_DOWN_MF <- GO_d14.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d14_DOWN'))
d14_DOWN_BP <- GO_d14.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd14_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d14_DOWN'))

# day 21 DOWNREGULATED
GO_d21.DOWN.05.a<-GO_d21.DOWN$category[GO_d21.DOWN$over_represented_pvalue<.05] # change twice here
GO_d21.DOWN.05<-data.frame(GO_d21.DOWN.05.a)
colnames(GO_d21.DOWN.05) <- c("category")
GO_d21.DOWN.05 <- merge(GO_d21.DOWN.05, GO_d21.DOWN, by="category") # change here
GO_d21.DOWN.05 <- GO_d21.DOWN.05[order(-GO_d21.DOWN.05$numDEInCat),]
GO_d21.DOWN.05$term <- as.factor(GO_d21.DOWN.05$term)

d21_DOWN_MF <- GO_d21.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('MF_d21_DOWN'))
d21_DOWN_BP <- GO_d21.DOWN.05 %>% 
  drop_na(ontology) %>% 
  mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) ),
         ontology = paste(ontology, 'd21_DOWN', sep = "_")) %>%
  dplyr::filter(ontology %in% ('BP_d21_DOWN'))

#======================================================================= #
# bind the rows of MF and BP in the clustered modules (those twith primary treatment effect in same pattern/directionality)
MF_DOWN <- rbind(d7_DOWN_MF, d14_DOWN_MF, d21_DOWN_MF) 
BP_DOWN <- rbind(d7_DOWN_BP, d14_DOWN_BP, d21_DOWN_BP)


MF_DOWN_plot  <- MF_DOWN %>% 
                          mutate(term = fct_reorder(term, ontology)) %>%
                          ggplot(aes(factor(ontology,level = c('MF_d7_DOWN', 'MF_d14_DOWN', 'MF_d21_DOWN')), term, fill= factor(ontology), alpha = over_represented_pvalue)) + 
                          geom_tile(fill = '#00BFC4') +
                          scale_alpha_continuous(range = c(1, 0.3)) +
                          theme_classic2() + 
                              ggtitle("GO Mol Fxn: Priming effect (> Exp in MODERATE history)") +
                              xlab("GO.ontology_Sampling.Day") +
                              ylab('') +
                              theme(legend.position="none") 
BP_DOWN_plot <- BP_DOWN %>% 
                          mutate(term = fct_reorder(term, ontology)) %>%
                          ggplot(aes(factor(ontology,level = c('BP_d7_DOWN', 'BP_d14_DOWN', 'BP_d21_DOWN')), term, fill= factor(ontology), alpha = over_represented_pvalue)) + 
                          geom_tile(fill = '#F8766D') +
                          scale_alpha_continuous(range = c(1, 0.3)) +
                          theme_classic2() + 
                              ggtitle("GO Biol Proc: Priming effect (> Exp in MODERATE history)") +
                              xlab("GO.ontology_Sampling.Day") +
                              ylab('') +
                              theme(legend.position="none") 
getwd()
pdf(paste("../Output/GO/goseq/GOplot_DESeq2_PrimEffect_Upregulated.pdf", sep =''), width=30, height=25)
print(ggarrange(MF_UP_plot, BP_UP_plot,         
                plotlist = NULL,
                ncol = 2,
                nrow = 1,
            labels = NULL))
dev.off()      


pdf(paste("../Output/GO/goseq/GOplot_DESeq2_PrimEffect_Downregulated.pdf", sep =''), width=30, height=25)
print(ggarrange(MF_DOWN_plot, BP_DOWN_plot,         
                plotlist = NULL,
                ncol = 2,
                nrow = 1,
            labels = NULL))
dev.off()      


```

Plotting: Day 0
```{r plots_Day0.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d0.UP$category[GO_d0.UP$over_represented_pvalue<.05] # change twice here
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d0.UP, by="category") # change here
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- # 
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d0.DOWN$category[GO_d0.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d0.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d0.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d0.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 0 Upregulated DEGs") +
  geom_label(aes(x = 0.5, y = 0.5, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 0 Downregulated DEGs") +
  geom_label(aes(x = 0.5, y = 1.5, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d0.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day0/GOplot_d0.Primary.pdf", GO.plot_d0.Primary, width = 21, height = 21, units = c("in"))
getwd()
```

Plotting: Day 7
```{r plots_Day7.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d7.UP$category[GO_d7.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d7.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d7.DOWN$category[GO_d7.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d7.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d7.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d7.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 7 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 2, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 7 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 2, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d7.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day7/GOplot_d7.Primary.pdf", GO.plot_d7.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Day 14
```{r plots_Day14.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d14.UP$category[GO_d14.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d14.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d14.DOWN$category[GO_d14.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d14.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d14.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d14.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 14 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 20, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 14 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 6, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d14.Primary<- grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day14/GOplot_d14.Primary.pdf", GO.plot_d14.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Day 21
```{r plots_Day21.Primary}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_d21.UP$category[GO_d21.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_d21.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
MFplot_UP <- UP_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_UP <- UP_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_UP <- UP_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_d21.DOWN$category[GO_d21.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_d21.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Processes
MFplot_DOWN <- DOWN_MF %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Molecular Function") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("Cellular Component") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, color="#69b3a2") +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="none") +
  xlab("") +
  ylab("") +
  ggtitle("Biological Process") + #add a main title
  theme(plot.title = element_text(face = 'bold', 
                                  size = 12, 
                                  hjust = 0)) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank())#Set the plot background

# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(DE_d21.primary.UPREG)[1] # call num upregulated genes
num_down <-dim(DE_d21.primary.DWNREG)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 21 Upregulated DEGs") +
  geom_label(aes(x = 2, y = 7, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Day 21 Downregulated DEGs") +
  geom_label(aes(x = 2, y = 3, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_d21.Primary<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("../Output/GO/goseq/Day21/GOplot_d21.Primary.pdf", GO.plot_d21.Primary, width = 21, height = 21, units = c("in"))
```

Plotting: Constant DEGs Days 7 14 21 affected by Primary treatment history
```{r Constant_All}
#How many enriched GO terms do we have
# class(GO_d14.all)
# head(GO_d14.all)
# tail(GO.wall)
# nrow(GO.wall)

# UPREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
UP_enriched.GO.05.a<-GO_Constant.UP$category[GO_Constant.UP$over_represented_pvalue<.05]
UP_enriched.GO.05<-data.frame(UP_enriched.GO.05.a)
colnames(UP_enriched.GO.05) <- c("category")
UP_enriched.GO.05 <- merge(UP_enriched.GO.05, GO_Constant.UP, by="category")
UP_enriched.GO.05 <- UP_enriched.GO.05[order(-UP_enriched.GO.05$numDEInCat),]
UP_enriched.GO.05$term <- as.factor(UP_enriched.GO.05$term)
head(UP_enriched.GO.05)

UP_MF <- subset(UP_enriched.GO.05, ontology=="MF")
UP_MF <- UP_MF[order(-UP_MF$numDEInCat),]
UP_CC <- subset(UP_enriched.GO.05, ontology=="CC")
UP_CC <- UP_CC[order(-UP_CC$numDEInCat),]
UP_BP <- subset(UP_enriched.GO.05, ontology=="BP")
UP_BP <- UP_BP[order(-UP_BP$numDEInCat),]

# Molecular Processes
head(UP_MF)

MFplot_UP <- UP_MF %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#00BFC4') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Mol Fxn: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

#Cellular Processes
CCplot_UP <- UP_CC %>% 
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#7CAE00') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Cell Comp: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

# Biological Processes 
BPplot_UP <- UP_BP %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#69b3a2') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Biol Proc: Constant DEGs UPREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 


# DOWNREGULATED DEGs ------------------------------------------------------------------------------- #
#Find only enriched GO terms that are statistically significant at cutoff 
DOWN_enriched.GO.05.a<-GO_Constant.DOWN$category[GO_Constant.DOWN$over_represented_pvalue<.05]
DOWN_enriched.GO.05<-data.frame(DOWN_enriched.GO.05.a)
colnames(DOWN_enriched.GO.05) <- c("category")
DOWN_enriched.GO.05 <- merge(DOWN_enriched.GO.05, GO_Constant.DOWN, by="category")
DOWN_enriched.GO.05 <- DOWN_enriched.GO.05[order(-DOWN_enriched.GO.05$numDEInCat),]
DOWN_enriched.GO.05$term <- as.factor(DOWN_enriched.GO.05$term)
head(DOWN_enriched.GO.05)

DOWN_MF <- subset(DOWN_enriched.GO.05, ontology=="MF")
DOWN_MF <- DOWN_MF[order(-DOWN_MF$numDEInCat),]
DOWN_CC <- subset(DOWN_enriched.GO.05, ontology=="CC")
DOWN_CC <- DOWN_CC[order(-DOWN_CC$numDEInCat),]
DOWN_BP <- subset(DOWN_enriched.GO.05, ontology=="BP")
DOWN_BP <- DOWN_BP[order(-DOWN_BP$numDEInCat),]

# Molecular Function
MFplot_DOWN <- DOWN_MF %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#00BFC4') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Mol Fxn: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

#Cellular Processes
CCplot_DOWN <- DOWN_CC %>% 
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#7CAE00') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Cell Comp: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 

# Biological Processes 
BPplot_DOWN <- DOWN_BP %>%
              mutate(term = fct_reorder(term, (-log10(over_represented_pvalue)) )) %>%
              ggplot(aes(ontology, term, fill= ontology, alpha = over_represented_pvalue)) + 
              geom_tile(fill = '#69b3a2') +
              scale_alpha_continuous(range = c(1, 0.3)) +
              theme_classic2() + 
              ggtitle("GO Biol Proc: Constant DEGs DOWNREGULATED") +
              xlab("") +
              ylab('') +
              theme(legend.position="none") 


# GOplot <- grid.arrange(MFplot, CCplot, BPplot, ncol=3, clip="off")
# ggsave("GOplot.pdf", GOplot, width = 21, height = 21, units = c("in"))

# Merge plots!
num_up   <- dim(Contant_upreg)[1] # call num upregulated genes
num_down <-dim(Contant_dwnreg)[1] # call num downregulated genes

GOplot.merge_UP <- UP_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Constant Upregulated DEGs") +
  geom_label(aes(x = 1, y = 3, label = paste(num_up, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_UP

GOplot.merge_DOWN <- DOWN_enriched.GO.05 %>% drop_na(ontology) %>% mutate(term = fct_reorder(term, numDEInCat)) %>%
  mutate(term = fct_reorder(term, ontology)) %>%
  ggplot( aes(x=term, y=numDEInCat) ) +
  geom_segment( aes(x=term ,xend=term, y=0, yend=numDEInCat), color="grey") +
  geom_point(size=3, aes(colour = ontology)) +
  coord_flip() +
  theme(
    panel.grid.minor.y = element_blank(),
    panel.grid.major.y = element_blank(),
    legend.position="bottom"
  ) +
  xlab("") +
  ylab("") +
  ggtitle("GO: Constant Downregulated DEGs") +
  geom_label(aes(x = 1, y = 3, label = paste(num_down, "DEGs"))) +
  theme_bw() + #Set background color 
  theme(panel.border = element_blank(), # Set border
        panel.grid.major = element_blank(), #Set major gridlines
        panel.grid.minor = element_blank(), #Set minor gridlines
        axis.line = element_line(colour = "black"), #Set axes color
        plot.background=element_blank()) #Set the plot background #set title attributes
# GOplot.merge_DOWN

GO.plot_Constant<-grid.arrange(GOplot.merge_UP, GOplot.merge_DOWN, nrow=2, clip="off")
ggsave("C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Output/GO/GOplot_Constant.Primary.pdf", GO.plot_Constant, width = 21, height = 21, units = c("in"))
```















### GO_MWU - not using this for the paper but did a bit of work here... Can review it for future data
Create all files for the GO_MWU.R script
output two-column csv of just the gene.id and log2fc 
```{r GO_MWU files - Primary_Treatment}
DE_d0.P <- data.frame(DE_d0.primary[,c(2,4)])
write.csv(DE_d0.P,"C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d0.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d7.P <- data.frame(DE_d7.primary[,c(2,4)])
write.csv(DE_d7.P,"C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d7.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d14.P <- data.frame(DE_d14.primary[,c(2,4)])
write.csv(DE_d14.P,"C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d14.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

DE_d21.P <- data.frame(DE_d21.primary[,c(2,4)])
write.csv(DE_d21.P,"C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/GO_MWU/d21.P.DE_GO.terms.csv",row.names=FALSE,quote=FALSE)

```

### Genewalk

link to tutorial: https://churchman.med.harvard.edu/genewalk/#tutorial
Citation (bioRxiv preprint): Ietswaart, R., Gyori, B. M., Bachman, J. A., Sorger, P. K., & Churchman, L. S. (2019). GeneWalk identifies relevant gene functions for a biological context using network representation learning. bioRxiv, 755579.
About: Genewalk is curently designed to perform function network analysis of DEGs with human and mouse identifiers

```{r prep genewalk files}
Geoduck.gene_HGNC.list <- Geoduck_annotation %>% dplyr::select(c("V1","V6")) # call the geoduck gene ID and the HGNC in the annotation file
colnames(Geoduck.gene_HGNC.list) <- c("gene.ID", "HGNC_symbol") # rename columns
# note that the HGNC column contains the HNC code with the model taxa as 'HUMAN, MOUSE, CHICK, etc.'
Geoduck.gene_HGNC.list$HGNC<- sapply(strsplit((Geoduck.gene_HGNC.list$HGNC), "_"), '[', 1) # this splits the HGNC by the delimiter "_" and calls just the HGNC code - note the model taxa is now absent from the dataframe! 


# change the col name of the gene.ID  in the DE datasets

# DAY 0 ------------------------------------------- #

# UPREGULATED

GW.d0.primary.UPREG <- merge((DE_d0.primary.UPREG[order(DE_d0.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d0.upreg.percHGNC <- ((dim(na.omit(GW.d0.primary.UPREG))[1]) / (dim(DE_d0.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d0.primary.DWNREG <- merge((DE_d0.primary.DWNREG[order(DE_d0.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d0.downreg.percHGNC <- ((dim(na.omit(GW.d0.primary.DWNREG))[1]) / (dim(DE_d0.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 7 ------------------------------------------- #

# UPREGULATED
GW.d7.primary.UPREG <- merge((DE_d7.primary.UPREG[order(DE_d7.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d7.upreg.percHGNC <- ((dim(na.omit(GW.d7.primary.UPREG))[1]) / (dim(DE_d7.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d7.primary.DWNREG <- merge((DE_d7.primary.DWNREG[order(DE_d7.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d7.downreg.percHGNC <- ((dim(na.omit(GW.d7.primary.DWNREG))[1]) / (dim(DE_d7.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 14 ------------------------------------------- #

# UPREGULATED
GW.d14.primary.UPREG <- merge((DE_d14.primary.UPREG[order(DE_d14.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d14.upreg.percHGNC <- ((dim(na.omit(GW.d14.primary.UPREG))[1]) / (dim(DE_d14.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d14.primary.DWNREG <- merge((DE_d14.primary.DWNREG[order(DE_d14.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC") # sort to order by greatest logfold change and merge with HGNC list
d14.downreg.percHGNC <- ((dim(na.omit(GW.d14.primary.DWNREG))[1]) / (dim(DE_d14.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID

# DAY 21 ------------------------------------------- #

# UPREGULATED
GW.d21.primary.UPREG <- merge((DE_d21.primary.UPREG[order(DE_d21.primary.UPREG$log2FoldChange, decreasing = TRUE),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d21.upreg.percHGNC <- ((dim(na.omit(GW.d21.primary.UPREG))[1]) / (dim(DE_d21.primary.UPREG)[1]) )*100 # percent of DEGs with HGNC ID

# DOWNREGULATED
GW.d21.primary.DWNREG <- merge((DE_d21.primary.DWNREG[order(DE_d21.primary.DWNREG$log2FoldChange),]), Geoduck.gene_HGNC.list, by = "gene.ID") %>% dplyr::select("HGNC")
d21.downreg.percHGNC <- ((dim(na.omit(GW.d21.primary.DWNREG))[1]) / (dim(DE_d21.primary.DWNREG)[1]) )*100 # percent of DEGs with HGNC ID


### OUTPUT Genewalk tables as csv wihtout header
GW.d0.primary.UPREG.om   <- na.omit(GW.d0.primary.UPREG)
GW.d0.primary.DWNREG.om  <- na.omit(GW.d0.primary.DWNREG)

GW.d7.primary.UPREG.om   <- na.omit(GW.d7.primary.UPREG)
GW.d7.primary.DWNREG.om  <- na.omit(GW.d7.primary.DWNREG) 

GW.d14.primary.UPREG.om  <- na.omit(GW.d14.primary.UPREG)
GW.d14.primary.DWNREG.om <- na.omit(GW.d14.primary.DWNREG)

GW.d21.primary.UPREG.om  <- na.omit(GW.d21.primary.UPREG)
GW.d21.primary.DWNREG.om <- na.omit(GW.d21.primary.DWNREG)

path_genewalk='C:/Users/samjg/Documents/My_Projects/Pgenerosa_TagSeq_Metabolomics/TagSeq/Analysis/Data/GO/Genewalk_files/' # call path
# output files 
write.table(GW.d0.primary.UPREG.om,paste(path_genewalk,"Day0_upreg.csv", sep =''),sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d0.primary.DWNREG.om, paste(path_genewalk,"Day0_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)

write.table(GW.d7.primary.UPREG.om,paste(path_genewalk,"Day7_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d7.primary.DWNREG.om,paste(path_genewalk,"Day7_downreg.csv", sep =''), sep=",", col.names=FALSE, row.names=FALSE, quote=FALSE)

write.table(GW.d14.primary.UPREG.om,paste(path_genewalk,"Day14_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d14.primary.DWNREG.om,paste(path_genewalk,"Day14_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)

write.table(GW.d21.primary.UPREG.om,paste(path_genewalk,"Day21_upreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
write.table(GW.d21.primary.DWNREG.om,paste(path_genewalk,"Day21_downreg.csv", sep =''), sep=",", col.names=FALSE,row.names=FALSE, quote=FALSE)
```

```{r Genewalk_knitr_table}
GW_summarytable <- data.frame(matrix(nrow = 4, ncol = 8)) # create a new data table
colnames(GW_summarytable)<-c('Day', 'Treatment', 'DEGs_upreg', 'numHGNC_upreg', 'percHGNC_upreg', 'DEGs_dwnreg', 'numHGNC_dwnreg', 'percHGNC_dwnreg') 
GW_summarytable$Day <- c("0","7","14","21")
GW_summarytable$Treatment <- "Ambient_v_Moderate"
GW_summarytable$DEGs_upreg <- c((dim(DE_d0.primary.UPREG)[1]), (dim(DE_d7.primary.UPREG)[1]), (dim(DE_d14.primary.UPREG)[1]), (dim(DE_d21.primary.UPREG)[1]))
GW_summarytable$numHGNC_upreg <- c((dim(na.omit(GW.d0.primary.UPREG))[1]),(dim(na.omit(GW.d7.primary.UPREG))[1]),(dim(na.omit(GW.d14.primary.UPREG))[1]),(dim(na.omit(GW.d21.primary.UPREG))[1]))
GW_summarytable$percHGNC_upreg <- c(d0.upreg.percHGNC,d7.upreg.percHGNC,d14.upreg.percHGNC,d21.upreg.percHGNC)
GW_summarytable$DEGs_dwnreg <- c((dim(DE_d0.primary.DWNREG)[1]), (dim(DE_d7.primary.DWNREG)[1]), (dim(DE_d14.primary.DWNREG)[1]), (dim(DE_d21.primary.DWNREG)[1]))
GW_summarytable$numHGNC_dwnreg<- c((dim(na.omit(GW.d0.primary.DWNREG))[1]),(dim(na.omit(GW.d7.primary.DWNREG))[1]),(dim(na.omit(GW.d14.primary.DWNREG))[1]),(dim(na.omit(GW.d21.primary.DWNREG))[1]))
GW_summarytable$percHGNC_dwnreg <- c(d0.downreg.percHGNC,d7.downreg.percHGNC,d14.downreg.percHGNC,d21.downreg.percHGNC)

knitr::kable(GW_summarytable, caption = "Summary table of Genewalk files - total DEGs and mapping abund HGNC IDs")
# Write csv
write.csv(GW_summarytable,paste(path_genewalk,"Summary_HGNC_hits.csv"))
```

